{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6e1a6c02",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "import re\n",
    "os.getcwd()\n",
    "from bs4 import BeautifulSoup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5661be5c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPIC_Guideline_for_abacavir_and_HLA_B\n",
      "CPIC_Guideline_for_aceclofenac_aspirin_diclofenac_dipyrone_indomethacin_lumiracoxib_nabumetone_naproxen_and_CYP2C9\n",
      "CPIC_Guideline_for_alfentanil_buprenorphine_codeine_fentanyl_hydrocodone_hydromorphone_levomethadone_morphine_naltrexone_remifentanil_sufentanil_tramadol_and_COMT_OPRM1\n",
      "CPIC_Guideline_for_allopurinol_and_HLA_B\n",
      "CPIC_Guideline_for_amikacin_gentamicin_kanamycin_paromomycin_plazomicin_streptomycin_tobramycin_and_MT_RNR1\n",
      "CPIC_Guideline_for_amitriptyline_and_CYP2C19_CYP2D6\n",
      "CPIC_Guideline_for_atazanavir_and_UGT1A1\n",
      "CPIC_Guideline_for_atomoxetine_and_CYP2D6\n",
      "CPIC_Guideline_for_azathioprine_and_NUDT15_TPMT\n",
      "CPIC_Guideline_for_capecitabine_and_DPYD\n",
      "CPIC_Guideline_for_carbamazepine_and_HLA_A_HLA_B\n",
      "CPIC_Guideline_for_celecoxib_flurbiprofen_ibuprofen_lornoxicam_and_CYP2C9\n",
      "CPIC_Guideline_for_citalopram_escitalopram_and_CYP2C19\n",
      "CPIC_Guideline_for_clomipramine_and_CYP2C19_CYP2D6\n",
      "CPIC_Guideline_for_clopidogrel_and_CYP2C19\n",
      "CPIC_Guideline_for_codeine_and_CYP2D6\n",
      "CPIC_Guideline_for_desflurane_enflurane_halothane_isoflurane_methoxyflurane_sevoflurane_succinylcholine_and_CACNA1S_RYR1\n",
      "CPIC_Guideline_for_desipramine_and_CYP2D6\n",
      "CPIC_Guideline_for_dexlansoprazole_and_CYP2C19\n",
      "CPIC_Guideline_for_diclofenac_ibuprofen_and_CYP2C8\n",
      "CPIC_Guideline_for_doxepin_and_CYP2C19_CYP2D6\n",
      "CPIC_Guideline_for_efavirenz_and_CYP2B6\n",
      "CPIC_Guideline_for_esomeprazole_rabeprazole_and_CYP2C19\n",
      "CPIC_Guideline_for_fluorouracil_and_DPYD\n",
      "CPIC_Guideline_for_fluvoxamine_and_CYP2D6\n",
      "CPIC_Guideline_for_fosphenytoin_phenytoin_and_CYP2C9_HLA_B\n",
      "CPIC_Guideline_for_hydrocodone_and_CYP2D6\n",
      "CPIC_Guideline_for_imipramine_and_CYP2C19_CYP2D6\n",
      "CPIC_Guideline_for_ivacaftor_and_CFTR\n",
      "CPIC_Guideline_for_lansoprazole_omeprazole_pantoprazole_and_CYP2C19\n",
      "CPIC_Guideline_for_meloxicam_and_CYP2C9\n",
      "CPIC_Guideline_for_mercaptopurine_and_NUDT15_TPMT\n",
      "CPIC_Guideline_for_methadone_oxycodone_and_COMT_CYP2D6_OPRM1\n",
      "CPIC_Guideline_for_nortriptyline_and_CYP2D6\n",
      "CPIC_Guideline_for_ondansetron_and_CYP2D6\n",
      "CPIC_Guideline_for_oxcarbazepine_and_HLA_B\n",
      "CPIC_Guideline_for_paroxetine_and_CYP2D6\n",
      "CPIC_Guideline_for_peginterferon_alfa_2a_peginterferon_alfa_2b_ribavirin_and_IFNL3\n",
      "CPIC_Guideline_for_piroxicam_and_CYP2C9\n",
      "CPIC_Guideline_for_rasburicase_and_G6PD\n",
      "CPIC_Guideline_for_sertraline_and_CYP2C19\n",
      "CPIC_Guideline_for_simvastatin_and_SLCO1B1\n",
      "CPIC_Guideline_for_tacrolimus_and_CYP3A5\n",
      "CPIC_Guideline_for_tamoxifen_and_CYP2D6\n",
      "CPIC_Guideline_for_tegafur_and_DPYD\n",
      "CPIC_Guideline_for_tenoxicam_and_CYP2C9\n",
      "CPIC_Guideline_for_thioguanine_and_NUDT15_TPMT\n",
      "CPIC_Guideline_for_tramadol_and_CYP2D6\n",
      "CPIC_Guideline_for_trimipramine_and_CYP2C19_CYP2D6\n",
      "CPIC_Guideline_for_tropisetron_and_CYP2D6\n",
      "CPIC_Guideline_for_voriconazole_and_CYP2C19\n",
      "CPIC_Guideline_for_warfarin_and_CYP2C9_CYP4F2_VKORC1\n",
      "CPNDS_Guideline_for_carbamazepine_and_HLA_A\n",
      "CPNDS_Guideline_for_carbamazepine_and_HLA_B\n",
      "CPNDS_Guideline_for_cisplatin_and_TPMT\n",
      "CPNDS_Guideline_for_codeine_and_CYP2D6\n",
      "CPNDS_Guideline_for_daunorubicin_doxorubicin_and_RARG_SLC28A3_UGT1A6\n",
      "CPNDS_Guideline_for_tamoxifen_and_CYP2D6\n",
      "CPNDS_Guideline_for_warfarin_and_CYP2C9\n",
      "DPWG_Guideline_for_abacavir_and_HLA_B\n",
      "DPWG_Guideline_for_acenocoumarol_and_CYP2C9\n",
      "DPWG_Guideline_for_acenocoumarol_and_VKORC1\n",
      "DPWG_Guideline_for_amiodarone_and_CYP2D6\n",
      "DPWG_Guideline_for_amitriptyline_and_CYP2C19\n",
      "DPWG_Guideline_for_amitriptyline_and_CYP2D6\n",
      "DPWG_Guideline_for_aripiprazole_and_CYP2D6\n",
      "DPWG_Guideline_for_atenolol_and_CYP2D6\n",
      "DPWG_Guideline_for_atomoxetine_and_CYP2D6\n",
      "DPWG_Guideline_for_atorvastatin_and_SLCO1B1\n",
      "DPWG_Guideline_for_azathioprine_and_NUDT15\n",
      "DPWG_Guideline_for_azathioprine_and_TPMT\n",
      "DPWG_Guideline_for_bisoprolol_and_CYP2D6\n",
      "DPWG_Guideline_for_brexpiprazole_and_CYP2D6\n",
      "DPWG_Guideline_for_capecitabine_and_DPYD\n",
      "DPWG_Guideline_for_carvedilol_and_CYP2D6\n",
      "DPWG_Guideline_for_citalopram_and_CYP2C19\n",
      "DPWG_Guideline_for_citalopram_escitalopram_and_CYP2D6\n",
      "DPWG_Guideline_for_clomipramine_and_CYP2C19\n",
      "DPWG_Guideline_for_clomipramine_and_CYP2D6\n",
      "DPWG_Guideline_for_clonidine_and_CYP2D6\n",
      "DPWG_Guideline_for_clopidogrel_and_CYP2C19\n",
      "DPWG_Guideline_for_clozapine_and_CYP2D6\n",
      "DPWG_Guideline_for_codeine_and_CYP2D6\n",
      "DPWG_Guideline_for_disopyramide_and_CYP2D6\n",
      "DPWG_Guideline_for_doxepin_and_CYP2C19\n",
      "DPWG_Guideline_for_doxepin_and_CYP2D6\n",
      "DPWG_Guideline_for_duloxetine_and_CYP2D6\n",
      "DPWG_Guideline_for_efavirenz_and_CYP2B6\n",
      "DPWG_Guideline_for_eliglustat_and_CYP2D6\n",
      "DPWG_Guideline_for_escitalopram_and_CYP2C19\n",
      "DPWG_Guideline_for_esomeprazole_and_CYP2C19\n",
      "DPWG_Guideline_for_flecainide_and_CYP2D6\n",
      "DPWG_Guideline_for_flucloxacillin_and_HLA_B\n",
      "DPWG_Guideline_for_flucytosine_and_DPYD\n",
      "DPWG_Guideline_for_fluorouracil_and_DPYD\n",
      "DPWG_Guideline_for_fluoxetine_and_CYP2D6\n",
      "DPWG_Guideline_for_flupenthixol_and_CYP2D6\n",
      "DPWG_Guideline_for_fluphenazine_and_CYP2D6\n",
      "DPWG_Guideline_for_fluvastatin_and_SLCO1B1\n",
      "DPWG_Guideline_for_fluvoxamine_and_CYP2C19\n",
      "DPWG_Guideline_for_fluvoxamine_and_CYP2D6\n",
      "DPWG_Guideline_for_gefitinib_and_CYP2D6\n",
      "DPWG_Guideline_for_glibenclamide_and_CYP2C9\n",
      "DPWG_Guideline_for_gliclazide_and_CYP2C9\n",
      "DPWG_Guideline_for_glimepiride_and_CYP2C9\n",
      "DPWG_Guideline_for_haloperidol_and_CYP2D6\n",
      "DPWG_Guideline_for_imipramine_and_CYP2C19\n",
      "DPWG_Guideline_for_imipramine_and_CYP2D6\n",
      "DPWG_Guideline_for_irinotecan_and_UGT1A1\n",
      "DPWG_Guideline_for_lansoprazole_and_CYP2C19\n",
      "DPWG_Guideline_for_mercaptopurine_and_NUDT15\n",
      "DPWG_Guideline_for_mercaptopurine_and_TPMT\n",
      "DPWG_Guideline_for_methylphenidate_and_CYP2D6\n",
      "DPWG_Guideline_for_metoprolol_and_CYP2D6\n",
      "DPWG_Guideline_for_mirtazapine_and_CYP2C19\n",
      "DPWG_Guideline_for_mirtazapine_and_CYP2D6\n",
      "DPWG_Guideline_for_moclobemide_and_CYP2C19\n",
      "DPWG_Guideline_for_nortriptyline_and_CYP2D6\n",
      "DPWG_Guideline_for_olanzapine_and_CYP2D6\n",
      "DPWG_Guideline_for_omeprazole_and_CYP2C19\n",
      "DPWG_Guideline_for_oxycodone_and_CYP2D6\n",
      "DPWG_Guideline_for_pantoprazole_and_CYP2C19\n",
      "DPWG_Guideline_for_paroxetine_and_CYP2D6\n",
      "DPWG_Guideline_for_phenprocoumon_and_CYP2C9\n",
      "DPWG_Guideline_for_phenprocoumon_and_VKORC1\n",
      "DPWG_Guideline_for_phenytoin_and_CYP2C9\n",
      "DPWG_Guideline_for_pimozide_and_CYP2D6\n",
      "DPWG_Guideline_for_prasugrel_and_CYP2C19\n",
      "DPWG_Guideline_for_propafenone_and_CYP2D6\n",
      "DPWG_Guideline_for_quetiapine_and_CYP2D6\n",
      "DPWG_Guideline_for_quinidine_and_CYP2D6\n",
      "DPWG_Guideline_for_rabeprazole_and_CYP2C19\n",
      "DPWG_Guideline_for_ribavirin_and_HLA_B\n",
      "DPWG_Guideline_for_risperidone_and_CYP2D6\n",
      "DPWG_Guideline_for_sertraline_and_CYP2C19\n",
      "DPWG_Guideline_for_sertraline_and_CYP2D6\n",
      "DPWG_Guideline_for_simvastatin_and_SLCO1B1\n",
      "DPWG_Guideline_for_siponimod_and_CYP2C9\n",
      "DPWG_Guideline_for_sotalol_and_CYP2D6\n",
      "DPWG_Guideline_for_tacrolimus_and_CYP3A5\n",
      "DPWG_Guideline_for_tamoxifen_and_CYP2D6\n",
      "DPWG_Guideline_for_tegafur_and_DPYD\n",
      "DPWG_Guideline_for_thioguanine_and_NUDT15\n",
      "DPWG_Guideline_for_thioguanine_and_TPMT\n",
      "DPWG_Guideline_for_ticagrelor_and_CYP2C19\n",
      "DPWG_Guideline_for_tolbutamide_and_CYP2C9\n",
      "DPWG_Guideline_for_tramadol_and_CYP2D6\n",
      "DPWG_Guideline_for_venlafaxine_and_CYP2D6\n",
      "DPWG_Guideline_for_voriconazole_and_CYP2C19\n",
      "DPWG_Guideline_for_warfarin_and_CYP2C9\n",
      "DPWG_Guideline_for_warfarin_and_VKORC1\n",
      "DPWG_Guideline_for_zuclopenthixol_and_CYP2D6\n",
      "PRO_Guideline_for_acenocoumarol_fluindione_warfarin_and_CYP2C9_VKORC1\n",
      "PRO_Guideline_for_allopurinol_and_HLA_B\n",
      "PRO_Guideline_for_antidepressants_and_CYP2C19_CYP2D6\n",
      "PRO_Guideline_for_azathioprine_and_TPMT\n",
      "PRO_Guideline_for_capecitabine_fluorouracil_and_DPYD\n",
      "PRO_Guideline_for_clopidogrel_and_CYP2C19\n",
      "PRO_Guideline_for_hmg_coa_reductase_inhibitors_simvastatin_and_SLCO1B1\n",
      "PRO_Guideline_for_irinotecan_and_UGT1A1\n",
      "PRO_Guideline_for_ivacaftor_and_CFTR\n",
      "PRO_Guideline_for_mercaptopurine_and_TPMT\n",
      "PRO_Guideline_for_methotrexate_and_ABCB1_MTHFR_SLC19A1_SLCO1B1\n",
      "PRO_Guideline_for_tacrolimus_and_CYP3A4_CYP3A5\n",
      "PRO_Guideline_for_tamoxifen_and_CYP2D6\n",
      "(5440, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>guideline_institute</th>\n",
       "      <th>guideline_name</th>\n",
       "      <th>drug</th>\n",
       "      <th>gene</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for abacavir and ...</td>\n",
       "      <td>abacavir</td>\n",
       "      <td>HLA_B</td>\n",
       "      <td>HLA-B*57:01:01</td>\n",
       "      <td>Presence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for allopurinol a...</td>\n",
       "      <td>allopurinol</td>\n",
       "      <td>HLA_B</td>\n",
       "      <td>HLA-B*58:01</td>\n",
       "      <td>Presence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 663A&gt;G</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 669T&gt;C</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 747A&gt;G</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 786G&gt;A</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 807A&gt;C</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 807A&gt;G</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 827A&gt;G</td>\n",
       "      <td>Normal risk of aminoglycoside-induced hearing ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 839A&gt;G</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  guideline_institute                                     guideline_name  \\\n",
       "0      CPIC_Guideline  Annotation of CPIC Guideline for abacavir and ...   \n",
       "1      CPIC_Guideline  Annotation of CPIC Guideline for allopurinol a...   \n",
       "2      CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "3      CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "4      CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "5      CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "6      CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "7      CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "8      CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "9      CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "\n",
       "          drug     gene       haplotype  \\\n",
       "0     abacavir    HLA_B  HLA-B*57:01:01   \n",
       "1  allopurinol    HLA_B     HLA-B*58:01   \n",
       "2     amikacin  MT_RNR1  MT-RNR1 663A>G   \n",
       "3     amikacin  MT_RNR1  MT-RNR1 669T>C   \n",
       "4     amikacin  MT_RNR1  MT-RNR1 747A>G   \n",
       "5     amikacin  MT_RNR1  MT-RNR1 786G>A   \n",
       "6     amikacin  MT_RNR1  MT-RNR1 807A>C   \n",
       "7     amikacin  MT_RNR1  MT-RNR1 807A>G   \n",
       "8     amikacin  MT_RNR1  MT-RNR1 827A>G   \n",
       "9     amikacin  MT_RNR1  MT-RNR1 839A>G   \n",
       "\n",
       "                                                term  \n",
       "0                                           Presence  \n",
       "1                                           Presence  \n",
       "2  Uncertain risk of aminoglycoside-induced heari...  \n",
       "3  Uncertain risk of aminoglycoside-induced heari...  \n",
       "4  Uncertain risk of aminoglycoside-induced heari...  \n",
       "5  Uncertain risk of aminoglycoside-induced heari...  \n",
       "6  Uncertain risk of aminoglycoside-induced heari...  \n",
       "7  Uncertain risk of aminoglycoside-induced heari...  \n",
       "8  Normal risk of aminoglycoside-induced hearing ...  \n",
       "9  Uncertain risk of aminoglycoside-induced heari...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folder = \"guideline_annotation\"\n",
    "guideline_list = []\n",
    "haplotype_list = []\n",
    "term_list = []\n",
    "drug_list = []\n",
    "gene_list = []\n",
    "guideline_name_list = []\n",
    "\n",
    "for path in os.listdir(folder):\n",
    "    if path.endswith(\".json\") and \"Annotation_of_\" in path:\n",
    "        with open(os.path.join(folder, path), \"r\", encoding=\"utf-8\") as f:\n",
    "            guideline_dict = json.load(f)\n",
    "\n",
    "        path = path.replace(\"Annotation_of_\", \"\").replace(\".json\", \"\")\n",
    "        try:\n",
    "            guideline, drug_gene = path.split(\"_for_\")\n",
    "        except:\n",
    "            continue\n",
    "        drugs, genes = drug_gene.split(\"_and_\")\n",
    "        drugs = list(filter(lambda x: x!= \"\", [x.strip() for x in drugs.split(\"_\")]))\n",
    "\n",
    "        tmp_haplotype_list = []\n",
    "        tmp_term_list = []\n",
    "        for guideline_gene in guideline_dict[\"guideline\"][\"guidelineGenes\"]:\n",
    "            for allele in guideline_gene[\"alleles\"]:\n",
    "                if \"haplotype\" in allele.keys():\n",
    "                    tmp_haplotype_list.append(allele[\"haplotype\"][\"symbol\"])\n",
    "                    tmp_term_list.append(allele[\"function\"][\"term\"])\n",
    "        \n",
    "        guideline_name = guideline_dict[\"guideline\"][\"name\"]\n",
    "        \n",
    "        guideline_name_list.extend([guideline_name] * len(tmp_haplotype_list) * len(drugs))\n",
    "        guideline_list.extend([guideline] * len(tmp_haplotype_list) * len(drugs))\n",
    "        gene_list.extend([genes] * len(tmp_haplotype_list) * len(drugs))\n",
    "        # data alignment\n",
    "        for drug in drugs:\n",
    "            drug_list.extend([drug] * len(tmp_haplotype_list))\n",
    "            haplotype_list.extend(tmp_haplotype_list)\n",
    "            term_list.extend(tmp_term_list)\n",
    "        print(path)\n",
    "        \n",
    "\n",
    "df_guideline = pd.DataFrame({\n",
    "    \"guideline_institute\": guideline_list,\n",
    "    \"guideline_name\": guideline_name_list,\n",
    "    \"drug\": drug_list,\n",
    "    \"gene\": gene_list,\n",
    "    \"haplotype\": haplotype_list,\n",
    "    \"term\": term_list\n",
    "})\n",
    "\n",
    "print(df_guideline.shape)\n",
    "df_guideline[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0ff7206f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'DPYD', 'CYP3A5', 'UGT1A1', 'TPMT', 'CACNA1S', 'RYR1', 'HLA_B', 'CYP2C9', 'CFTR', 'IFNL3', 'MT_RNR1', 'CYP2C19', 'CYP2D6', 'NUDT15', 'SLCO1B1', 'HLA_A', 'CYP2B6', 'G6PD'}\n"
     ]
    }
   ],
   "source": [
    "print(set(['CYP2C19', 'CYP2D6', 'NUDT15', 'TPMT', 'CYP2C9', 'CFTR', 'CYP2D6', 'G6PD',\n",
    "           'SLCO1B1', 'UGT1A1', 'DPYD', 'HLA_B', 'HLA_A', 'HLA_B', 'CYP2C9', 'HLA_B',\n",
    "           'NUDT15', 'CYP3A5', 'IFNL3', 'CYP2C19', 'CYP2B6', 'MT_RNR1', 'CACNA1S', 'RYR1']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "56e658f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_guideline.assign(\n",
    "    data_source=[\"guideline\"] * len(df_guideline)\n",
    ").to_csv(\"processed/guideline_drug_variant_annotation.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9b93e4f7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7d7614df",
   "metadata": {},
   "outputs": [],
   "source": [
    "html = \"<h3 id=\\\"november-2018-update\\\">November 2018 Update</h3>\\n<p>The Royal Dutch Pharmacists Association - Pharmacogenetics Working Group has evaluated therapeutic dose recommendations for aripiprazole based on CYP2D6 genotype. They recommend reducing maximum dose of aripiprazole for CYP2D6 poor metabolizer (PM).</p>\\n<p><em>Wording in table taken from the <a href=\\\"https://api.pharmgkb.org/v1/download/file/attachment/DPWG_November_2018.pdf\\\">Dutch guidelines November 2018 update</a></em></p>\\n<table class=\\\"table\\\">\\n<thead>\\n<tr><th>Allele/Genotype/Phenotype</th><th>Drug</th><th>Description</th><th>Recommendation</th></tr>\\n</thead>\\n<tbody>\\n<tr><td>CYP2D6 UM</td><td>aripiprazole</td><td>-</td><td>NO action is needed for this gene-drug interaction. The genetic variation decreases the plasma concentration of the sum of aripiprazole and the active metabolite dehydroaripiprazole to a limited degree. There is no evidence that this increases the risk of reduced effectiveness.</td></tr>\\n<tr><td>CYP2D6 IM</td><td>aripiprazole</td><td>-</td><td>NO action is needed for this gene-drug interaction.The genetic variation increases the plasma concentration of the sum of aripiprazole and the active metabolite dehydroaripiprazole to a limited degree. There is insufficient evidence that this increases the risk of side effects.</td></tr>\\n<tr><td>CYP2D6 PM</td><td>aripiprazole</td><td>The risk of side effects is increased. The genetic variation leads to an increase in the sum of the plasma concentrations of aripiprazole and the active metabolite.</td><td>Administer no more than 10 mg/day or 300 mg/month (67-75% of the standard maximum dose of aripiprazole).</td></tr>\\n</tbody>\\n</table>\\n<p><a rel=\\\"noopener noreferrer\\\" href=\\\"https://www.g-standaard.nl/risicoanalyse/B0001543.PDF\\\" target=\\\"_blank\\\">Read for more information about this recommendation</a>, <a rel=\\\"noopener noreferrer\\\" href=\\\"https://www.knmp.nl/producten/gebruiksrecht-g-standaard/medicatiebewaking-g-standaard/background-information-pharmacogenetics\\\" target=\\\"_blank\\\">Read about gene information from DPWG</a></p>\\n<h3 id=\\\"2011-guideline\\\">2011 Guideline</h3>\\n<p>The Royal Dutch Pharmacists Association - Pharmacogenetics Working Group (DPWG) has evaluated therapeutic dose recommendations for aripiprazole based on CYP2D6 genotypes [Article:<a href=\\\"/pmid/21412232\\\">21412232</a>].  They recommend reducing maximum dose of aripiprazole for patients carrying poor metabolizer alleles of CYP2D6.</p>\\n<table class=\\\"table\\\">\\n<thead>\\n<tr><th>Phenotype (Genotype)</th><th>Therapeutic Dose Recommendation</th><th>Level of Evidence</th><th>Clinical Relevance</th></tr>\\n</thead>\\n<tbody>\\n<tr><td>PM (two inactive (*3-*8, *11-*16, *19-*21, *38, *40, *42) alleles)</td><td>Reduce maximum dose to 10 mg/day (67% of the maximum recommended daily dose).</td><td>Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints.</td><td>Clinical effect (S): long-standing discomfort (48-168 hr) without permanent injury e.g. failure of therapy with tricyclic antidepressants, atypical antipsychotic drugs; extrapyramidal side effects; parkinsonism; ADE resulting from increased bioavailability of tricyclic antidepressants, metoprolol, propafenone (central effects e.g. dizziness); INR 4.5-6.0; neutropenia 1.0-1.5x109/l; leucopenia 2.0-3.0x109/l; thrombocytopenia 50-75x109/l.</td></tr>\\n<tr><td>IM (two decreased-activity (*9, *10, *17, *29, *36, *41) alleles or carrying one active (*1, *2, *33, *35) and one inactive (*3-*8, *11-*16, *19-*21, *38, *40, *42) allele, or carrying one decreased-activity (*9, *10, *17, *29, *36, *41) allele and one inactive (*3-*8, *11-*16, *19-*21, *38, *40, *42) allele)</td><td>No recommendations.</td><td>Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints.</td><td>Minor clinical effect (S): QTc prolongation (&lt;450 ms female, &lt;470 ms male); INR increase &lt; 4.5; Kinetic effect (S)</td></tr>\\n<tr><td>UM (a gene duplication in absence of inactive (*3-*8, *11-*16, *19-*21, *38, *40, *42) or decreased-activity (*9, *10, *17, *29, *36, *41) alleles)</td><td>No recommendations.</td><td>--</td><td>--</td></tr>\\n</tbody>\\n</table>\\n<ul>\\n<li>*See <a href=\\\"http://www.pharmgkb.org/page/dpwg\\\">Methods </a> or [Article:<a href=\\\"/pmid/18253145\\\">18253145</a>] for definition of &quot;good quality.&quot;</li>\\n<li>S: statistically significant difference.</li>\\n<li>Please see attached PDF for detailed information about the evaluated studies: <a href=\\\"https://api.pharmgkb.org/v1/download/file/attachment/aripiprazole_CYP2D6_271111.pdf\\\">Aripiprazole CYP2D6 </a></li>\\n</ul>\\n\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0091fb57",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1ee9f324",
   "metadata": {},
   "outputs": [],
   "source": [
    "soup = BeautifulSoup(html, \"html.parser\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8dc6c090",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_header = []\n",
    "header = soup.find_all(\"table\")[0].find(\"tr\")\n",
    "for items in header:\n",
    "    try:\n",
    "        list_header.append(items.get_text())\n",
    "    except:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "44334479",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Allele/Genotype/Phenotype', 'Drug', 'Description', 'Recommendation']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ab4c8f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "HTML_data = soup.find_all(\"table\")[0].find_all(\"tr\")[1:]\n",
    "data = []\n",
    "\n",
    "for element in HTML_data:\n",
    "    sub_data = []\n",
    "    for sub_element in element:\n",
    "        try:\n",
    "            sub_data.append(sub_element.get_text())\n",
    "        except:\n",
    "            continue\n",
    "    data.append(sub_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "170c5cc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['CYP2D6 UM',\n",
       "  'aripiprazole',\n",
       "  '-',\n",
       "  'NO action is needed for this gene-drug interaction. The genetic variation decreases the plasma concentration of the sum of aripiprazole and the active metabolite dehydroaripiprazole to a limited degree. There is no evidence that this increases the risk of reduced effectiveness.'],\n",
       " ['CYP2D6 IM',\n",
       "  'aripiprazole',\n",
       "  '-',\n",
       "  'NO action is needed for this gene-drug interaction.The genetic variation increases the plasma concentration of the sum of aripiprazole and the active metabolite dehydroaripiprazole to a limited degree. There is insufficient evidence that this increases the risk of side effects.'],\n",
       " ['CYP2D6 PM',\n",
       "  'aripiprazole',\n",
       "  'The risk of side effects is increased. The genetic variation leads to an increase in the sum of the plasma concentrations of aripiprazole and the active metabolite.',\n",
       "  'Administer no more than 10 mg/day or 300 mg/month (67-75% of the standard maximum dose of aripiprazole).']]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6cb45a0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele/Genotype/Phenotype</th>\n",
       "      <th>Drug</th>\n",
       "      <th>Description</th>\n",
       "      <th>Recommendation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CYP2D6 UM</td>\n",
       "      <td>aripiprazole</td>\n",
       "      <td>-</td>\n",
       "      <td>NO action is needed for this gene-drug interac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CYP2D6 IM</td>\n",
       "      <td>aripiprazole</td>\n",
       "      <td>-</td>\n",
       "      <td>NO action is needed for this gene-drug interac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CYP2D6 PM</td>\n",
       "      <td>aripiprazole</td>\n",
       "      <td>The risk of side effects is increased. The gen...</td>\n",
       "      <td>Administer no more than 10 mg/day or 300 mg/mo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Allele/Genotype/Phenotype          Drug  \\\n",
       "0                 CYP2D6 UM  aripiprazole   \n",
       "1                 CYP2D6 IM  aripiprazole   \n",
       "2                 CYP2D6 PM  aripiprazole   \n",
       "\n",
       "                                         Description  \\\n",
       "0                                                  -   \n",
       "1                                                  -   \n",
       "2  The risk of side effects is increased. The gen...   \n",
       "\n",
       "                                      Recommendation  \n",
       "0  NO action is needed for this gene-drug interac...  \n",
       "1  NO action is needed for this gene-drug interac...  \n",
       "2  Administer no more than 10 mg/day or 300 mg/mo...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(data = data, columns = list_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0a8ec0c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_tables(html, gene, drug, organization, link):\n",
    "    # save columns which contain these pharses\n",
    "    column_template = [\"phenotype\", \"diplotype\", \"genotype\", \"implication\", \"description\", \"recommendation\", \"dosing\"]\n",
    "    soup = BeautifulSoup(html, \"html.parser\")\n",
    "    tables = soup.find_all(\"table\")\n",
    "    titles = soup.find_all(\"h3\")\n",
    "    df_list = []\n",
    "    for i in range(len(tables)):\n",
    "        try:\n",
    "            table = tables[i]\n",
    "            title = titles[i].get_text()\n",
    "        except Exception as e:\n",
    "            print(\"missing title, gene {}, drug {}, organization {}\".format(gene, drug, organization))\n",
    "            continue\n",
    "        list_header = []\n",
    "        header = table.find(\"tr\")\n",
    "        for items in header:\n",
    "            try:\n",
    "                t = items.get_text()\n",
    "                if t == \"\\n\":\n",
    "                    continue\n",
    "                list_header.append(t)\n",
    "            except:\n",
    "                continue\n",
    "\n",
    "        HTML_data = table.find_all(\"tr\")[1:]\n",
    "        data = []\n",
    "        for element in HTML_data:\n",
    "            sub_data = []\n",
    "            for sub_element in element:\n",
    "                try:\n",
    "                    t = sub_element.get_text()\n",
    "                    if t == \"\\n\":\n",
    "                        continue\n",
    "                    sub_data.append(t)\n",
    "                except:\n",
    "                    continue\n",
    "            data.append(sub_data)\n",
    "\n",
    "        try:\n",
    "            df_tmp = pd.DataFrame(data=data, columns=list_header)\n",
    "            save_column = []\n",
    "            for column in df_tmp.columns:\n",
    "                if any([True if x in column.lower() else False for x in column_template]):\n",
    "                    save_column.append(column)\n",
    "            df_tmp = df_tmp[save_column]\n",
    "            \n",
    "            df_tmp = df_tmp.assign(\n",
    "                gene=[gene] * len(df_tmp)\n",
    "            ).assign(\n",
    "                drug=[drug] * len(df_tmp)\n",
    "            ).assign(\n",
    "                organization=[organization] * len(df_tmp)\n",
    "            ).assign(\n",
    "                title=[title] * len(df_tmp)\n",
    "            ).assign(\n",
    "                link=[link] * len(df_tmp)\n",
    "            )\n",
    "            df_list.append(df_tmp)\n",
    "        except Exception as e:\n",
    "            print(\"parsing html failed, gene {}, drug {}, organization {}\".format(gene, drug, organization))\n",
    "            continue\n",
    "    return df_list\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fdee11e7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "missing title, gene CYP2D6, drug tamoxifen, organization CPNDS_Guideline\n",
      "missing title, gene CYP2D6, drug haloperidol, organization DPWG_Guideline\n",
      "missing title, gene CYP2D6, drug imipramine, organization DPWG_Guideline\n",
      "missing title, gene CYP2D6, drug metoprolol, organization DPWG_Guideline\n",
      "missing title, gene CYP2D6, drug nortriptyline, organization DPWG_Guideline\n",
      "missing title, gene CYP2D6, drug paroxetine, organization DPWG_Guideline\n",
      "missing title, gene CYP2D6, drug propafenone, organization DPWG_Guideline\n",
      "missing title, gene CYP2C9_VKORC1, drug acenocoumarol_fluindione_warfarin, organization PRO_Guideline\n"
     ]
    }
   ],
   "source": [
    "tables = []\n",
    "folder = \"guideline_annotation\"\n",
    "for path in os.listdir(folder):\n",
    "    if path.endswith(\".json\") and \"Annotation_of_\" in path:\n",
    "        with open(os.path.join(folder, path), \"r\", encoding=\"utf-8\") as f:\n",
    "            guideline_dict = json.load(f)\n",
    "            \n",
    "        path = path.replace(\"Annotation_of_\", \"\").replace(\".json\", \"\")\n",
    "        try:\n",
    "            guideline, drug_gene = path.split(\"_for_\")\n",
    "        except:\n",
    "            continue\n",
    "        drugs, genes = drug_gene.split(\"_and_\")\n",
    "        html = guideline_dict[\"guideline\"][\"textMarkdown\"][\"html\"]\n",
    "        link = guideline_dict[\"guideline\"][\"@id\"]\n",
    "        \n",
    "        tables.extend(parse_tables(guideline_dict[\"guideline\"][\"textMarkdown\"][\"html\"], \n",
    "                                   gene=genes, drug=drugs, organization=guideline, link=link))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbd91738",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e8d0f78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['phenotype',\n",
       " 'diplotype',\n",
       " 'genotype',\n",
       " 'implication',\n",
       " 'description',\n",
       " 'recommendation']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[\"phenotype\", \"diplotype\", \"genotype\", \"implication\", \"description\", \"recommendation\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "de4c86fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_category_list = []\n",
    "phenotype_list = []\n",
    "diplotype_list = []\n",
    "genotype_list = []\n",
    "implication_list = []\n",
    "description_list = []\n",
    "recommendation_list = []\n",
    "dosing_list = []\n",
    "gene_list = []\n",
    "drug_list = []\n",
    "organization_list = []\n",
    "title_list = []\n",
    "link_list = []\n",
    "for table in tables:\n",
    "    phenotype_column = \"\"\n",
    "    diplotype_column = \"\"\n",
    "    genotype_column = \"\"\n",
    "    implication_column = \"\"\n",
    "    description_column = \"\"\n",
    "    recommendation_column = \"\"\n",
    "    dosing_column = \"\"\n",
    "    for col in table.columns:\n",
    "        if \"phenotype\" in col.lower():\n",
    "            phenotype_column = col\n",
    "        if \"diplotype\" in col.lower():\n",
    "            diplotype_column = col\n",
    "        if \"genotype\" in col.lower():\n",
    "            genotype_column = col\n",
    "        if \"implication\" in col.lower():\n",
    "            implication_column = col\n",
    "        if \"description\" in col.lower():\n",
    "            description_column = col\n",
    "        if \"recommendation\" in col.lower() and \"classification\" not in col.lower():\n",
    "            recommendation_column = col\n",
    "        if \"dosing\" in col.lower():\n",
    "            dosing_column = col\n",
    "            \n",
    "    for index, row in table.iterrows():\n",
    "        if phenotype_column != \"\":\n",
    "            phenotype = row[phenotype_column]\n",
    "        else:\n",
    "            continue\n",
    "        \n",
    "        if diplotype_column != \"\":\n",
    "            diplotype = row[diplotype_column]\n",
    "        else:\n",
    "            diplotype = \"\"\n",
    "        \n",
    "        if genotype_column != \"\":\n",
    "            genotype = row[genotype_column]\n",
    "        else:\n",
    "            genotype = \"\"\n",
    "\n",
    "        if implication_column != \"\":\n",
    "            implication = row[implication_column]\n",
    "        else:\n",
    "            implication = \"\"\n",
    "            \n",
    "        if description_column != \"\":\n",
    "            description = row[description_column]\n",
    "        else:\n",
    "            description = \"\"\n",
    "            \n",
    "        if recommendation_column != \"\":\n",
    "            recommendation = row[recommendation_column]\n",
    "        else:\n",
    "            recommendation = \"\"\n",
    "            \n",
    "        if dosing_column != \"\":\n",
    "            dosing = row[dosing_column]\n",
    "        else:\n",
    "            dosing = \"\"\n",
    "        \n",
    "        gene = row[\"gene\"]\n",
    "        drug = row[\"drug\"]\n",
    "        organization = row[\"organization\"]\n",
    "        title = row[\"title\"]\n",
    "\n",
    "        if any([True if x in phenotype.lower() and \"meta\" in phenotype.lower() else False \\\n",
    "                for x in [\"intermediate\", \"normal\", \"poor\", \"rapid\", \"ultrarapid\", \"extensive\"]]) or \\\n",
    "           any(True if x in phenotype else False \\\n",
    "                for x in [\"IM\", \"PM\", \"UM\", \"NM\", \"RM\", \"EM\"]):\n",
    "            pass\n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "        if genotype == \"\" and implication == \"\" and description == \"\" and recommendation == \"\" and dosing == \"\":\n",
    "            continue\n",
    "            \n",
    "        if \"_\" in gene:\n",
    "            genes = gene.split(\"_\")\n",
    "            filtered_gene = list(filter(lambda x: x in phenotype, genes))\n",
    "            if len(filtered_gene) > 0:\n",
    "                gene = filtered_gene[0] \n",
    "                \n",
    "        if \"Drug\" in table.columns and row[\"Drug\"] != \"\":\n",
    "            drug = row[\"Drug\"]\n",
    "        elif \"_\" in drug:\n",
    "            drugs = drug.split(\"_\")\n",
    "            filtered_drug = list(filter(lambda x: x in phenotype or x in recommendation or x in implication, drugs))\n",
    "            if len(filtered_drug) > 0:\n",
    "                drug = filtered_drug[0] \n",
    "        \n",
    "        phenotype_category = \"\"\n",
    "        if \"poor\" in phenotype.lower() or \"PM\" in phenotype:\n",
    "            phenotype_category = \"Poor Metabolizer\"\n",
    "        elif \"intermediate\" in phenotype.lower() or \"IM\" in phenotype:\n",
    "            phenotype_category = \"Intermediate metabolizer\"\n",
    "        elif \"normal\" in phenotype.lower() or \"NM\" in phenotype:\n",
    "            phenotype_category = \"Normal metabolizer\"\n",
    "        elif (\"rapid\" in phenotype.lower() and \"ultra\" not in phenotype.lower()) or \"RM\" in phenotype:\n",
    "            phenotype_category = \"Rapid metabolizer\"\n",
    "        elif (\"rapid\" in phenotype.lower() and \"ultra\" in phenotype.lower()) or \"UM\" in phenotype:\n",
    "            phenotype_category = \"Ultrarapid metabolizer\"\n",
    "        elif \"extensive\" in phenotype.lower() or \"EM\" in phenotype:\n",
    "            phenotype_category = \"Extensive metabolizer\"\n",
    "            \n",
    "        phenotype_category_list.append(phenotype_category)\n",
    "        phenotype_list.append(phenotype)\n",
    "        diplotype_list.append(diplotype)\n",
    "        genotype_list.append(genotype)\n",
    "        implication_list.append(implication)\n",
    "        description_list.append(description)\n",
    "        recommendation_list.append(recommendation)\n",
    "        dosing_list.append(dosing)\n",
    "        gene_list.append(gene)\n",
    "        drug_list.append(drug)\n",
    "        organization_list.append(organization)\n",
    "        title_list.append(title)\n",
    "        link_list.append(row[\"link\"])\n",
    "            \n",
    "df_pheno_drug = pd.DataFrame({\n",
    "    \"phenotype_category\": phenotype_category_list,\n",
    "    \"phenotype\": phenotype_list,\n",
    "    \"diplotype\": diplotype_list,\n",
    "    \"genotype\": genotype_list,\n",
    "    \"implication\": implication_list,\n",
    "    \"description\": description_list,\n",
    "    \"recommendation\": recommendation_list,\n",
    "    \"dosing\": dosing_list,\n",
    "    \"gene\": gene_list,\n",
    "    \"drug\": drug_list,\n",
    "    \"organization\": organization_list,\n",
    "    \"title\": title_list,\n",
    "    \"link\": link_list,\n",
    "    \"data_source\": [\"guildeline\"] * len(link_list)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "17c19998",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for diplotype\n",
    "phenotype_category_list = [] \n",
    "phenotype_list = []\n",
    "diplotype_list = []\n",
    "genotype_list = []\n",
    "implication_list = []\n",
    "description_list = []\n",
    "recommendation_list = []\n",
    "gene_list = []\n",
    "drug_list = []\n",
    "organization_list = []\n",
    "title_list = []\n",
    "link_list = []\n",
    "\n",
    "for index, row in df_pheno_drug.iterrows():\n",
    "    diplotype = row[\"diplotype\"]\n",
    "    diplotypes = re.split(\",|;| \", diplotype)\n",
    "    diplotypes = list(filter(lambda x: x != \"\" and x.count(\"*\") == 2, diplotypes))\n",
    "    if len(diplotypes) == 0:\n",
    "        continue\n",
    "        \n",
    "    for dip in diplotypes:\n",
    "        dip_name = \"{} {}\".format(row[\"gene\"], dip)\n",
    "        recommendation = row[\"dosing\"] if row[\"dosing\"] != \"\" else row[\"recommendation\"]\n",
    "        phenotype_category_list.append(row[\"phenotype_category\"])\n",
    "        phenotype_list.append(row[\"phenotype\"])\n",
    "        diplotype_list.append(dip_name)\n",
    "        genotype_list.append(row[\"genotype\"])\n",
    "        implication_list.append(row[\"implication\"])\n",
    "        description_list.append(row[\"description\"])\n",
    "        recommendation_list.append(row[\"recommendation\"])\n",
    "        gene_list.append(row[\"gene\"])\n",
    "        drug_list.append(row[\"drug\"])\n",
    "        organization_list.append(row[\"organization\"])\n",
    "        title_list.append(row[\"title\"])\n",
    "        link_list.append(row[\"link\"])\n",
    "        \n",
    "pd.DataFrame({\n",
    "    \"phenotype_category\": phenotype_category_list,\n",
    "    \"phenotype\": phenotype_list,\n",
    "    \"diplotype\": diplotype_list,\n",
    "    \"genotype\": genotype_list,\n",
    "    \"implication\": implication_list,\n",
    "    \"description\": description_list,\n",
    "    \"recommendation\": recommendation_list,\n",
    "    \"gene\": gene_list,\n",
    "    \"drug\": drug_list,\n",
    "    \"organization\": organization_list,\n",
    "    \"title\": title_list,\n",
    "    \"link\": link_list,\n",
    "    \"data_source\": [\"guildeline\"] * len(link_list)\n",
    "}).to_csv(\"processed/diplotype_drug_relation.csv\", index=False)\n",
    "\n",
    "df_pheno_drug.to_csv(\"processed/phenotype_drug_relation.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "d532677b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "473"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_pheno_drug)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2386caa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "a8349fd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1892"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(link_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "860082f8",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>diplotype</td>\n",
       "      <td>16574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chemical</td>\n",
       "      <td>932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene</td>\n",
       "      <td>2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>haplotype</td>\n",
       "      <td>1526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rsID</td>\n",
       "      <td>6412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nmpa_drug</td>\n",
       "      <td>33608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>61065</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        node  count\n",
       "0  diplotype  16574\n",
       "1   chemical    932\n",
       "2       gene   2013\n",
       "3  haplotype   1526\n",
       "4       rsID   6412\n",
       "5  nmpa_drug  33608\n",
       "6      total  61065"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame({\n",
    "    \"node\": [\"diplotype\", \"chemical\", \"gene\", \"haplotype\", \"rsID\", \"nmpa_drug\", \"total\"],\n",
    "    \"count\": [16574, 932, 2013, 1526, 6412, 33608, 61065]\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b7119569",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edge</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>clinical_annotation</td>\n",
       "      <td>8318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>diplotype_consist_of</td>\n",
       "      <td>32420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>diplotype_metabolizer</td>\n",
       "      <td>10291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>drug_label</td>\n",
       "      <td>29628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>guideline_annotation</td>\n",
       "      <td>5797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>haplotype_rsID_related</td>\n",
       "      <td>785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>mutation_at</td>\n",
       "      <td>8183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>research_annotation</td>\n",
       "      <td>4291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>drug_consist_of</td>\n",
       "      <td>33608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>total</td>\n",
       "      <td>133321</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     edge   count\n",
       "0     clinical_annotation    8318\n",
       "1    diplotype_consist_of   32420\n",
       "2   diplotype_metabolizer   10291\n",
       "3              drug_label   29628\n",
       "4    guideline_annotation    5797\n",
       "5  haplotype_rsID_related     785\n",
       "6             mutation_at    8183\n",
       "7     research_annotation    4291\n",
       "8         drug_consist_of   33608\n",
       "9                   total  133321"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame({\n",
    "    \"edge\": [\"clinical_annotation\", \"diplotype_consist_of\", \"diplotype_metabolizer\", \"drug_label\",\n",
    "             \"guideline_annotation\", \"haplotype_rsID_related\", \"mutation_at\", \"research_annotation\",\n",
    "             \"drug_consist_of\", \"total\"],\n",
    "    \"count\": [8318, 32420, 10291, 29628, 5797, 785, 8183, 4291, 33608, 133321]\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "046c9bbc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
