{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "648ae453",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'D:\\\\pgkb_graph'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "from collections import defaultdict\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7e7da333",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CACNA1S functionality\\CACNA1S_allele_functionality_reference.xlsx\n",
      "CYP2B6 functionality\\CYP2B6_allele_functionality_reference.xlsx\n",
      "CYP2C19 functionality\\CYP2C19_allele_functionality_reference.xlsx\n",
      "CYP2C9 functionality\\CYP2C9_allele_functionality_reference.xlsx\n",
      "CYP2D6 functionality\\CYP2D6_allele_functionality_reference.xlsx\n",
      "CYP3A5 functionality\\CYP3A5_allele_functionality_reference.xlsx\n",
      "DPYD functionality\\DPYD_allele_functionality_reference.xlsx\n",
      "MT-RNR1 functionality\\MT-RNR1_allele_functionality_reference.xlsx\n",
      "NUDT15 functionality\\NUDT15_allele_functionality_reference.xlsx\n",
      "RYR1 functionality\\RYR1_allele_functionality_reference.xlsx\n",
      "SLCO1B1 functionality\\SLCO1B1_allele_functionality_reference.xlsx\n",
      "TPMT functionality\\TPMT_allele_functionality_reference.xlsx\n",
      "UGT1A1 functionality\\UGT1A1_allele_functionality_reference.xlsx\n"
     ]
    }
   ],
   "source": [
    "functionality_folder = \"functionality\"\n",
    "\n",
    "mentioned_genes = [\"CACNA1S\", \"CYP2B6\", \"CYP2C9\", \"CYP2C19\", \"CYP2D6\", \"CYP3A5\",\n",
    "                   \"DPYD\", \"MT-RNR1\", \"NUDT15\", \"RYR1\", \"SLCO1B1\", \"TPMT\",\n",
    "                   \"UGT1A1\"]\n",
    "\n",
    "functionality_path_dict = {g: \"\" for g in mentioned_genes}\n",
    "for path in os.listdir(functionality_folder):\n",
    "    for gene in mentioned_genes:\n",
    "        if \"{}_allele_functionality\".format(gene) in path:\n",
    "            functionality_path_dict[gene] = os.path.join(functionality_folder, path)\n",
    "for key,value in sorted(functionality_path_dict.items(), key=lambda x: x[0]):\n",
    "    print(key, value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ad7031af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# xl = pd.ExcelFile('functionality\\CYP2B6_allele_functionality_reference.xlsx')\n",
    "# print(xl.sheet_names)  # see all sheet names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d00397ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CACNA1S\n",
      "{'rs772226819 ': {'function_status': 'Uncertain functione', 'activity_score': '', 'pmid': '19825159, 28259615', 'evidence': '', 'finding': 'positive IVCT ', 'nucleotide': 'c.520C>T', 'protein': 'p.R174W'}, '': {'function_status': '', 'activity_score': '', 'pmid': 15201141, 'evidence': '', 'finding': 'Similar to that observed for MH mutations in RYR1, the R1086H MH mutation in the 1S-subunit of the skeletal muscle DHPR increases the sensitivity of the calcium release mechanism to activation by both pharmacological (caffeine) and endogenous (voltage sensor) activators. \\nThreshold for caffeine-induced calcium release is approximately fivefold lower in R1086H- expressing myotubes. Increased “sensitivity” of RYR1 to activation by caffeine may be a consequence of an elevated resting calcium level in R1086H- expressing myotubes. The data suggest that the intracellular III-IV linker of the DHPR acts as a negative regulatory module for release channel activation and that the R1086H MH mutation in 1S disrupts this critical negative allosteric regulatory mechanism. \\n', 'nucleotide': '', 'protein': ''}, 'rs1800559 ': {'function_status': 'Uncertain functione', 'activity_score': '', 'pmid': '9199552, 11260227, 12411788, 16163667', 'evidence': '', 'finding': 'positive IVCT ', 'nucleotide': 'c.3257G>A', 'protein': 'p.R1086H'}}\n",
      "CYP2B6\n",
      "{'*1': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '21659470; 30201214', 'evidence': '', 'finding': '30201214: efavirenz (in vitro)', 'nucleotide': '', 'protein': ''}, '*2': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '12642465;23399569;23254426;22471906;21896912;26655325', 'evidence': '', 'finding': '12642465: 7-EFC (in vitro, no effect); 23399569: Efavirenz (in vivo, marginal increase for *2/*2, only one individual) ; 23254426: Efavirenz (in vivo, no effect on EFV concentration); 22471906: Efavirenz (in vivo, no effect on EFV concentration) ; 21896912: Efavirenz (in vivo, no effect on EFV concentration); 26655325: Efavirenz (in vivo, no effect on EFV concentration) ', 'nucleotide': '', 'protein': ''}, '*3': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '12642465; 26655325', 'evidence': '', 'finding': '12642465: 7-EFC (in vitro, no effect); 26655325: Efavirenz (in vivo, no effect on EFV concentration) ', 'nucleotide': '', 'protein': ''}, '*4': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '12642465; 21746968; 14515060; 17015050; 17235330; 26107645; 23399569; 23254426; 26389554; 21659470; 21821736; 21659470', 'evidence': '', 'finding': '12642465: 7-EFC (in vitro, increase); 21746968: artemether (in vitro, increase); 14515060: bupropion (in vivo, increase); 17015050: nicotine (in vivo, increase); 17235330: efavirenz (in vivo, increase) ; 26107645: efavirenz (in vivo, decrease - see comments); 23399569: efavirenz (in vivo, decrease - see comments); 23254426: efavirenz (in vivo, decrease - see comments); 26389554: methadone (in vivo, increase); 21659470: efavirenz (in vitro, no change), bupropion (in vitro, decrease); 21821736: efavirenz (in vitro, increase), cyclophosphamide (in vitro, decrease); 21659470: 7-EFC (in vitro, decrease)', 'nucleotide': '', 'protein': ''}, '*5': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '12642465; 21659470; 17559344; 21659470; 26655325; 23990572; 23399569; 23254426; 22354160; 20841522; 20662624', 'evidence': '', 'finding': '12642465: 7-EFC (in vitro, higher activity); 21659470: efavirenz (in vitro, increase), bupropion (in vitro, no increase); 17559344: efavirenz (in vitro, no effect) ; 21659470: 7-EFC (in vitro, decrease); 26655325: efavirenz (in vivo, no effect on EFV concentration); 23990572: efavirenz (in vivo, no effect on EFV concentration); 23399569: efavirenz (in vivo, no effect on EFV concentration); 23254426: efavirenz (in vivo, no effect on EFV concentration); 22354160: efavirenz (in vivo, no effect on EFV concentration); 20841522: efavirenz (in vivo, no effect on EFV concentration); 20662624: efavirenz (in vivo, no effect on EFV concentration)', 'nucleotide': '', 'protein': ''}, '*6': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '17559344; 22232427; 21659470; 15622315; 26655325; 12642465; 30201214', 'evidence': '', 'finding': '17559344: efavirenz (in vitro, decrease) ; 22232427: efavirenz (in vitro, decrease); 21659470: efavirenz, bupropion (in vitro, decrease); 15622315:  efavirenz (in vivo, decrease); 26655325: efavirenz (in vivo, decrease); 12642465: 7-EFC (in vitro, higher activity); 30201214: efavirenz (in vitro, decrease) ', 'nucleotide': '', 'protein': ''}, '*7': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '12642465; 21659470; 30201214; 31324697', 'evidence': '', 'finding': '12642465: 7-EFC (in vitro, higher activity); 21659470: (in vitro, efavirenz (increase); 7-EFC (decrease), buproprion (slight decrease)); 30201214 (in vitro, efavirenz (increase)); 31324697: efavirenz (in vitro, decrease) ', 'nucleotide': '', 'protein': ''}, '*8': {'function_status': 'No function', 'activity_score': '', 'pmid': '21659470; 20517174; 21746968; 15190123; 30201214', 'evidence': '', 'finding': '21659470: efavirenz, bupropion, 7-EFC (in vitro, no activity); 20517174: 7-EFC (in vitro, no activity); 21746968: artemether (in vitro, no activity); 15190123: no protein expression (in vitro); 30201214: efavirenz (in vitro, no expression)', 'nucleotide': '', 'protein': ''}, '*9': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 21659470, 'evidence': '', 'finding': '21659470: efavirenz, bupropion (in vitro)', 'nucleotide': '', 'protein': ''}, '*10': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '30201214; 21746968', 'evidence': '', 'finding': '30201214; efavirenz (in vitro, decrease); 21746968: artemether (in vitro)', 'nucleotide': '', 'protein': ''}, '*11': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '20517174; 21746968; 15190123; 27045425; 26779253; 25303294; 23734829; 19916993; 23778320; 30201214', 'evidence': '', 'finding': '20517174: 7-EFC, selegiline (in vitro, inactive); 21746968: artemether (in vitro, significant decrease); 15190123: bupropion (in vitro, almost complete deficient activity); 27045425: efavirenz (in vivo, decrease); 26779253: efavirenz (in vivo, no change); 25303294: efavirenz (in vivo, decrease, *6 and *11 combined); 23734829: efavirenz (in vivo, decrease); 19916993: efavirenz (in vivo, decrease); 23778320: efavirenz (in vivo, no change); 30201214: efavirenz (in vitro, no expression)', 'nucleotide': '', 'protein': ''}, '*12': {'function_status': 'No function', 'activity_score': '', 'pmid': '20517174; 21746968; 30201214', 'evidence': '', 'finding': '20517174: 7-EFC (in vitro, inactive); 21746968: artemether (in vitro, not detectable activity); 30201214: efavirenz (in vitro, no expression)', 'nucleotide': '', 'protein': ''}, '*13': {'function_status': 'No function', 'activity_score': '', 'pmid': '20517174; 17559344', 'evidence': '', 'finding': '20517174: 7-EFC (in vitro, inactive); 17559344: Efavirenz (in vitro, decrease)', 'nucleotide': '', 'protein': ''}, '*14': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '21746968; 17559344; 15190123; 20517174; 30201214', 'evidence': '', 'finding': '21746968: artemether ( in vitro); 17559344: Efavirenz (in vitro, decrease); 15190123: bupropion (in vitro, decrease); 20517174: selegiline  (in vitro, decrease); 30201214: efavirenz (in vitro, decrease)', 'nucleotide': '', 'protein': ''}, '*15': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '20517174; 21746968; 17559344; 15190123; 20662624; 30201214', 'evidence': '', 'finding': '20517174: 7-EFC, selegiline (in vitro, inactive); 21746968: artemether (in vitro, severely decreased); 17559344: Efavirenz (in vitro, decrease); 15190123: bupropion (iin vitro, nactive); 20662624: Efavirenz (in vivo, no effect on EFV concentration for 1172T>A) ); 30201214: efavirenz (in vitro, no expression)', 'nucleotide': '', 'protein': ''}, '*17': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '16272958; 26655325', 'evidence': '', 'finding': '16272958: bupropion (in vitro); 26655325: efavirenz (in vivo, no effect on EFV concentration)', 'nucleotide': '', 'protein': ''}, '*18': {'function_status': 'No function', 'activity_score': '', 'pmid': '21746968; 20517174; 26655325; 16272958; 30201214; 16495778', 'evidence': '', 'finding': '21746968: artemether (in vitro); 20517174: 7-EFC (in vitro, inactive); 26655325: efavirenz (in vivo); 16272958: bupropion (in vitro, decrease 50%); 30201214: efavirenz (in vitro, no expression); from *16 - 16495778: bupropion (in vitro, no change), efavirenz (in vivo, decrease', 'nucleotide': '', 'protein': ''}, '*19': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '21746968; 16272958; 30201214', 'evidence': '', 'finding': '21746968: artemether (in vitro, increase); 16272958:in vitro, decreased expression in COS-1 cells ; 30201214: efavirenz (in vitro, decrease)', 'nucleotide': '', 'protein': ''}, '*20': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '21746968; 16272958', 'evidence': '', 'finding': '21746968: artemether (in vitro); 16272958: bupropion (in vitro)', 'nucleotide': '', 'protein': ''}, '*21': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '20517174; 21746968; 16272958; 30201214', 'evidence': '', 'finding': '20517174: 7-EFC (in vitro, inactive); 21746968: artemether (in vitro, not detectable activity); 16272958: bupropion (in vitro, decrease); 30201214: efavirenz (in vitro, no expression)', 'nucleotide': '', 'protein': ''}, '*22': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '17559344; 15722458', 'evidence': '', 'finding': '17559344: Efavirenz, bupropion (in vitro, increase); 15722458: bupropion (in vitro, increase)', 'nucleotide': '', 'protein': ''}, '*23': {'function_status': 'Unknown function', 'activity_score': '', 'pmid': 15499182, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*24': {'function_status': 'No function', 'activity_score': '', 'pmid': '20517174; 21746968; 30201214', 'evidence': '', 'finding': '20517174: 7-EFC (in vitro, inactive); 21746968: artemether (in vitro, no activity); 30201214: efavirenz (in vitro, no expression)', 'nucleotide': '', 'protein': ''}, '*25': {'function_status': 'Unknown function', 'activity_score': '', 'pmid': 15499182, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*26': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '17918089; 30201214', 'evidence': '', 'finding': '17918089: efavirenz (in vitro, was found as *6/*26 with decreased efavirenz metabolism, no separate *26 analysis); 30201214: efavirenz (in vitro, increase)', 'nucleotide': '', 'protein': ''}, '*27': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '21746968; 17235330', 'evidence': '', 'finding': '21746968: artemether (in vitro, significantly decreased); 17235330: bupropion (in vitro, significant decrease, more than *6)', 'nucleotide': '', 'protein': ''}, '*28': {'function_status': 'No function', 'activity_score': '', 'pmid': '20517174; 17235330', 'evidence': '', 'finding': '20517174: 7-EFC (in vitro, inactive); 17235330', 'nucleotide': '', 'protein': ''}, '*29': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 17885627, 'evidence': '', 'finding': '17885627: Efavirenz (in vivo, decrease); partial deletion, contain 516G>T (Q172H) and M198T (decreased function alleles)', 'nucleotide': '', 'protein': ''}, '*30': {'function_status': 'No function', 'activity_score': '', 'pmid': '28181240; 23164804', 'evidence': '', 'finding': '28181240: bupropion (in vitro), cyclophosphamide (in vivo); 23164804: contain the exon 7 p.R378X nonsense alteration characteristic of the CYP2B6*28 allele', 'nucleotide': '', 'protein': ''}, '*31': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 26655325, 'evidence': '', 'finding': '26655325: Efavirenz (in vivo, no effect on EFV concentration)', 'nucleotide': '', 'protein': ''}, '*32': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '26655325; 30201214', 'evidence': '', 'finding': '26655325: Efavirenz (in vivo, no effect on EFV concentration); 30201214: efavirenz (in vitro, increase)', 'nucleotide': '', 'protein': ''}, '*33': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '23418033; 30201214', 'evidence': '', 'finding': '23418033: ; 30201214: efavirenz (in vitro, increase)', 'nucleotide': '', 'protein': ''}, '*34': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23418033; 30201214', 'evidence': '', 'finding': '23418033: contains 516G>T (decreased activity), 785A>G (increased),R487S (uncertain or increased activity) and *22 (increased function); 30201214: efavirenz (in vitro, increase)', 'nucleotide': '', 'protein': ''}, '*35': {'function_status': 'No function', 'activity_score': '', 'pmid': '23418033; 30201214', 'evidence': '', 'finding': '23418033: contains 4 LOF variants. Recombinant expression in COS-1 cells and functional characterization using the upropion and efavirenz (EFV) as substrates showed complete or almost complete loss-of-function for variants p.G110V, p.I114T, p.V183G, and p.F213L; 30201214: efavirenz (in vitro, no expression)', 'nucleotide': '', 'protein': ''}, '*36': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 23418033, 'evidence': '', 'finding': 'no direct functional data for the combined allele, but contain 516G>T (decreased activity), 785A>G (increased function), -82T>C; (*22)(increased expression)', 'nucleotide': '', 'protein': ''}, '*37': {'function_status': 'No function', 'activity_score': '', 'pmid': '23418033; 30201214', 'evidence': '', 'finding': '23418033:contain 516G>T (decreased activity), 785A>G (increased function) and 548T>G (no function); 30201214: efavirenz (in vitro, no expression)', 'nucleotide': '', 'protein': ''}, '*38': {'function_status': 'No function', 'activity_score': '', 'pmid': 23640958, 'evidence': '', 'finding': '23640958: efavirenz (in vivo, 1 case, contains *6 and a rare truncated protein)', 'nucleotide': '', 'protein': ''}}\n",
      "CYP2C19\n",
      "{'*1': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '7487078, 32602114, 22027650', 'evidence': 'Definitve', 'finding': 'CYP2C19*1 is assigned normal function based on definitive evidence. CYP2C19*1 has been studied for more than 20 years and is the most frequently observed allele (7487078, 32602114). CYP2C19*1 has repeatedly demonstrated normal function with numerous substrates and its association with normal function is well-established (22027650). ', 'nucleotide': '', 'protein': ''}, '*2': {'function_status': 'No function', 'activity_score': '', 'pmid': '8195181, 22027650', 'evidence': 'Definitve', 'finding': 'CYP2C19*2 is assigned no function based on definitive evidence. CYP2C19*2 has been studied for more than 20 years and its association with impaired function is well-established (8195181, 22027650). CYP2C19*2 is defined by splicing defects resulting in a truncated protein lacking a heme binding region, making it a nonfunctional protein. CYP2C19*2 was first identified in a homozygous subject demonstrating decreased mephenytoin metabolism and was subsequently identified in seventeen additional subjects demonstrating decreased mephenytoin metabolism (8195181). Therefore, consensus among experts was no function based on definitive evidence.', 'nucleotide': '', 'protein': ''}, '*3': {'function_status': 'No function', 'activity_score': '', 'pmid': '7969038, 9103550, 22027650', 'evidence': 'Definitve', 'finding': 'CYP2C19*3 is assigned no function based on definitive evidence. CYP2C19*3 has been studied for more than 20 years and its association with impaired function is well-established (7969038, 9103550, 22027650). CYP2C19*3 is defined by an early stop codon resulting in a truncated protein lacking a heme binding region and a majority of the substrate binding region, making it a nonfunctional protein. CYP2C19*3 was first identified in a homozygous subject demonstrating decreased mephenytoin metabolism and was subsequently identified in six additional subjects demonstrating decreased mephenytoin metabolism (7969038).', 'nucleotide': '', 'protein': ''}, '*4': {'function_status': 'No function', 'activity_score': '', 'pmid': '9435198, 21358751', 'evidence': 'Limited', 'finding': 'CYP2C19*4 is assigned no function based on limited evidence in heterozygous subjects and an in vitro study. Two subjects with CYP2C19 *2/*4 genotype demonstrated decreased mephenytoin metabolism compared to wildtype subjects based on urine metabolite concentrations. This correlated with the in vitro experiment which found CYP2C19*4 protein was not expressed but mRNA levels were similar to wildtype indicating CYP2C19*4, which is defined by a SNP in the initiation codon, is not properly translated (9435198). Thus, the CYP2C19*4 defining SNP would result in a nonfunctional protein, even in the presence of a SNP associated with increased enzyme function (21358751). Therefore, consensus among experts was no function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*5': {'function_status': 'No function', 'activity_score': '', 'pmid': '9103550, 10022751, 21325430, 25001882', 'evidence': 'Moderate', 'finding': 'CYP2C19*5 is assigned no function based on moderate evidence in heterozygous subjects and in vitro studies.  CYP2C19*5 was first described in a subject demonstrating decreased mephenytoin metabolism compared to wildtype subjects and similar mephenytoin metabolism compared to CYP2C19 *2/*2, CYP2C19 *2/*3, and CYP2C19 *3/*3 subjects based on S/R mephenytoin ratio in urine metabolites  (9103550) . One CYP2C19*2/*5 subject similarly demonstrated decreased phenytoin metabolism based on S/R mephenytoin ratio in urine metabolites. This result was correlated with an in vitro study which showed CYP2C19*5 had a 40% decrease in enzyme expression, 0.4% enzyme activity with a mephenytoin substrate, and nondetectable enzyme activity with a tolbutamide substrate compared to wildtype (10022751). Two additional in vitro studies using different cell systems found CYP2C19*5 had nondetectable enzyme activity with mephenytoin, omeprazole, and clopidogrel substrates (21325430, 25001882). Therefore, consensus among experts was no function based on moderate evidence.', 'nucleotide': '', 'protein': ''}, '*6': {'function_status': 'No function', 'activity_score': '', 'pmid': '9732415, 21325430, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*6 is assigned no function based on limited evidence in a heterozygous subject and in vitro studies. CYP2C19*6 was first described in a subject demonstrating decreased mephenytoin metabolism based on urine metabolite concentrations. This correlated with an in vitro study which found CYP2C19*6 had nondetectable enzyme activity with mephenytoin and tolbutamide substrates (9732415). Two additional in vitro studies using different cell systems found CYP2C19*6 had nondetectable enzyme activity with mephenytoin, omeprazole, and clopidogrel substrates (21325430, 25001882). Therefore, consensus among experts was no function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*7': {'function_status': 'No function', 'activity_score': '', 'pmid': 10411572, 'evidence': 'Limited', 'finding': 'CYP2C19*7 is assigned no function based on limited evidence in a heterozygous subject. CYP2C19*7 is defined by a splicing defect and was described in a subject demonstrating decreased metabolism of mephenytoin based on urine metabolite concentrations after three separate phenotyping studies (10411572). Therefore, consensus among experts was no function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*8': {'function_status': 'No function', 'activity_score': '', 'pmid': '10411572, 21325430, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*8 is assigned no function based on limited evidence in a heterozygous subject and in vitro studies. CYP2C19*8 was first described in a subject demonstrating decreased metabolism of mephenytoin based on urine metabolite concentrations. This correlated with an in vitro study which found CYP2C19*8 had decreased enzyme activity compared to wildtype with mephenytoin and tolbutamide substrates (10411572). Two additional in vitro studies using different cell systems found CYP2C19*8 had nondetectable enzyme activity with mephenytoin, omeprazole, and clopidogrel substrates (21325430, 25001882). Therefore, consensus among experts was no function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*9': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '12464799, 21325430, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*9 is assigned decreased function based on limited evidence. Three in vitro studies using different cell systems found CYP2C19*9 demonstrated decreased enzyme activity compared to wildtype with mephenytoin, omeprazole, and clopidogrel substrates (12464799, 21325430, 25001882). Therefore, consensus among experts was decreased function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*10': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '12464799, 19661214, 21325430, 25001882, 24945780', 'evidence': 'Limited', 'finding': 'CYP2C19*10 is assigned decreased function based on limited evidence. Five in vitro studies using different cell systems found CYP2C19*10 demonstrated decreased enzyme activity compared to wildtype with mephenytoin, omeprazole, and clopidogrel substrates (12464799, 19661214, 21325430, 25001882, 24945780). Therefore, consensus among experts was decreased function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*11': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '12464799, 21325430, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*11 is assigned normal function based on limited evidence. Three in vitro studies using different cell systems found CYP2C19*11 demonstrated similar enzyme activity compared to wildtype with mephenytoin, omeprazole, and clopidogrel substrates (12464799, 21325430, 25001882). Therefore, consensus among experts was normal function based on limited evidence as it was determined the phenotypes containing this allele would be clinically similar to phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*12': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': 12464799, 'evidence': 'Inadequate', 'finding': 'CYP2C19*12 is assigned uncertain function due to inadequate evidence. In one in vitro study CYP2C19*12 was poorly expressed and enzyme activity could not be assessed with a mephenytoin substrate (12464799). CYP2C19*12 is defined by the conversion of the stop codon to cysteine, and it is unclear if this protein is unstable in vivo. Therefore, consensus among experts was not achieved and uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*13': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '12464799, 21325430, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*13 is assigned normal function based on limited evidence. Three in vitro studies using different cell systems found CYP2C19*13 demonstrated similar enzyme activity compared to wildtype with mephenytoin, omeprazole, and clopidogrel substrates (12464799, 21325430, 25001882). Therefore, consensus among experts was normal function based on limited evidence as it was determined the phenotypes containing this allele would be clinically similar to phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*14': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '21325430, 25001882', 'evidence': 'Inadequate', 'finding': 'CYP2C19*14 is assigned uncertain function due to inadequate evidence. In one in vitro study CYP2C19*14 demonstrated similar enzyme activity compared to wildtype with mephenytoin and omeprazole substrates (21325430). In a separate in vitro study CYP2C19*14 demonstrated decreased activity compared to wildtype with mephenytoin and clopidogrel substrates (25001882). Therefore, consensus among experts was not achieved and uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*15': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '21325430, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*15 is assigned normal function based on limited evidence. Two in vitro studies using different cell systems found CYP2C19*15 demonstrated similar enzyme activity compared to wildtype with mephenytoin and omeprazole substrates (21325430, 25001882). Therefore, consensus among experts was normal function based on limited evidence as it was determined the phenotypes containing this allele would be clinically similar to phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*16': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '15499191, 21325430, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*16 is assigned normal function based on limited evidence. A subject with CYP2C19 *2/*16 genotype demonstrated decreased mephobarbital excretion compared to heterozygous CYP2C19 *2 or CYP2C19*3 subjects but greater mephobarbital excretion compared to subjects carrying two no function alleles based on urine metabolites after a 200mg dose (15499191). Two in vitro studies using different cell systems found CYP2C19*16 had nondetectable or decreased enzyme activity compared to wildtype with mephenytoin, omeprazole, and clopidogrel substrates (25001882, 21325430). Therefore, consensus among experts was decreased function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*17': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '16413245, 17625515, 20083681', 'evidence': 'Strong', 'finding': 'CYP2C19*17 is assigned increased function based on strong evidence in homozygous and heterozygous subjects. Subjects with CYP2C19 *1/*17 and CYP2C19 *17/*17 genotype demonstrated increased omeprazole metabolism compared to wildtype subjects based on plasma metabolite measurements. The predicted AUC of omeprazole in CYP2C19*17/*17 patients was 63% of the predicted AUC in wildtype subjects. Additionally, in a separate population, CYP2C19*17/*17 subjects demonstrated increased mephenytoin metabolism compared to wildtype and CYP2C19 *1/*17 subjects. A murine study found increased gene transcription of CYP2C19*17 likely results in the observed increased enzyme activity (16413245). CYP2C19 *17/*17 patients taking escitalopram had 42% serum escitalopram concentrations compared to wildtype (17625515). CYP2C19 *17/*17 patients taking clopidogrel had lower ADP induced platelet aggregation likely as a result of increased clopidogrel metabolism to its active metabolite compared to wildtype and CYP2C19 *1/*17 patients (20083681). Therefore, consensus among experts was increased function based on strong evidence.', 'nucleotide': '', 'protein': ''}, '*18': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '17455109, 18312490, 21325430, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*18 is assigned normal function based on limited evidence. Four in vitro studies found CYP2C19*18 demonstrated similar intrinsic clearance compared to wildtype with mephenytoin, omeprazole, and clopidogrel substrates (17455109, 18312490, 25001882, 21325430). Therefore, consensus among experts was normal function based on limited evidence as it was determined the phenotypes containing this allele would be clinically similar to phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*19': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '17455109, 18312490, 21325430, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*19 is assigned decreased function based on limited evidence. In vitro studies found CYP2C19*19 had decreased intrinsic clearance compared to wildtype with mephenytoin, omeprazole, and clopidogrel substrates (17455109, 21325430, 25001882). Although two in vitro studies found CYP2C19*19 demonstrated similar intrinsic clearance compared to wildtype with omeprazole and mephenytoin, the experts did not find these results convincing as CYP2C19*19  also demonstrated higher Km values compared to wildtype in these studies, suggesting CYP2C19*19 has decreased substrate affinity (18312490, 21325430). Additionally, decreased intrinsic clearance compared to wildtype with these two substrates was previously reported as mentioned above. Therefore, consensus among experts was decreased function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*22': {'function_status': 'No function', 'activity_score': '', 'pmid': '25001882, 19164093', 'evidence': 'Limited', 'finding': 'CYP2C19*22 is assigned no function based on limited in vitro and computational evidence. An in vitro study found CYP2C19*22 had 8% enzyme activity compared to wildtype with a clopidogrel substrate and nondetectable enzyme activity compared to wildtype with a mephenytoin substrate (25001882). Crystal structure analysis shows the Arg186Pro substitution in CYP2C19*22 disrupts hydrogen bonds and alters protein structure resulting in decreased heme affinity and ultimately enzyme activity (19164093). Therefore, consensus among experts was no function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*23': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '25001882, 27578295, 30464555, 31713493', 'evidence': 'Inadequate', 'finding': 'CYP2C19*23 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C19*23 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (25001882, 27578295, 30464555, 31713493). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*24': {'function_status': 'No function', 'activity_score': '', 'pmid': '25001882, 27578295, 31713493', 'evidence': 'Limited', 'finding': 'CYP2C19*24 is assigned no function based on limited evidence. In vitro studies found CYP2C19*24 demonstrated nondetectable enzyme activity compared to wildtype with mephenytoin, clopidogrel, and omeprazole substrates (25001882, 27578295, 31713493). Additionally, CYP2C19*24 had less holoprotein compared to wildtype suggesting the protein is less stable resulting in little to no enzyme activity (27578295). Therefore, consensus among experts was no function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*25': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '25001882, 19636337, 31713493', 'evidence': 'Limited', 'finding': 'CYP2C19*25 is assigned decreased function based on limited in vitro and computational evidence. An in vitro study found CYP2C19*25 had decreased intrinsic clearance compared to wildtype with clopidogrel and mephenytoin, and omeprazole substrates (25001882, 31713493). A computational analysis of the Phe448Leu amino acid substitution in CYP2C19*25 predicted the change as deleterious. An examination of the structure showed the amino acid substitution, which is important for the heme-binding region, could disrupt the hydrophobic pocket in the region (19636337).  Therefore, consensus among experts was decreased function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*26': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '19661214, 25001882', 'evidence': 'Limited', 'finding': 'CYP2C19*26 is assigned decreased function based on limited evidence in a heterozygous subject and in vitro studies. CYP2C19*26 was first described in a subject also carrying the CYP2C19*2 allele and demonstrated decreased omeprazole metabolism compared to wildtype and CYP2C19 *1/*2 or CYP2C19*1/*3 subjects based on the plasma metabolite ratio. This result was correlated with an in vitro study which showed CYP2C19*26 demonstrated decreased intrinsic clearance compared to wildtype with mephenytoin and omeprazole substrates (19661214). An additional in vitro study using a different cell system similarly found CYP2C19*26 demonstrated decreased intrinsic clearance compared to wildtype with mephenytoin and clopidogrel substrates (25001882). Therefore, consensus among experts was decreased function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*28': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '25001882, 20712527', 'evidence': 'Limited', 'finding': 'CYP2C19*28 is assigned normal function based on limited in vitro and computational evidence. An in vitro study found CYP2C19*28 had similar intrinsic clearance compared to wildtype with a mephenytoin substrate (25001882). A computational analysis including splice site, mFold and nonsynonymous mutation analysis, did not find a deleterious impact on enzyme function (20712527). Although CYP2C19*28 was found to have increased intrinsic clearance compared to wildtype with a clopidogrel substrate in vitro (25001882), the experts did not find this result convincing as computational analysis and in vitro data with mephenytoin did not support increased function. Therefore, consensus among experts was normal function based on limited evidence as it was determined the phenotypes containing this allele would be clinically similar to phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*29': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '26153442, 30464555, 29098786', 'evidence': 'Inadequate', 'finding': 'CYP2C19*29 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C19*29 demonstrated modest decreased enzyme activity compared to wildtype with mephenytoin, omeprazole, and voriconazole substrates (26153442, 30464555). One in vitro study found increased clearance relative to wild-type for nebivolol (29098786). Consensus was not achieved among experts as data were sparse and conflicting and it was not convincing the magnitude of decreased activity was clinically significant. Therefore, uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*30': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '26153442, 30464555, 29098786', 'evidence': 'Inadequate', 'finding': 'CYP2C19*30 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C19*30 demonstrated decreased enzyme activity compared to wildtype with mephenytoin, omeprazole, voriconazole, and nebivolol substrates (26153442, 30464555, 29098786). Consensus was not achieved among experts as data were sparse. Therefore, uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*31': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '26153442, 30464555, 29098786', 'evidence': 'Inadequate', 'finding': 'CYP2C19*31 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C19*31 demonstrated decreased enzyme activity compared to wildtype with mephenytoin, omeprazole, voriconazole, and nebivolol substrates (26153442, 30464555, 29098786). Consensus was not achieved among experts as data were sparse and not consistent among substrates. Therefore, uncertain function was assigned. ', 'nucleotide': '', 'protein': ''}, '*32': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '26153442, 30464555, 29098786', 'evidence': 'Inadequate', 'finding': 'CYP2C19*32 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C19*32 demonstrated enzyme activity ranging from decreased to similar activity compared to wildtype with various substrates (26153442, 30464555, 29098786). Consensus was not achieved among experts and therefore uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*33': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '26153442, 30464555, 29098786', 'evidence': 'Inadequate', 'finding': 'CYP2C19*33 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C19*33 demonstrated decreased enzyme activity compared to wildtype with mephenytoin, omeprazole, and voriconazole substrates (26153442, 30464555, 29098786). Consensus was not achieved among experts as data were sparse and not consistent among substrates. Therefore, uncertain function was assigned', 'nucleotide': '', 'protein': ''}, '*34': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': 23799451, 'evidence': 'Inadequate', 'finding': 'CYP2C19*34 is assigned uncertain function due to inadequate evidence. CYP2C19*34 was identified in a patient heterozygous for the CYP2C19*2 allele who experienced adverse effects with venlafaxine and a supratherapeutic nortriptyline level. CYP2C19 enzyme function was not adequately assessed as CYP2D6 is involved in metabolism of antidepressants such as nortriptyline and this patient was a CYP2D6 intermediate metabolizer (23799451). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*35': {'function_status': ' No function', 'activity_score': '', 'pmid': 26021325, 'evidence': 'Limited', 'finding': 'CYP2C19*35 is assigned no function based on limited evidence. CYP2C19*35 is defined by a splicing defect resulting in a truncated protein (26021325). This splicing defect is also found in CYP2C19*2 which definitively has no function. Although subjects with CYP2C19 *1/*35 did not demonstrate significantly decreased enzyme activity compared to wildtype based on urine metabolites after mephenytoin 100mg once (26021325), the experts did not find this result convincing due to the small sample size. Therefore, consensus among experts was no function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*36': {'function_status': ' No function', 'activity_score': '', 'pmid': 31260137, 'evidence': 'Limited', 'finding': 'CYP2C19*36 is assigned no function based on limited evidence. CYP2C19*36 is a complete deletion of the CYP2C19 gene and was identified by interrogation of multiple databases of clinical chromosomal microarray data (31260137). Therefore, because a gene deletion would not result in protein formation, consensus among experts was no function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*37': {'function_status': ' No function', 'activity_score': '', 'pmid': 31260137, 'evidence': 'Limited', 'finding': 'CYP2C19*37 is assigned no function based on limited evidence. CYP2C19*37 is a partial deletion of the CYP2C19 gene and was identified by interrogation of multiple databases of clinical chromosomal microarray data (31260137). Therefore, because a partial gene deletion is not expected to result in a functional protein, consensus among experts was no function based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*38': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '2009263, 32602114', 'evidence': 'Definitve', 'finding': 'CYP2C19*38 is assigned normal function based on definitive evidence. CYP2C19*38 was originally catalogued as CYP2C19*1.001 (i.e. CYP2C19*1A) (2009263). Due to the standardized nomenclature system this allele received its own star designation and is now catalogued as CYP2C19*38 (32602114). Therefore, because CYP2C19*38 does not contain any single nucleotide polymorphisms and the new star designation was a result of nomenclature standardization, consensus among experts was normal function.', 'nucleotide': '', 'protein': ''}}\n",
      "CYP2C9\n",
      "{'*1': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*2': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '29283396; 20150829; 1563752; 12742136; 10413320; 27179628; 25775139; 19082874; 30088221; 15637526', 'evidence': 'Definitive', 'finding': 'CYP2C9*2 is assigned decreased function based on definitive evidence in homozygous and heterozygous patients and in vitro experimental data. CYP2C9*2 has been studied for more than 20 years and its association with impaired function compared to wildtype is well-established (29283396, 20150829, 15637526). The phenotype for patients homozygous for the CYP2C9*2 variant was updated from poor metabolizer to intermediate metabolizer after reevaluation of prior and new evidence demonstrating CYP2C9*2 has more enzyme activity than CYP2C9*3, which results in in a similar phenotype for CYP2C9*1/*3 and CYP2C9 *2/*2 patients (12742136, 10413320, 27179628, 25775139, 19082874, 30088221), both of which can be distinguished from those homozygous for CYP2C9*3/*3. Therefore, consensus among experts was decreased function with an activity value of 0.5 based on definitive evidence.  ', 'nucleotide': '', 'protein': ''}, '*3': {'function_status': 'No function', 'activity_score': '', 'pmid': '8809086; 29283396; 20150829; 15637526; 12742136; 10413320; 27179628; 25775139; 19082874; 30088221', 'evidence': 'Definitive', 'finding': 'CYP2C9*3 is assigned no function based on definitive evidence in homozygous and heterozygous patients and in vitro experimental data. CYP2C9*3 has been studied for more than 20 years and its association with impaired function compared to wildtype is well-established (8809086, 29283396, 20150829, 15637526). This allele was previously assigned “decreased function” but after reevaluation of prior and new evidence, CYP2C9*3 is assigned “no function” because its activity is much more decreased compared to CYP2C9*2 for multiple substrates, resulting in the lowest possible designation for the phenotypes observed in homozygous patients (12742136, 10413320, 27179628, 25775139, 19082874, 30088221). Therefore, consensus among experts was no function with an activity value of 0 based on definitive evidence.  ', 'nucleotide': '', 'protein': ''}, '*4': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23752738; 21228733', 'evidence': 'Limited', 'finding': 'CYP2C9*4 is assigned decreased function based on limited evidence in heterozygous patients and in vitro studies. One patient on a stable dose of warfarin with CYP2C9*1/*4 genotype required a lower weekly warfarin dose compared to CYP2C9*1/*1 patients (21228733). In an in vitro study CYP2C9*4 demonstrated 16% enzyme activity with a warfarin substrate and 22% enzyme activity with a tolbutamide substrate compared to wildtype. The CYP2C9*4 defining SNP occurs in the substrate recognition site which likely contributes to the loss of enzyme activity (23752738).  Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*5': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23752738; 15289788; 11455026; 16220110; 17504998; 21811894; 11901091', 'evidence': 'Strong', 'finding': 'CYP2C9*5 is assigned decreased function based on strong evidence in heterozygous patients and in vitro studies. Three subjects carrying the CYP2C9*5 allele (CYP2C9*1/*5, CYP2C9*5/*6, CYP2C9*5/*8 genotype) had decreased losartan metabolism compared to wildtype based on urinary metabolites after a single 25mg dose (15289788). Two CYP2C9 *1/*5 subjects, one CYP2C9 *5/*6 subject, and one CYP2C9*5/*8 subject had lower phenytoin metabolism compared to wildtype based on urinary excretion of a phenytoin metabolite after a single 300mg dose. Additionally, the CYP2C9 *5/*6 subject demonstrated the lowest phenytoin metabolism as the only subject of the 109 subject cohort to have a urinary metabolite concentration below the limit of quantification (16220110). One CYP2C9 *5/*8 demonstrated the lowest acenocoumarol metabolism based on plasma concentrations in a 36 subject cohort (21811894). One CYP2C9*1/*5 subject demonstrated decreased indisulam metabolism with an AUC nearly two times greater than wildtype subjects (17504998). These results correlated with in vitro pharmacokinetic studies which demonstrated decreased clearance of warfarin, diclofenac, flurbiprofen, naproxen, and piroxicam compared to wildtype (11455026, 11901091). A separate in vitro study found CYP2C9*5 has decreased enzyme activity compared to wildtype with warfarin and tolbutamide substrates that (23752738). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on strong evidence.  ', 'nucleotide': '', 'protein': ''}, '*6': {'function_status': 'No function', 'activity_score': '', 'pmid': '23752738; 15289788; 11740344; 16220110; 17895500; 21811894; 15094935', 'evidence': 'Strong', 'finding': 'CYP2C9*6 is assigned no function based on strong supporting evidence in heterozygous and homozygous patients. This allele was first identified in a CYP2C9*6/*6 patient experiencing phenytoin toxicity after recently starting the medication. Based on serial phenytoin plasma concentrations after discontinuing phenytoin, the patient had an estimated 17% phenytoin clearance compared to wildtype (11740344). In a separate case report, a patient with CYP2C9*6/*6 genotype stable on a low weekly warfarin dose demonstrated impaired warfarin metabolism with an S:R warfarin ratio 6x greater than expected (15094935). The S:R warfarin ratio of this patient was the highest among 52 patients on a stable dose of warfarin with a CYP2C9 genotype of *1/*1, *1/*2, *1/*3, *2/*2, or *2/*3 (17895500).  Two subjects (CYP2C9*1/*6 and CYP2C9*5/*6 genotype, respectively) had decreased losartan metabolism compared to wildtype subjects based on urinary metabolites after a single 25mg dose (15289788). Three CYP2C9*1/*6 subjects and one CYP2C9*5/*6 subjects had lower phenytoin metabolism compared to wildtype based on urinary excretion of a phenytoin metabolite after a single 300mg dose. Additionally, the CYP2C9*5/*6 subject demonstrated the lowest phenytoin metabolism as the only subject of the 109 subject cohort to have a urinary metabolite concentration below the limit of quantification (16220110). CYP2C9*6 is defined by a deletion resulting in a frameshift and not included in in vitro studies as it is it is a null allele resulting in no protein formation (23752738). Although one CYP2C9*1/*6 subject had a similar acenocoumarol metabolite plasma concentration compared to wildtype subjects 8 and 24 hours after a single 8mg dose (21811894), the experts did not deem this result as convincing conflicting evidence based on the small sample size. Therefore, consensus among experts was no function with an activity value of 0 based on strong evidence.  ', 'nucleotide': '', 'protein': ''}, '*7': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '15284535; 23752738', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*7 is assigned uncertain function based on inadequate evidence due to conflicting in vitro data and no clinical data. In an in vitro study CYP2C9*7 demonstrated 47% warfarin clearance compared to wildtype, however, similar tolbutamide clearance compared to wildtype was demonstrated in the same study as well as a separate study (23752738, 15284535). Therefore, consensus among experts was not achieved and uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*8': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23752738; 22378156; 23959274; 16220110; 21811894; 23844998; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25924705; 27163851; 27199745; 25951663; 24077631; 15289788; 28758225; 23376925; 15284535', 'evidence': 'Definitive', 'finding': 'CYP2C9*8 is assigned decreased function based on definitive evidence in heterozygous patients and in vitro studies. Ten CYP2C9*1/*8 genotype patients on a stable warfarin dose had 28% lower S-warfarin clearance compared to wildtype patients which correlated with the in vitro kinetic study demonstrating the CYP2C9*8 protein had 30% reduced warfarin clearance compared to wildtype (22378156). The CYP2C9*8 allele was associated with decreased phenytoin metabolism in 19 subjects (CYP2C9*1/*8, CYP2C9*8/*9, CYP2C9*8/*11, or CYP2C9*5/*8) compared to wildtype subjects based on urinary excretion of a phenytoin metabolite after a single 300mg dose (16220110). Four CYP2C9*1/*8 subjects had decreased diclofenac metabolism compared to wildtype subjects based on the urinary metabolite ratio after a single 50mg dose (23959274). One in vitro study showed CYP2C9*8 had decreased mRNA expression compared to wildtype, resulting in decreased function and metabolism of substrates (23376925). Decreased metabolism of multiple substrates compared to wildtype was demonstrated in a series of in vitro studies (23752738, 27199745, 23844998, 24118918, 25142737, 25144335, 25476996, 25832633, 25884291, 25924705, 27163851, 25951663, 24077631). Although an in vitro study showed similar diclofenac metabolism compared to wildtype (22378156), the experts did not find this conflicting because four CYP2C9*1/*8 subjects exhibited decreased diclofenac metabolism as previously described (23959274). Additionally, the experts did not deem two in vitro studies where CYP2C9*8 demonstrated similar to increased enzyme activity compared to wildtype (28758225, 15284535) as conflicting evidence because both studies used the same system, and it could not be ruled out that these results which differed from the majority of in vitro studies were due to the specific system. Although six CYP2C9*1/*8 subjects had similar acenocoumarol metabolite plasma concentrations compared to wildtype subjects 8 and 24 hours after a single 8mg dose (21811894), the experts did not deem this result as conflicting evidence based on the small sample size. Similarly, two CYP2C9 *1/*8 subjects had a similar urinary losartan-metabolite ratio compared to wildtype after a single 25mg dose and the experts did not deem this result as convincing conflicting evidence based on the small sample size (15289788). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on definitive evidence.  ', 'nucleotide': '', 'protein': ''}, '*9': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '23752738; 15284535; 21811894; 28758225; 16220110; 19164093; 20716240', 'evidence': 'Strong', 'finding': 'CYP2C9*9 is assigned normal function based on strong evidence in heterozygous patients, homozygous patients, and in vitro studies. Twenty-one subjects with CYP2C9 *1/*9 genotype and three subjects with CYP2C9 *9/*9 genotype had similar urinary excretion of a phenytoin metabolite compared to CYP2C9 *1/*1 subjects following a phenytoin 300mg dose, demonstrating CYP2C9*9 is not associated with decreased phenytoin metabolism (16220110). Seven subjects with CYP2C9 *1/*9 genotype had similar acenocoumarol plasma concentrations 8 and 24 hours after an 8mg dose compared to CYP2C9 *1/*1 subjects (21811894). Seven patients on a stable dose of warfarin with CYP2C9 *1/*9 genotype had a similar maintenance warfarin dose compared to CYP2C9 *1/*1 patients (20716240). In vitro studies demonstrated CYP2C9*9 intrinsic clearance or enzyme activity was similar compared to wildtype using warfarin and tolbutamide substrates (23752738, 15284535). Although one in vitro study showed CYP2C9*9  demonstrated decreased enzyme activity with a 7-ethoxycoumarin substrate compared to wildtype (28758225), the experts did not find this result convincing as no clinical difference was seen in patients taking acenocoumarol or warfarin, or in another in vitro study using a warfarin substrate as previously described. Although one study using the polyphen computational program to predict the functional effect of the SNP defining CYP2C9*9 predicted the substitution as possibly damaging to enzyme function, the authors also noted that this effect has not been shown in clinical phenotypic studies (19164093). Therefore, consensus among experts was normal function with an activity value of 1 based on strong evidence.', 'nucleotide': '', 'protein': ''}, '*10': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '23752738; 28758225; 15284535 ', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*10 is assigned uncertain function based on conflicting in vitro data and no clinical data. Although CYP2C9*10 demonstrated decreased warfarin clearance (23752738) it also demonstrated similar activity with a tolbutamide substrate (15284535) and activity ranging from decreased to normal with 7-ethoxycoumarin, flavanone, and steroid substrates (28758225) compared to wildtype. Therefore, consensus among experts was not achieved and uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*11': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '15284535; 23752738; 15289788; 15970795; 25951663; 16220110; 21811894; 23844998; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25924705; 27163851; 27199745; 24077631', 'evidence': 'Definitive', 'finding': 'CYP2C9*11 is assigned decreased function based on definitive evidence in heterozygous patients and in vitro studies. Four subjects carrying the CYP2C9*11 allele (n=3 CYP2C9*1/*11, n=1 CYP2C9*8/*11) demonstrated decreased phenytoin metabolism compared to wildtype subjects based on urinary excretion of a phenytoin metabolite after a single 300mg dose (16220110). Three CYP2C9*1/*11 patients on a stable dose of warfarin required a 33% lower warfarin dose compared to CYP2C9*1/*1 patients which correlated with the in vitro investigation where CYP2C9*11 had low levels of enzyme expression, decreased half-life, and 70% catalytic efficiency compared to wildtype; these results are suggestive of improper protein folding leading to decreased stability and reduced enzyme activity (15970795). An examination of the crystal structure showed the substitution which defines the CYP2C9*11 allele disrupts the sole extrahelical hydrogen bonding interaction, destabilizing the structure (15284535). In vitro studies showed CYP2C9*11 demonstrated decreased metabolism of tolbutamide, warfarin, glimepiride, carvedilol, fluoxetine, flurbiprofen, mestranol, diclofenac, and losartan compared to wildtype (15284535, 23752738, 24118918, 25476996, 25884291, 27199745, 25144335, 25924705, 25951663, 23844998). Although other in vitro studies showed CYP2C9*11 had similar activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*2 and/or CYP2C9*3 which definitively have decreased function and no function, respectively, also demonstrated similar activity compared to wildtype in these studies (25142737, 25832633, 27163851). The experts did not find the result of an in vitro study which showed CYP2C9*11 had similar activity compared to wildtype with a tolbutamide substrate (24077631) convincing because this result was not replicated as two separate in vitro studies which used two different systems and showed CYP2C9*11 demonstrated decreased clearance compared to wildtype, as described previously (15284535, 23752738). The experts also did not deem the results of a study which found no difference in losartan metabolism among four subjects carrying the CYP2C9*11 allele nor a study which found no difference in acenocoumarol metabolism among one subject with CYP2C9*1/*11 genotype as convincing conflicting evidence due to the small sample size (15289788, 21811894). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on definitive evidence.', 'nucleotide': '', 'protein': ''}, '*12': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '15284535; 23752738; 28758225; 23688605', 'evidence': 'Limited', 'finding': 'CYP2C9*12 is assigned decreased function based on limited evidence in heterozygous patients and in vitro studies. Four patients on a stable dose of warfarin with CYP2C9*1/*12 genotype required a lower weekly warfarin dose compared to CYP2C9*1/*1 patients (23688605). In vitro studies similarly found CYP2C9*12 demonstrated decreased enzyme activity with warfarin and tolbutamide substrates compared to wildtype (23752738, 15284535, 28758225). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*13': {'function_status': 'No function', 'activity_score': '', 'pmid': '23752738; 16308280; 15764711; 25951663; 19541829; 16740353; 20814155; 19604036; 21395648; 21726410; 21842338; 22735459; 23844998; 24118918; 24464553; 25142737; 25476996; 25614704; 25832633; 25884291; 27163851; 25144335; 27377818; 27864660; 15606435; 24077631', 'evidence': 'Definitive', 'finding': 'CYP2C9*13 is assigned no function based on definitive evidence in heterozygous patients. Subjects with CYP2C9*1/*13 genotype repeatedly demonstrated decreased clearance and increased AUC compared to wildtype subjects and similar clearance and AUC compared to CYP2C9*1/*3 subjects in multiple studies with different substrates (19604036, 21395648, 21726410, 21842338, 22735459, 27377818, 27864660, 15606435). Two studies correlated in vitro data with clinical data and also found increased AUC and decreased clearance compared to wildtype with lornoxicam and tolbutamide (15764711, 25614704). Additionally, a CYP2C9*3/*13 subject had a greater decrease in tolbutamide clearance compared to the CYP2C9*1/*13 subject (25614704). This was also demonstrated in another CYP2C9*3/*13 subject with decreased lornoxicam clearance and increased AUC compared to wildtype, CYP2C9 *1/*3, and CYP2C9*1/*2 subjects (15606435). Multiple in vitro studies demonstrate CYP2C9*13 exhibits decreased metabolism comparable to that observed for CYP2C9*3 for numerous substrates (23844998, 24118918, 24464553, 25142737, 25476996, 25832633, 25884291, 27163851, 24077631, 25951663, 19541829, 20814155, 23752738, 16308280, 25144335) and one model suggests the decreased metabolism across substrates is due to a side chain rotation which blocks the catalytic site (16740353). Therefore, consensus among experts was no function with an activity value of 0 based on definitive evidence.  ', 'nucleotide': '', 'protein': ''}, '*14': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '16099926; 23844998; 24118918; 25142737; 25144335; 25476996; 25951663; 25832633; 25884291; 25924705; 27163851; 24077631; 24956244; 19077919; 15371982; 23752738', 'evidence': 'Moderate', 'finding': 'CYP2C9*14 is assigned decreased function based on moderate evidence in heterozygous patients and in vitro studies. Separate studies of patients on a stable dose of warfarin with CYP2C9 *1/*14 genotype required a lower warfarin dose compared to CYP2C9*1/*1 patients (24956244, 19077919, 15371982). The VKORC1 and CYP4F2 variants associated with increased warfarin dose requirements were also observed in one of these patients which further suggested decreased warfarin metabolism and warfarin dose was due to the CYP2C9 *1/*14 genotype (24956244). In vitro studies similarly found CYP2C9*14 demonstrated decreased clearance of warfarin, phenytoin, tolbutamide, glimepiride, carvedilol, propofol, flurbiprofen, mestranol, diclofenac, and losartan (23752738, 27163851, 16099926, 24077631, 24118918, 25476996, 25832633, 25144335, 25924705, 25951663, 23844998). Although two in vitro studies showed CYP2C9*14 had similar activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*2 and/or CYP2C9*3 which definitively have decreased function and no function, respectively, demonstrated increased activity compared to wildtype in these studies (25142737, 25884291). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on moderate evidence.', 'nucleotide': '', 'protein': ''}, '*15': {'function_status': 'No function', 'activity_score': '', 'pmid': '16099926; 23752738', 'evidence': 'Limited', 'finding': 'CYP2C9*15 is assigned no function based on limited evidence. CYP2C9*15 is defined by a nonsense mutation which results in a premature stop codon resulting in a significant truncation of the protein. In an in vitro study, CYP2C9*15 did not result in detectable protein expression (16099926). CYP2C9*15 was not included in subsequent in vitro studies because it is a null allele resulting in no protein formation (23752738). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal and decreased function alleles. ', 'nucleotide': '', 'protein': ''}, '*16': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '16099926; 24077631; 24118918; 24464553; 25142737; 25951663; 25144335; 25476996; 25614704; 25832633; 25884291; 25924705; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*16 is assigned decreased function based on limited evidence in a heterozygous patient and in vitro studies. One subject with CYP2C9*1/*16 genotype exhibited decreased clearance compared to wildtype subjects based on tolbutamide plasma concentrations after one 500mg dose. This result correlated with the in vitro study where CYP2C9*16 demonstrated similar enzyme expression but 3.4% enzyme activity with a tolbutamide substrate compared to wildtype (25614704). Multiple in vitro studies similarly found CYP2C9*16 demonstrated decreased clearance with various substrates compared to wildtype (16099926, 24077631, 24118918, 24464553, 25951663, 25144335, 25476996, 25832633, 25884291, 25924705, 27163851, 23844998). Although one in vitro study showed CYP2C9*16 had similar activity compared to wildtype, the experts did not find this result convincing as in this study CYP2C9*2 and CYP2C9*3 which definitively have decreased function and no function, respectively, demonstrated increased activity compared to wildtype (25142737).  Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*17': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '23752738; 16099926', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*17 is assigned uncertain function based on inadequate evidence due to conflicting in vitro data and no clinical data. In an in vitro study CYP2C9*17 demonstrated 32% warfarin clearance compared to wildtype, however, similar tolbutamide clearance compared to wildtype was demonstrated in the same study as well as a separate study (23752738, 16099926). Therefore, consensus among experts was not achieved and uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*18': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '23752738; 16099926', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*18 is assigned uncertain function due to inadequate evidence. In one in vitro study, CYP2C9*18 protein expression could not be detected in an E.coli system (16099926). CYP2C9*18 protein expression was detected in a separate study using COS-7 cells and no enzyme activity was detected with warfarin and tolbutamide substrates (23752738). Therefore, consensus among experts was uncertain function due to insufficient evidence. ', 'nucleotide': '', 'protein': ''}, '*19': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '23752738; 23844998; 24118918; 25142737; 25476996; 25144335; 25951663; 25614704; 25832633; 25884291; 27163851; 24077631; 16099926', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*19 is assigned uncertain function due to inadequate evidence. One subject with CYP2C9*1/*19 genotype exhibited decreased clearance compared to wildtype subjects based on tolbutamide plasma concentrations and urine metabolite concentrations after one 500mg dose. This result correlated with the in vitro study where CYP2C9*19 demonstrated 31% enzyme activity with a tolbutamide substrate compared to wildtype (25614704). Multiple in vitro studies similarly found CYP2C9*19 demonstrated decreased clearance with various substrates compared to wildtype (23752738, 23844998, 24118918, 25142737, 25476996, 25144335, 25951663, 25832633, 25884291, 27163851, 24077631), however, one in vitro study showed CYP2C9*19 had similar activity compared to wildtype with a tolbutamide substrate (16099926). Consensus was not achieved among experts and therefore uncertain function was assigned but experts agreed additional evidence may change the assignment to decreased function.', 'nucleotide': '', 'protein': ''}, '*20': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': 23752738, 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*20 is assigned uncertain function due to inadequate evidence. In one in vitro study, CYP2C9*20 demonstrated decreased enzyme activity compared to wildtype with a warfarin and tolbutamide substrate (23752738). Therefore, consensus among experts was uncertain function due to insufficient evidence. ', 'nucleotide': '', 'protein': ''}, '*21': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': 23752738, 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*21 is assigned uncertain function due to inadequate evidence. In one in vitro study, CYP2C9*21 demonstrated no enzyme activity compared to wildtype with a warfarin and tolbutamide substrate (23752738). Therefore, consensus among experts was uncertain function due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '*22': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': 23752738, 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*22 is assigned uncertain function due to inadequate evidence. In one in vitro study, CYP2C9*22 demonstrated decreased enzyme activity compared to wildtype with a warfarin and tolbutamide substrate (23752738). Therefore, consensus among experts was uncertain function due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '*23': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23752738; 23844998; 24118918; 25951663; 25142737; 25144335; 25476996; 25832633; 25884291; 25924705; 27163851; 24077631', 'evidence': 'Limited', 'finding': 'CYP2C9*23 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*23 demonstrated decreased clearance compared to wildtype (23844998, 24118918, 25144335, 25476996, 23752738, 25832633, 25884291, 25924705, 24077631). Although three in vitro studies showed CYP2C9*23 had similar activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*2 which definitively has decreased function, demonstrated increased or similar activity compared to wildtype in these studies (25142737, 25951663, 27163851). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*24': {'function_status': 'No function', 'activity_score': '', 'pmid': '23752738; 17332144; 16543980', 'evidence': 'Limited', 'finding': 'CYP2C9*24 is assigned decreased function based on limited evidence in a heterozygous patient and in vitro studies. A patient on a stable dose of warfarin with CYP2C9 *2/*24 genotype had the lowest daily warfarin dose requirement and warfarin clearance based on plasma concentrations compared to patients with CYP2C9 *1/*2 genotype and matched for age, warfarin indication and co-administrated medications (16543980). An in vitro study similarly found CYP2C9*24 exhibited no enzyme activity with warfarin or tolbutamide substrates (23752738). No CYP2C9*24 protein was detected in a separate in vitro study suggesting the enzyme may be improperly folded (17332144). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*25': {'function_status': 'No function', 'activity_score': '', 'pmid': '23752738; 16788382', 'evidence': 'Limited', 'finding': 'CYP2C9*25 is assigned no function based on limited evidence. CYP2C9*25 is defined by a deletion which leads to a frameshift resulting in an early stop codon and truncated protein. This was demonstrated in an in vitro study where CYP2C9*25 had similar mRNA levels compared to wildtype but no protein expression (16788382). CYP2C9*25 was not included in subsequent in vitro studies because it is it is a null allele resulting in no protein formation (23752738). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal and decreased function alleles. ', 'nucleotide': '', 'protein': ''}, '*26': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23752738; 19541829; 16788382', 'evidence': 'Limited', 'finding': 'CYP2C9*26 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*26 demonstrated decreased clearance compared to wildtype. These studies used different cell systems and substrates included warfarin, tolbutamide, diclofenac, losartan, and glimepiride (23752738, 19541829, 16788382). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*27': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '23752738; 16788382; 23844998; 24118918; 25142737; 25144335; 25951663; 25476996; 25614704; 25832633; 25884291; 25924705; 27163851; 24077631', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*27 is assigned uncertain function due to inadequate evidence. One subject with CYP2C9 *1/*27 genotype exhibited similar clearance compared to wildtype subjects based on tolbutamide plasma concentrations after one 500mg dose (25614704). In vitro studies showed conflicting results with CYP2C9*27 exhibiting enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24118918, 23844998, 25476996, 25832633, 24077631, 25144335, 25951663, 23752738, 16788382, 25884291, 25614704, 25924705, 27163851, 25142737). Consensus was not achieved among experts and therefore uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*28': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23752738;19541829; 16788382', 'evidence': 'Limited', 'finding': 'CYP2C9*28 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*28 demonstrated decreased clearance compared to wildtype. These studies used different cell systems and substrates included warfarin, tolbutamide, diclofenac, losartan, and glimepiride (23752738, 19541829, 16788382). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*29': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23752738; 16788382; 23844998; 24118918; 25142737; 25951663; 25144335; 25476996; 25614704; 25832633; 25884291; 25924705; 27163851; 24077631', 'evidence': 'Limited', 'finding': 'CYP2C9*29 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*29 demonstrated decreased clearance compared to wildtype (23752738, 23844998, 24118918, 25951663, 25144335, 25476996, 25832633, 25884291, 25924705, 24077631). Although two in vitro studies showed CYP2C9*29 had similar activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*2 which definitively has decreased function, demonstrated increased or similar activity compared to wildtype in these studies (25142737, 27163851). Other studies, however, also showed CYP2C9*29 has similar clearance compared to wildtype (16788382, 25614704). Although the experts recognized there is conflicting evidence, the experts did not find it convincing as it was determined the activity of CYP2C9*29 toward most substrates was similar to CYP2C9*2 which definitively has decreased function. Therefore, consensus among experts was decreased function with an activity value of 0.5.  ', 'nucleotide': '', 'protein': ''}, '*30': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23752738; 19541829; 16788382', 'evidence': 'Limited', 'finding': 'CYP2C9*30 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*30 demonstrated decreased clearance compared to wildtype. These studies used different cell systems and substrates included warfarin, tolbutamide, diclofenac, losartan, and glimepiride (23752738, 19541829, 16788382). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*31': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '23752738; 24118918; 23844998; 25142737; 25144335; 25951663; 25476996; 25832633; 25884291; 25924705; 27163851; 27199745; 24077631', 'evidence': 'Limited', 'finding': 'CYP2C9*31 is assigned decreased function based on limited evidence. Multiple in vitro studies using different cell systems found CYP2C9*31 demonstrated decreased clearance for numerous substrates compared to wildtype (23752738, 24118918, 23844998, 25144335, 25951663, 25476996, 25832633, 25884291, 27163851, 25924705, 27199745, 24077631). Although one in vitro study showed CYP2C9*31 had similar activity compared to wildtype, the experts did not find this result convincing as in this study CYP2C9*2 and CYP2C9*3 which definitively have decreased function and no function, respectively, demonstrated increased activity compared to wildtype (25142737). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*32': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': 23752738, 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*32 is assigned uncertain function due to inadequate evidence. In one in vitro study, CYP2C9*32 demonstrated similar enzyme activity compared to wildtype with a warfarin and tolbutamide substrate (23752738). Therefore, consensus among experts was uncertain function due to insufficient evidence. ', 'nucleotide': '', 'protein': ''}, '*33': {'function_status': 'No function', 'activity_score': '', 'pmid': '23752738; 18971529; 23844998; 24118918; 25142737; 25951663; 25144335; 25476996; 25832633; 25884291; 27163851; 19541829; 24077631', 'evidence': 'Limited', 'finding': 'CYP2C9*33 is assigned no function based on limited evidence. Multiple in vitro studies using different cell systems found CYP2C9*33 demonstrated decreased clearance for numerous substrates compared to wildtype and similar clearance compared to CYP2C9*3 which definitively has no function (23752738, 18971529, 23844998, 24118918, 25142737, 25951663, 25144335, 25476996, 25832633, 25884291, 27163851, 19541829, 24077631). Although one patient with CYP2C9 *1/*33 genotype had their blood pressure lowered with losartan, the experts did not find this result convincing due to the small sample size (18971529). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*34': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '23752738; 18971529; 19541829; 23844998; 24118918; 25142737; 25144335; 25951663; 25476996; 25614704; 25832633; 25884291; 25924705; 27163851; 24077631', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*34 is assigned uncertain function due to inadequate evidence. Although one subject with CYP2C9*1/*34 genotype exhibited decreased clearance compared to wildtype subjects based on tolbutamide plasma concentrations after one 500mg dose (25614704), in vitro studies found CYP2C9*34 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (23752738, 18971529, 23844998, 19541829, 24118918, 25142737, 25476996, 25614704, 25832633, 25884291, 25924705, 27163851, 24077631, 25144335, 25951663). Consensus was not achieved among experts and therefore uncertain function was assigned..', 'nucleotide': '', 'protein': ''}, '*35': {'function_status': 'No function', 'activity_score': '', 'pmid': '23752738; 24322786; 2145143', 'evidence': 'Limited', 'finding': 'CYP2C9*35 is assigned no function based on limited evidence in one heterozygous subject and in vitro studies. CYP2C9*35 was first described in a patient requiring a very low weekly warfarin dose of 2.25mg (2145143). No variant was observed in VKORC1 or CYP4F2 in this patient and a subsequent in silico study showed a decreased interaction with reductase resulting in impaired enzyme activity. In vitro studies similarly found CYP2C9*35 demonstrated no enzyme activity with warfarin, tolbutamide, and diclofenac substrates (23752738, 24322786). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal and decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*36': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25951663; 25924705; 27163851; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*36 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C9*36 demonstrated enzyme activity ranging from markedly decreased to markedly increased compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25884291, 25951663, 25924705, 27163851, 23844998). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*37': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25951663; 25476996; 25832633; 25884291; 25924705; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*37 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*37 demonstrated decreased clearance compared to wildtype with multiple substrates (24077631, 24118918, 25951663, 25476996, 25832633, 25924705, 27163851, 23844998). Although two in vitro studies showed CYP2C9*37 had similar activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*3 which definitively has no function, demonstrated increased activity compared to wildtype in these studies (25142737, 25884291).  An additional in vitro study showed CYP2C9*37 had similar activity compared to wildtype (25144335), however, the experts did not find this result convincing as the function of CYP2C9*37 demonstrated decreased function toward the majority of substrates. Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*38': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25951663; 25476996; 25832633; 25884291; 25924705; 27163851; 28758225; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*38 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*38 demonstrated decreased clearance compared to wildtype for multiple substrates (24077631, 24118918, 25144335, 25951663, 25476996, 25832633, 25884291, 25924705, 23844998). Although two in vitro studies showed CYP2C9*38 had similar activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*2 which definitively has decreased function, demonstrated increased activity and similar activity compared to wildtype in these studies (25142737, 27163851). Additionally, the experts did not deem an in vitro study where CYP2C9*38 demonstrated activity ranging from decreased to increased compared to wildtype with 7-ethoxycoumarin, flavanone, and steroid substrates (28758225) as convincing evidence because it could not be ruled out that these results which differed from the majority of in vitro studies were due to the specific system used in the study. Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*39': {'function_status': 'No function', 'activity_score': '', 'pmid': '24077631; 24118918 ;25142737; 25144335; 25476996; 25832633; 25951663; 25884291; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*39 is assigned no function based on limited evidence. Multiple in vitro studies found CYP2C9*39 demonstrated decreased clearance with numerous substrates compared to wildtype (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25951663, 25884291, 27163851, 23844998). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function or decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*40': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25951663; 25144335; 25476996; 25614704; 25832633; 25884291; 25924705; 27163851; 28758225; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*40 is assigned uncertain function due to inadequate evidence in one heterozygous subject and in vitro studies. One subject with CYP2C9 *1/*40 genotype exhibited decreased AUC and half-life but similar oral clearance compared to wildtype subjects based on tolbutamide plasma concentrations after one 500mg dose (25614704). In vitro studies found CYP2C9*40 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24077631, 24118918, 23844998, 25142737, 25144335, 25951663, 25476996, 25832633, 25884291, 25924705, 27163851, 28758225, 25614704). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*41': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25951663; 25144335; 25476996; 25614704; 25832633; 25884291; 25924705; 27163851; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*41 is assigned uncertain function due to inadequate evidence in one heterozygous subject and in vitro studies. One subject with CYP2C9 *1/*41 genotype exhibited increased AUC but similar oral clearance compared to wildtype subjects based on tolbutamide plasma concentrations after one 500mg dose (25614704). In vitro studies found CYP2C9*41 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25951663, 25476996, 25832633, 25884291, 25924705, 27163851, 23844998, 25614704). Therefore, consensus among experts was uncertain function due to inadequate evidence. ', 'nucleotide': '', 'protein': ''}, '*42': {'function_status': 'No function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25614704; 25832633; 25951663; 25884291; 25924705; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*42 is assigned no function based on limited evidence in one heterozygous patient and in vitro studies. One subject with CYP2C9*1/*42 genotype exhibited twice the AUC and less than half the oral clearance compared to wildtype subjects based on tolbutamide plasma concentrations after one 500mg dose (25614704). Multiple in vitro studies found CYP2C9*42 demonstrated markedly decreased clearance compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25951663, 25884291, 25924705, 27163851, 23844998, 25614704). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function or decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*43': {'function_status': 'No function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25951663; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*43 is assigned no function based on limited evidence in in vitro studies. Multiple in vitro studies found CYP2C9*43 demonstrated markedly decreased clearance compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25884291, 25951663, 27163851, 23844998). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function or decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*44': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25951663; 25884291; 25924705; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*44 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*44 demonstrated decreased clearance compared to wildtype (24118918, 25476996, 25144335, 25832633, 25951663, 25884291, 25924705, 23844998, 24077631). Although two in vitro studies showed CYP2C9*44 had similar activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*2 which definitively has decreased function, demonstrated increased or similar activity compared to wildtype in these studies (25142737, 27163851). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*45': {'function_status': 'No function', 'activity_score': '', 'pmid': '24077631; 24118918; 25144335; 25951663; 25142737; 25476996; 25614704; 25884291; 25832633; 25924705; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*45 is assigned no function based on limited evidence in one heterozygous patient and in vitro studies. One subject with CYP2C9*1/*45 genotype exhibited nearly twice the AUC and half the oral clearance compared to wildtype subjects based on tolbutamide plasma concentrations after one 500mg dose (25614704). Multiple in vitro studies found CYP2C9*45 demonstrated markedly decreased clearance compared to wildtype with various substrates (24077631, 24118918, 25144335, 25951663, 25476996, 25832633, 25924705, 27163851, 23844998, 25614704). Although two in vitro studies showed CYP2C9*45 had similar or increased activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*3 which definitively has no function, demonstrated increased or similar activity compared to wildtype in these studies (25142737, 25884291). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function or decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*46': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25614704; 25832633; 25884291; 25951663; 25924705; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*46 is assigned decreased function based on limited evidence in one heterozygous patient and in vitro studies. One subject with CYP2C9*1/*46 genotype exhibited increased AUC and decreased oral clearance compared to wildtype subjects based on tolbutamide plasma concentrations after one 500mg dose (25614704). Multiple in vitro studies found CYP2C9*46 demonstrated decreased clearance compared to wildtype with various substrates (24077631, 24118918, 25476996, 25832633, 25884291, 25951663, 25924705, 23844998, 25614704). Although two in vitro studies showed CYP2C9*46 had similar activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*2 which definitively has decreased function, demonstrated increased or similar activity compared to wildtype in these studies (25142737, 27163851). An additional in vitro study showed CYP2C9*46 had similar activity compared to wildtype (25144335), however, the experts did not find this result convincing as the function of CYP2C9*46 in vivo and in vitro demonstrated decreased function toward the majority of substrates. Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*47': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25924705; 25951663; 27163851; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*47 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C9*47 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25884291, 25924705, 25951663, 27163851, 23844998). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*48': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25614704; 25832633; 25884291; 25951663; 25924705; 27163851; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*48 is assigned uncertain function due to inadequate evidence in one heterozygous subject and in vitro studies. One subject with CYP2C9 *1/*48 genotype exhibited increased AUC and decreased oral clearance compared to wildtype subjects based on tolbutamide plasma concentrations after one 500mg dose (25614704). In vitro studies found CYP2C9*48 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25884291, 25951663, 25924705, 27163851, 23844998, 25614704). Consensus was not achieved among experts and therefore uncertain function was assigned.', 'nucleotide': '', 'protein': ''}, '*49': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25951663; 27163851; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*49 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C9*49 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25884291, 25951663, 27163851, 23844998). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*50': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25951663; 25476996; 25832633; 25884291; 25924705; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*50 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*50 demonstrated decreased clearance compared to wildtype for multiple substrates (24077631, 24118918, 25476996, 25144335, 25951663, 25832633, 25884291, 27163851, 25142737, 25924705, 23844998). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*51': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25951663; 25476996; 25832633; 25884291; 25924705; 27163851; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*51 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C9*51 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25951663, 25476996, 25832633, 25884291, 25924705, 27163851, 23844998). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*52': {'function_status': 'No function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25951663; 25884291; 25924705; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*52 is assigned no function based on limited evidence in in vitro studies. Multiple in vitro studies found CYP2C9*52 demonstrated markedly decreased clearance compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25951663, 25884291, 25924705, 27163851, 23844998). Therefore, consensus among experts was no function with an activity value of 0 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function or decreased function alleles.', 'nucleotide': '', 'protein': ''}, '*53': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25951663; 25924705; 27163851; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*53 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C9*53 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25951663, 25884291, 25924705, 27163851, 23844998). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*54': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25951663; 25924705; 27163851; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*54 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C9*54 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25951663, 25884291, 25924705, 27163851, 23844998). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*55': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25951663; 25924705; 27163851; 23844998', 'evidence': 'Limited', 'finding': 'CYP2C9*55 is assigned decreased function based on limited evidence. Multiple in vitro studies found CYP2C9*55 demonstrated decreased clearance compared to wildtype (24077631, 24118918, 25144335, 25476996, 25832633, 25951663, 25924705, 27163851, 23844998). Although two in vitro studies showed CYP2C9*55 had similar or increased activity compared to wildtype, the experts did not find these results convincing as in these studies CYP2C9*3 which definitively has no function, demonstrated increased or similar activity compared to wildtype in these studies (25142737, 25884291). Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}, '*56': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '24077631; 24118918; 25142737; 25144335; 25476996; 25832633; 25884291; 25951663; 25924705; 27163851; 23844998', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*56 is assigned uncertain function due to inadequate evidence. In vitro studies found CYP2C9*56 demonstrated enzyme activity ranging from decreased to increased compared to wildtype with various substrates (24077631, 24118918, 25142737, 25144335, 25476996, 25832633, 25951663, 25884291, 25924705, 27163851, 23844998). Therefore, consensus among experts was uncertain function due to inadequate evidence.', 'nucleotide': '', 'protein': ''}, '*57': {'function_status': 'Unknown function', 'activity_score': '', 'pmid': '', 'evidence': 'No evidence', 'finding': 'CYP2C9*57 is assigned unknown function due to no literature describing function of this allele. Therefore, consensus among experts was unknown function due to no available evidence.', 'nucleotide': '', 'protein': ''}, '*58': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '25142737; 25075423; 25884291', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*58 is assigned uncertain function due to inadequate evidence. CYP2C9*58 was first described in a patient requiring a low weekly warfarin dose of 9mg (25075423). No variant was observed in VKORC1 or CYP4F2 in this patient and a subsequent in silico study showed changes in hydrogen bonding which may affect the secondary structure of the protein. In vitro studies found CYP2C9*58 demonstrated decreased enzyme activity with multiple substrates (25142737, 25075423, 25884291). Consensus was not achieved among experts and therefore uncertain function was assigned. ', 'nucleotide': '', 'protein': ''}, '*59': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '25142737; 25994031; 25884291', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*59 is assigned uncertain function due to inadequate evidence. CYP2C9*59 was first described in a patient requiring a low daily warfarin dose of 1.5mg (25994031). No variant was observed in VKORC1 or CYP4F2 in this patient and a subsequent in silico study showed destabilization of the interaction with the heme group may cause reduction in catalytic activity in this allele. In vitro studies found CYP2C9*59 demonstrated decreased enzyme activity with multiple substrates (25142737, 25994031, 25884291). Consensus was not achieved among experts and therefore uncertain function was assigned. ', 'nucleotide': '', 'protein': ''}, '*60': {'function_status': 'Uncertain function', 'activity_score': '', 'pmid': '25142737; 26255664; 25884291', 'evidence': 'Inadequate evidence', 'finding': 'CYP2C9*60 is assigned uncertain function due to inadequate evidence. CYP2C9*60 was first described in a patient requiring a low daily warfarin dose of 1.5mg (26255664). In vitro studies found CYP2C9*60 demonstrated enzyme activity ranging from decreased to similar activity compared to wildtype with various substrates (25142737, 26255664, 25884291). Consensus was not achieved among experts and therefore uncertain function was assigned. ', 'nucleotide': '', 'protein': ''}, '*61': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 30518301, 'evidence': 'Limited', 'finding': 'CYP2C9*61 is assigned decreased function based on limited evidence in a heterozygous patient. CYP2C9*61 was first described in a patient requiring a low daily warfarin dose of 3mg (30518301). CYP2C9*61 is defined by two SNPs, one of which is the defining SNP for CYP2C9*2. Therefore, consensus among experts was decreased function with an activity value of 0.5 based on limited evidence as it was determined the function of this allele is similar to CYP2C9*2 and phenotypes containing this allele would be clinically distinct from phenotypes comprising normal function alleles.', 'nucleotide': '', 'protein': ''}}\n",
      "CYP2D6\n",
      "{}\n",
      "CYP3A5\n",
      "{'*1': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 12893984.0, 'evidence': '', 'finding': '12893984: testosterone, nifedipine (in vitro)', 'nucleotide': '', 'protein': ''}, '*2': {'function_status': 'Unknown function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*3': {'function_status': 'No function', 'activity_score': '', 'pmid': 15179407.0, 'evidence': '', 'finding': '15179407: midalozam (in vivo)', 'nucleotide': '', 'protein': ''}, '*4': {'function_status': 'Unknown function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*5': {'function_status': 'Unknown function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*6': {'function_status': 'No function', 'activity_score': '', 'pmid': 11279519.0, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*7': {'function_status': 'No function', 'activity_score': '', 'pmid': 14515058.0, 'evidence': '', 'finding': '14515058: midazolam (in vitro)', 'nucleotide': '', 'protein': ''}, '*8': {'function_status': 'Unknown function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*9': {'function_status': 'Unknown function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}}\n",
      "DPYD\n",
      "{'Reference': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, 'c.1905+1G>A (*2A)': {'function_status': 'No function', 'activity_score': '', 'pmid': 23328581.0, 'evidence': 'Strong', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1898delC (*3)': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1601G>A (*4)': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 23588312.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (ex vivo)', 'nucleotide': '', 'protein': ''}, 'c.1627A>G (*5)': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 23328581.0, 'evidence': 'Strong', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2194G>A (*6)': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 23328581.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.295_298delTCAT (*7)': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.703C>T (*8)': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.85T>C (*9A)': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 18452418.0, 'evidence': 'Strong', 'finding': 'Dihydrouracil (in vivo)', 'nucleotide': '', 'protein': ''}, 'c.2657G>A (*9B)': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2983G>T (*10)': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1003G>T (*11)': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1156G>T (*12)': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1679T>G (*13)': {'function_status': 'No function', 'activity_score': '', 'pmid': 23328581.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1129-5923C>G,  c.1236G>A (HapB3)': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 28295243.0, 'evidence': 'Strong', 'finding': '5-fluorouracil (ex vivo)', 'nucleotide': '', 'protein': ''}, 'c.2846A>T': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Strong', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.557A>G': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.62G>A': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.496A>G': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1218G>A': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1896T>C': {'function_status': 'Normal function ', 'activity_score': '', 'pmid': 23588312.0, 'evidence': 'Moderate', 'finding': '5-fluorouracil (ex vivo)', 'nucleotide': '', 'protein': ''}, 'c.46C>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.61C>T': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.313G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.343A>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.451A>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.498G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.601A>C': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.632A>G': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.775A>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 25410891.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.868A>G': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.929T>C': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.934C>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.967G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1024G>A': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1057C>T': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1108A>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1181G>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1180C>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1260T>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1278G>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1294G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1314T>G': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1349C>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1358C>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 23588312.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1403C>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1475C>T': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1484A>G': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1519G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1543G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1577C>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1615G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1682G>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1775G>A': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1774C>T': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1777G>A': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1796T>C': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1905C>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1906A>C': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.1990G>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2021G>A': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2161G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2186C>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2195T>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2279C>T': {'function_status': 'Decreased function ', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2303C>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2336C>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2482G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2582A>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2623A>C': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2639G>T': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2656C>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2872A>G': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2915A>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2921A>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2933A>G': {'function_status': 'No function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2978T>G': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.2977C>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.3049G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.3061G>C': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 24648345.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (in vitro)', 'nucleotide': '', 'protein': ''}, 'c.3067C>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 23588312.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (ex vivo)', 'nucleotide': '', 'protein': ''}, 'c.525G>A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 23588312.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (ex vivo)', 'nucleotide': '', 'protein': ''}, 'c.1371C>T': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 23588312.0, 'evidence': 'Limited', 'finding': '5-fluorouracil (ex vivo)', 'nucleotide': '', 'protein': ''}}\n",
      "MT-RNR1\n",
      "{'Reference': {'function_status': 'Normal risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 27654872, 'evidence': '', 'finding': '27654872: N/A', 'nucleotide': '', 'protein': ''}, '1095T>C': {'function_status': 'increased risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': '11079536, 11313749, 15555598, 15841390, 16875663, 21205314', 'evidence': 'Moderate', 'finding': 'MT-RNR1 1095T>C is assigned an increased risk of aminoglycoside-induced hearing loss based on moderate evidence. Screening of individuals or probands of individuals with aminoglycoside-hearing looss identified 1095T>C in a number of cases (11079536, 11313749, 15555598, 15841390, 16875663, 21205314). Therefore, consensus among experts was this variant is associated with increased risk of aminoglycoside-induced hearing loss based on moderate evidence. ', 'nucleotide': '', 'protein': ''}, '1494C>T': {'function_status': 'increased risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': '14681830, 16380089, 17085680, 17434445, 17698030, 17698299, 18830133, 19682603, 20100600, 21205314, 27397648, 29876232, 15722487, 14681831, 19687236, 28103015, 25474306', 'evidence': 'Strong', 'finding': 'MT-RNR1 1494C>T is assigned an increased risk of aminoglycoside-induced hearing loss based on strong evidence. Screening of individuals or probands of individuals with aminoglycoside-hearing looss identified 1494C>T in a number of cases (14681830, 16380089, 17085680, 17434445,  17698030,  17698299, 18830133, 19682603,  20100600, 21205314, 27397648, 29876232). LCLs and/or cybrind cell lines with 1494C>T mutation have reduced growth in the presence of aminoglycosides (14681830, 15722487).  Paromomycin shows increased binding to rRNA constructs carrying the T allele (19687236).  Increased Kd for neomycin binding to A-site constructs carrying the T allele compared to those carrying the C allele (28103015). Cybrid lines with the 1494C>T mutation grown in gentamicin have reduced ATP production and altered mitochondria morphology (25474306). Therefore, consensus among experts was this variant is associated with increased risk of aminoglycoside-induced hearing loss based on strong evidence. ', 'nucleotide': '', 'protein': ''}, '1555A>G': {'function_status': 'increased risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': '7689389, 8285309, 8414970, 9039999, 9040738, 9111378, 9164619, 9391883, 9490575, 9779807, 9831149, 9950117, 10633132, 10661905, 11174959, 11857751, 11870684, 12031626, 12920080, 14699607, 14755216, 15708009, 15841390, 15917167, 16168391, 16375862, 16458854, 16935512, 16955413, 17341440, 17637808, 17723226, 17999439, 18386806, 18790089, 18820594, 18830133, 19376484, 19818876, 19835846, 20100600, 20111055, 20353758, 21162657, 21205314, 21504270, 22475488, 23256547, 24252789, 25744662, 26497601, 27397648, 27427311, 29805548, 24703164, 25155176, 8687424, 9315872, 12127547, 19687236, 28103015, 10915767, 12054632, 21828074', 'evidence': 'Definitive', 'finding': 'MT-RNR1 1555A>G is assigned an increased risk of aminoglycoside-induced hearing loss based on definitive evidence. Screening of individuals or probands of individuals with aminoglycoside-hearing loss identified 1494C>T in a number of cases (7689389, 8285309, 8414970, 9039999, 9040738, 9111378, 9164619, 9391883, 9490575, 9779807, 9831149, 9950117, 10633132, 10661905, 11174959, 11857751, 11870684, 12031626, 12920080, 14699607, 14755216, 15708009, 15841390, 15917167, 16168391, 16375862, 16458854, 16935512, 16955413, 17341440, 17637808, 17723226, 17999439, 18386806, 18790089, 18820594, 18830133, 19376484, 19818876, 19835846, 20100600, 20111055, 20353758, 21162657, 21205314, 21504270, 22475488, 23256547, 24252789, 25744662, 26497601, 27397648, 27427311, 29805548, 24703164, 25155176). There is inhibiton of translation in fibroblasts and cybrid cell lines with 1555A>G mutation (8687424). Aminoglycosides bind to 12S rRNA carrying the G allele, but not to rRNA with the A allele (9315872, 12127547, 19687236, 28103015). LCLs with the 1555A>G mutation have reduced growth and translation when grown in paromomycin (10915767). No effect of 1555A>G mutation in fibroblast lines. Cybrids with 1555A>G had slower growth in neomycin or paromomycin (12054632, 21828074). Therefore, consensus among experts was this variant is associated with increased risk of aminoglycoside-induced hearing loss based on definitive evidence. ', 'nucleotide': '', 'protein': ''}, '827A>G': {'function_status': 'normal risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': '16782057, 18261986, 18830133, 20353758', 'evidence': 'Limited', 'finding': 'MT-RNR1 827A>G is assigned a normal risk of aminoglycoside-induced hearing loss based on limited evidence. There are weak evidence proposing an association with aminoglycoside-induced hearing loss (16782057, 18261986, 20353758). However,  there is one study showing patients with 827A>G mutation with a history of exposure to aminoglycosides with normal hearing (18830133) and this variant has too high a population frequency in certain haplogroups to be classified as associated with aminoglycoside-induced hearing loss. Therefore, consensus among experts was this variant is associated with normal risk of aminoglycoside-induced hearing loss based on limited evidence. ', 'nucleotide': '', 'protein': ''}, '663A>G': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': '8414970, 7689389, 7550368', 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 663A>G is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. This variant has been identified in individuals with aminoglycoside-induced hearing loss but also in  controls after aminoglycoside exposure (8414970, 7689389, 7550368). Therefore, consensus among experts is that there is uncertain risk of aminoglycoside-induced hearing loss associated with this variant.', 'nucleotide': '', 'protein': ''}, '669T>C': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': '17637808, 20353758', 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 669T>C is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. There are weak evidence proposing an association with aminoglycoside-induced hearing loss and this variant (17637808, 20353758). Therefore, consensus among experts was uncertain function due to insufficient evidence. ', 'nucleotide': '', 'protein': ''}, '747A>G': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 21205314, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 747A>G is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals with aminoglycoside-hearing loss identified this variant in one study (21205314). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence. ', 'nucleotide': '', 'protein': ''}, '786G>A': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 23969527, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 786G>A is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals with aminoglycoside-hearing loss identified this variant in one study (23969527). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '807A>G': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 18830133, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 807A>G is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (18830133). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '807A>C': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 18830133, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 807A>C is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (18830133). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '839A>G': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 21205314, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 839A>G is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (21205314). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '896A>G': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 18830133, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 896A>G is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (18830133). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '930A>G': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 18830133, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 930A>G is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (18830133). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '951G>A': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': '18830133, 18851951', 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 951G>A is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in two studies (18830133, 18851951). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '960C>del': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 18851951, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 960C>del is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (18851951). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '961T>G': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 18830133, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 961T<G is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. One report of normal hearing after aminoglycoside hearing loss in an invidual carrying this variant (18830133).  Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '961T>del': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 15917167, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 961T>del is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. In one study, 961T>del was detected in subjects in all four cohorts (no statistically significant difference in allele frequency between the groups). Pedigree analysis showed that T961del and the hearing loss phenotype did not co-segregate (15917167).  Therefore, consensus among experts is that there is uncertain risk of aminoglycoside-induced hearing loss associated with this variant.', 'nucleotide': '', 'protein': ''}, '961T>del+Cn': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': '10326749, 12037390, 7550368, 20171168', 'evidence': '', 'finding': 'MT-RNR1 961>del+Cn is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence.  Screening of individuals  with aminoglycoside-hearing loss identified this variant in three studies and one case report (10326749, 12037390, 7550368); however, one study found this variant in patients without aminoglycoside-induced hearing loss (20171168). Therefore, consensus among experts is that there is uncertain risk of aminoglycoside-induced hearing loss associated with this variant.', 'nucleotide': '', 'protein': ''}, '988G>A': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 18830133, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR 988G>A is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (18830133). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '1189T>C': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 22030963, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 1189T>Cis assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (22030963). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '1243T>C': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 7550368, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 1243T>C is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (7550368). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '1520T>C': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 7550368, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 1520T>C is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (7550368). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '1537C>T': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 17637808, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 1537C>T is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Screening of individuals  with aminoglycoside-hearing loss identified this variant in one study (17637808). Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}, '1556C>T': {'function_status': 'uncertain risk of aminoglycoside-induced hearing loss', 'activity_score': '', 'pmid': 15141753, 'evidence': 'Inadequate evidence', 'finding': 'MT-RNR1 1556C>T is assigned an uncertain risk of aminoglycoside-induced hearing loss based on inadequate evidence. Case report of of a patient with streptomycin-induced tinnitus carrying this variant (15141753) . Therefore, consensus among experts was uncertain risk of aminoglycoside-induced hearing loss due to insufficient evidence.', 'nucleotide': '', 'protein': ''}}\n",
      "NUDT15\n",
      "{'*1': {'function_status': 'Normal Function', 'activity_score': '', 'pmid': 26878724, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*2': {'function_status': 'No Function', 'activity_score': '', 'pmid': 26878724, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*3': {'function_status': 'No Function', 'activity_score': '', 'pmid': 26878724, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*4': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 26878724, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*5': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 26878724, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*6': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 26878724, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*7': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 28659275, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*8': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 28659275, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*9': {'function_status': 'No Function', 'activity_score': '', 'pmid': '28659275; 30728528', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}}\n",
      "RYR1\n",
      "{'rs193922747 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '9066328, 16163667, 17710899, 20681998', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.103T>C', 'protein': 'p.C35R'}, '': {'function_status': '', 'activity_score': '', 'pmid': 28337975, 'evidence': '', 'finding': 'decreases voltage-gated calcium release and resting cytosolic calcium levels in mice; increases endoplasmic reticulum stress/unfolded protein response, enhances Ca2+ uptake/ROS production by interfibrillar mitochondria, activates proapoptotic pathways and decreases protein synthesis', 'nucleotide': '', 'protein': ''}, 'rs193922748': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 23459219, 'evidence': '', 'finding': 'increased sensitivity to 4-chloro-m-cresol when transiently expressed in HEK-293 cells', 'nucleotide': 'c.130C>T', 'protein': 'p.R44C          '}, 'rs118192161': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 7547049, 'evidence': '', 'finding': 'positive caffeine and halothane contracture test', 'nucleotide': 'c.487C>T', 'protein': 'p.R163C'}, 'rs193922753 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 16163667, 'evidence': '', 'finding': 'increased sensitivity to caffeine and halothane', 'nucleotide': 'c.488G>T', 'protein': 'p.R163L'}, 'rs1801086 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 1354642, 'evidence': '', 'finding': 'positive caffeine and halothane contracture test', 'nucleotide': 'c.742G>A/C', 'protein': 'p.G248R'}, 'rs193922762': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 12883402, 'evidence': '', 'finding': 'positive CHCT ', 'nucleotide': 'c.982C>T', 'protein': 'p.R328W'}, 'rs121918592': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 16163667, 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.1021G>C ', 'protein': 'p.G341R'}, 'rs121918592 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '8012359, 8825043, 8902717, 10051009, 12059893, 12124989, 12151923, 12411788, 16163667, 17710899, 18564801, 19648156, 20681998, 24433488, 25960145, 29101530', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.1021G>A', 'protein': 'p.G341R'}, 'rs193922764 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 23459219, 'evidence': '', 'finding': 'higher intensity of calcium release in response to 4-chloro-m-cresol compared with WT at 4-chloro-m-cresol concentration of 50–500 μM; lower EC50 (21.2–170.4 μM) for 4-CmC activation than WT (453.2 μM) showing higher sensitivity', 'nucleotide': 'c.1201C>T', 'protein': 'p.R401C'}, 'rs118192116 ': {'function_status': 'Increased function ', 'activity_score': '', 'pmid': 9334205, 'evidence': '', 'finding': 'increased sensitivity to caffeine and halothane compared to vector-transfected HEK-293 analyzed by cellular Ca2+ photometry  (for caffeine: less strong than R163C, G248R, G341R, Y522S, R552W, R614C/L, R2163C/H, G2434R, R2435H, R2458H but similar to R2458C)', 'nucleotide': 'c.1209C>G', 'protein': 'p.I403M'}, 'rs118192162 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 7829078, 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.1565A>C', 'protein': 'p.Y522S'}, 'rs111888148 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '19191329, 19191333', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.1589G>A', 'protein': 'p.R530H'}, 'rs193922768': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 23459219, 'evidence': '', 'finding': 'increased sensitivity to 4-chloro-m-cresol when transiently expressed in HEK-293 cells', 'nucleotide': 'c.1597C>T', 'protein': 'p.R533C'}, 'rs144336148': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 23459219, 'evidence': '', 'finding': 'increased sensitivity to 4-chloro-m-cresol when transiently expressed in HEK-293 cells', 'nucleotide': 'c.1598G>A ', 'protein': 'p.R533H '}, 'rs193922770 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '9138151, 25658027', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.1654C>T', 'protein': 'p.R552W'}, 'rs118192172 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '1774074, 7547049', 'evidence': '', 'finding': 'positive caffeine and halothane contracture test', 'nucleotide': 'c.1840C>T', 'protein': 'p.R614C'}, 'rs193922772 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 16917943, 'evidence': '', 'finding': 'associated with MH (supplemental material)', 'nucleotide': 'c.1841G>T', 'protein': 'p.R614L'}, 'rs118192175 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '9497245, 10484775, 12124989', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.6487C>T', 'protein': 'p.R2163C'}, 'rs118192163 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '9497245, 12124989, 12208234, 16835904, 19648156', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.6488G>A', 'protein': 'p.R2163H'}, 'rs118192176 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '9497245, 10484775, 11668625, 12059893, 12709367, 19648156, 20681998, 24433488, 25735680', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.6502G>A', 'protein': 'p.V2168M'}, 'rs118192177 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 10484775, 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.6617C>G', 'protein': 'p.T2206R'}, 'rs112563513 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '19191329, 19648156, 23736090, 24433488', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.7007G>A', 'protein': 'p.R2336H'}, 'rs121918596 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 27857962, 'evidence': '', 'finding': 'p.ΔGlu2348 RyR1 variant expressed in HEK 293 cells was hypersensitive compared to WT when stimulated with 4-chloro-m-cresol', 'nucleotide': 'c.7039_7041delGAG/c.7042_7044delGAG', 'protein': 'p.E2348del'}, 'rs193922802 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 11525881, 'evidence': '', 'finding': 'altered calcium dependence and increased caffeine sensitivity in HEK cells expressing the variant', 'nucleotide': 'c.7048G>A', 'protein': 'p.A2350T'}, 'rs193922803 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '15210166, 19648156, 24361844, 25256590', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.7063C>T', 'protein': 'p.R2355W'}, 'rs193922807': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '14641996, 15210166, 24433488', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.7124G>C', 'protein': 'p.G2375A'}, 'rs193922809 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 12059893, 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.7282G>A', 'protein': 'p.A2428T'}, 'rs121918593 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 9334205, 'evidence': '', 'finding': 'increased sensitivity to caffeine and halothane ', 'nucleotide': 'c.7300G>A', 'protein': 'p.G2434R'}, 'rs28933396 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 9334205, 'evidence': '', 'finding': 'increased sensitivity to caffeine and halothane ', 'nucleotide': 'c.7304G>A', 'protein': 'p.R2435H'}, 'rs118192124': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '10823104, 12059893, 24433488, 25086907', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.7354C>T', 'protein': 'p.R2452W'}, 'rs193922816 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '10484775, 10612851, 12411788, 16163667, 24433488', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.7360C>T', 'protein': 'p.R2454C'}, 'rs118192122 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '10051009, 10484775, 12059893, 12709367, 16163667, 17081152, 19648156, 20681998, 24433488, 25960145', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.7361G>A', 'protein': 'p.R2454H'}, 'rs28933397 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 9334205, 'evidence': '', 'finding': 'increased sensitivity to caffeine and halothane ', 'nucleotide': 'c.7372C>T', 'protein': 'p.R2458C'}, 'rs121918594 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '9450902, 18564801, 19648156, 29355282', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.7373G>A', 'protein': 'p.R2458H'}, 'rs118192178': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 16621918, 'evidence': '', 'finding': 'enhanced CICR', 'nucleotide': 'c.7522C>G         ', 'protein': 'p.R2508G'}, 'rs118192178 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 16621918, 'evidence': '', 'finding': 'enhanced CICR in CCD patient with MH episode history', 'nucleotide': 'c.7522C>T', 'protein': 'p.R2508C'}, 'rs193922818': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 16621918, 'evidence': '', 'finding': 'enhanced CICR in CCD patient with MH family history', 'nucleotide': 'c.7523G>A', 'protein': 'p.R2508H'}, 'rs193922832  ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 16917943, 'evidence': '', 'finding': 'associated with MH (supplemental material)', 'nucleotide': 'c.9310G>A', 'protein': 'p.E3104K'}, 'rs193922843 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 16917943, 'evidence': '', 'finding': 'associated with MH (supplemental material)', 'nucleotide': 'c.11969G>T', 'protein': 'p.G3990V'}, 'rs118192167': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 11063719, 'evidence': '', 'finding': 'positive IVCT in subjects with with congenital myopathy with rods and cores; mutant channel was significantly more sensitive to gating by caffeine in HEK cells; resting Ca2+ concentrations were significantly elevated in the HEK-293 cells expressing the mutant channel; amplitude of Ca2+ release from the Ca2+ store is diminished significantly in HEK cells expressing the Y4796C mutant channel', 'nucleotide': 'c.14387A>G', 'protein': 'p.Y4796C'}, 'rs121918595': {'function_status': 'Increased function', 'activity_score': '', 'pmid': '10888602, 16163667, 18564801, 19648156, 25960145', 'evidence': '', 'finding': 'positive IVCT', 'nucleotide': 'c.14477C>T', 'protein': 'p.T4826I'}, 'rs193922876 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 18212565, 'evidence': '', 'finding': 'positive IVCT; more sensitive to activation with 4-chloro-m-cresol of B Lymphocytes', 'nucleotide': 'c.14497C>T', 'protein': 'p.H4833Y'}, 'rs193922878 ': {'function_status': 'Increased function', 'activity_score': '', 'pmid': 11928716, 'evidence': '', 'finding': 'increased caffeine-induced calcium release in CHO cells', 'nucleotide': 'c.14512C>G', 'protein': 'p.L4838V'}, 'rs118192168': {'function_status': 'Increased function ', 'activity_score': '', 'pmid': 16917943, 'evidence': '', 'finding': 'associated with MH (supplemental material)', 'nucleotide': 'c.14545G>A', 'protein': 'p.V4849I'}, 'rs63749869 ': {'function_status': 'Uncertain functione', 'activity_score': '', 'pmid': 11741831, 'evidence': '', 'finding': '4-chloro-m-cresol resulted in almost no increase in the [Ca2+]i was observed for 4861 while  in contrast, the addition of 300 μM 4-chloro-m-cresol to lymphoblastoid cells from control individuals or from an individual harbouring the MHS mutation Val2168Met caused a substantial increase in the [Ca2+]i. a release of calcium from intracellular stores in the absence of any pharmacological activators of RYR; significantly smaller thapsigargin-sensitive intracellular calcium stores compared to lymphoblasts from control individuals; normal sensitivity of the calcium release to the RYR inhibitor dantrolene', 'nucleotide': 'c.14582G>A', 'protein': 'p.R4861H'}, 'rs118192170 ': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': 10097181, 'evidence': '', 'finding': 'I4898T led to a simultaneous increase in intracellular calcium concentrations and decrease in sarcoplasmic reticulum luminal calcium concentrations suggesting a calcium “leak” from the SR lumen; I4898T led to increased sensitivity to calcium but decreased binding of ryanodine vs. normal RYR1 and decreased response to caffeine; positive IVCT', 'nucleotide': 'c.14693T>C', 'protein': 'p.I4898T'}}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SLCO1B1\n",
      "{'*1A': {'function_status': 'Normal function', 'activity_score': '', 'pmid': 16568260, 'evidence': '', 'finding': '16568260: pravastatin (in vivo)', 'nucleotide': '', 'protein': ''}, '*1B': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '16568260; 15970799; 16198652; 16513443; 11477075', 'evidence': '', 'finding': '16568260: pravastatin (in vivo); 15970799: pravastatin, atorvastatin, cerivastatin (in vitro); 16198652: rosuvastatin (in vivo); 16513443: atrasentan (in vivo, in vitro); 11477075: estrone sulfate, estradiol 17beta-d-glucuronide (in vitro)', 'nucleotide': '', 'protein': ''}, '*2': {'function_status': 'Possible decreased function', 'activity_score': '', 'pmid': '11477075; 12490595', 'evidence': '', 'finding': '11477075: estrone sulfate, estradiol 17beta-d-glucuronide (in vitro); 12490595: rifampin (in vitro)', 'nucleotide': '', 'protein': ''}, '*3': {'function_status': 'Possible decreased function', 'activity_score': '', 'pmid': '11477075; 12490595', 'evidence': '', 'finding': '11477075: estrone sulfate, estradiol 17beta-d-glucuronide (in vitro); 12490595: rifampin (in vitro)', 'nucleotide': '', 'protein': ''}, '*4': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*5': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '15970799; 11477075; 12490595; 16697742; 16513443; 22147369; 23469684; 17108811; 15116054; 16678544; 17473846; 21386754', 'evidence': '', 'finding': '15970799: pravastatin, atorvastatin, cerivastatin (in vitro); 11477075: estrone sulfate, estradiol 17beta-d-glucuronide (in vitro); 12490595: rifampin (in vitro); 16697742: rosuvastatin (in vitro); 16513443: atrasentan (in vivo, in vitro); 22147369: methotrexate (in vitro); 23469684: simvastatin (in vivo); 17108811: simvastatin (in vivo); 15116054: pravastatin ( in vivo); 16678544: pravastatin (in vivo); 17473846: atorvastatin, rosuvastatin (in vivo); 21386754: cerivastatin (in vivo)', 'nucleotide': '', 'protein': ''}, '*6': {'function_status': 'Possible decreased function', 'activity_score': '', 'pmid': '11477075; 12490595', 'evidence': '', 'finding': '11477075: estrone sulfate, estradiol 17beta-d-glucuronide (in vitro); 12490595: rifampin (in vitro)', 'nucleotide': '', 'protein': ''}, '*7': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*8': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*9': {'function_status': 'Possible decreased function', 'activity_score': '', 'pmid': '11477075; 12490595; 16697742; 16513443', 'evidence': '', 'finding': '11477075: estrone sulfate, estradiol 17beta-d-glucuronide (in vitro); 12490595: rifampin (in vitro); 16697742: rosuvastatin (in vitro); 16513443: atrasentan (in vivo, in vitro)', 'nucleotide': '', 'protein': ''}, '*10': {'function_status': 'Possible decreased function', 'activity_score': '', 'pmid': '11477075; 12490595', 'evidence': '', 'finding': '11477075: estrone sulfate, estradiol 17beta-d-glucuronide (in vitro); 12490595: rifampin (in vitro)', 'nucleotide': '', 'protein': ''}, '*11': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*12': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*13': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*14': {'function_status': 'Possible increased function', 'activity_score': '', 'pmid': '22147369; 23311897', 'evidence': '', 'finding': '22147369: methotrexate (in vivo); 23311897: atorvastatin (in vivo)', 'nucleotide': '', 'protein': ''}, '*15': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*16': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*17': {'function_status': 'Decreased function', 'activity_score': '', 'pmid': '15970799; 15226675', 'evidence': '', 'finding': '15970799: pravastatin, atorvastatin, cerivastatin (in vitro); 15226675: pravastatin (in vivo)', 'nucleotide': '', 'protein': ''}, '*18': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*19': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*20': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*21': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*22': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*23': {'function_status': 'Possible decreased function', 'activity_score': '', 'pmid': 22147369, 'evidence': '', 'finding': '22147369: methotrexate (in vitro)', 'nucleotide': '', 'protein': ''}, '*24': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*25': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*26': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*27': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*28': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*29': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*30': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*31': {'function_status': 'Possible decreased function', 'activity_score': '', 'pmid': 22147369, 'evidence': '', 'finding': '22147369: methotrexate (in vitro)', 'nucleotide': '', 'protein': ''}, '*32': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*33': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*34': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*35': {'function_status': 'Possible increased function', 'activity_score': '', 'pmid': '22147369; 23311897', 'evidence': '', 'finding': '22147369: methotrexate (in vitro); 23311897: atorvastatin (in vivo)', 'nucleotide': '', 'protein': ''}, '*36': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}}\n",
      "TPMT\n",
      "{'*1': {'function_status': 'Normal Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*2': {'function_status': 'No Function', 'activity_score': '', 'pmid': '786267;16220112;9177237;18708949;8644731   ', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*3A': {'function_status': 'No Function', 'activity_score': '', 'pmid': '16220112;9177237;18708949;8644731;8561894', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*3B': {'function_status': 'No Function', 'activity_score': '', 'pmid': '16220112;18708949;8644731;8561894', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*3C': {'function_status': 'No Function', 'activity_score': '', 'pmid': '16220112;18708949;8644731;8561894', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*4': {'function_status': 'No Function', 'activity_score': '', 'pmid': 9246020, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*5': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;16220112;9246020', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*6': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;16220112;9246020', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*7': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;16220112;9246020;13679074', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*8': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;16220112;27564568;9931346', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*9': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;18602085;15226673;16220112', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*10': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;13679074;10833476;16220112', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*11': {'function_status': 'No Function', 'activity_score': '', 'pmid': '18708949;12835738;16220112', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*12': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;13679074;27564568;16220112', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*13': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;13679074;16220112', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*14': {'function_status': 'No Function', 'activity_score': '', 'pmid': '18708949;15083071', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*15': {'function_status': 'No Function', 'activity_score': '', 'pmid': 15083071, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*16': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;15652243;25564374;15226673', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*17': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;15226673', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*18': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;15226673', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*19': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;15652243', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*20': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;16917910', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*21': {'function_status': 'Uncertain Function ', 'activity_score': '', 'pmid': '18708949;18602085;16917910', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*22': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18708949;16917910', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*23': {'function_status': 'No Function', 'activity_score': '', 'pmid': '17885628;18708949', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*24': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '18602085;18708949', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*25': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 18602085, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*26': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 19660010, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*27': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 19945438, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*28': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 20945351, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*29': {'function_status': 'No Function', 'activity_score': '', 'pmid': 22447118, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*30': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '23407052;16946561', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*31': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '23407052;20881512', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*32': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': '25564374;23252716', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*33': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 23252716, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*34': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 23252716, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*35': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': 23252704, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*36': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': 23252704, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*37': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 24710034, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*38': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': 25564374, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*39': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 27703193, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*40': {'function_status': 'Uncertain Function', 'activity_score': '', 'pmid': 27564568, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*41': {'function_status': 'No Function', 'activity_score': '', 'pmid': 28462921, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}}\n",
      "UGT1A1\n",
      "{'*1': {'function_status': 'Normal function', 'activity_score': '', 'pmid': '7565971; 9653159; 15179405', 'evidence': '', 'finding': '15179405: SNS-38 (in vivo) - based on protein quantification', 'nucleotide': '', 'protein': ''}, '*6': {'function_status': 'Decreased Function', 'activity_score': '', 'pmid': '7715297; 8280139; 15179405; 11906189', 'evidence': '', 'finding': '15179405: SNS-38 (in vivo) - based on protein quantification; 11906189: bilirubin (in vivo) - based on protein quantification', 'nucleotide': '', 'protein': ''}, '*27': {'function_status': 'Decreased Function', 'activity_score': '', 'pmid': '7565971; 7715297; 15179405', 'evidence': '', 'finding': '7715297: bilirubin (in vivo) - based on protein quantification; 15179405: SNS-38 (in vivo) - based on protein quantification', 'nucleotide': '', 'protein': ''}, '*28': {'function_status': 'Decreased Function', 'activity_score': '', 'pmid': '9653159; 15179405', 'evidence': '', 'finding': '15179405: SNS-38 (in vivo) - based on protein quantification', 'nucleotide': '', 'protein': ''}, '*36': {'function_status': 'Increased Function', 'activity_score': '', 'pmid': 9653159, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*37': {'function_status': 'Decreased Function', 'activity_score': '', 'pmid': 9653159, 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*80': {'function_status': 'Unknown Function', 'activity_score': '', 'pmid': '', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*80+*28': {'function_status': 'Decreased Function', 'activity_score': '', 'pmid': '21309756; 19414484; 22085899; 19419973', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}, '*80+*37': {'function_status': 'Decreased Function', 'activity_score': '', 'pmid': '21309756; 19414484; 22085899; 19419973', 'evidence': '', 'finding': '', 'nucleotide': '', 'protein': ''}}\n"
     ]
    }
   ],
   "source": [
    "def parse_functionality(path):\n",
    "    xl = pd.ExcelFile(path)\n",
    "    allele_function_dict = {}\n",
    "    \n",
    "    for sheet in xl.sheet_names:\n",
    "        if \"function\" in sheet.lower():\n",
    "            ### read and handle functionalities\n",
    "            df_functionality = pd.read_excel(xl, sheet_name=sheet, skiprows=1).fillna(\"\")\n",
    "            allele_column = \"\"\n",
    "            function_status_column = \"\"\n",
    "            activity_score_column = \"\"\n",
    "            pmid_column = \"\"\n",
    "            evidence_column = \"\"\n",
    "            finding_column = \"\"\n",
    "            nucleotide_column = \"\"\n",
    "            protein_column = \"\"\n",
    "\n",
    "            for column in df_functionality.columns:\n",
    "                if \"rsid\" in column.lower() or \"allele/\" in column.lower():\n",
    "                    allele_column = column\n",
    "                elif \"functional status\" in column.lower() or \"clinical function status\" in column.lower():\n",
    "                    function_status_column = column\n",
    "                elif \"activity score\" in column.lower():\n",
    "                    activity_score_column = column\n",
    "                elif \"pmid\" in column.lower() or \"reference\" in column.lower():\n",
    "                    pmid_column = column\n",
    "                elif \"evidence\" in column.lower():\n",
    "                    evidence_column = column\n",
    "                elif \"finding\" in column.lower():\n",
    "                    finding_column = column\n",
    "                elif \"nucleotide\" in column.lower():\n",
    "                    nucleotide_column = column\n",
    "                elif \"protein\" in column.lower():\n",
    "                    protein_column = column\n",
    "\n",
    "            for index, row in df_functionality.iterrows():\n",
    "                if allele_column != \"\":\n",
    "                    allele = row[allele_column]\n",
    "                else:\n",
    "                    continue\n",
    "                \n",
    "                function_status = row[function_status_column] if function_status_column != \"\" else \"\"\n",
    "                activity_score = row[activity_score_column] if activity_score_column != \"\" else \"\"\n",
    "                pmid = row[pmid_column] if pmid_column != \"\" else \"\"\n",
    "                evidence = row[evidence_column] if evidence_column != \"\" else \"\"\n",
    "                finding = row[finding_column] if finding_column != \"\" else \"\"\n",
    "                nucleotide = row[nucleotide_column] if nucleotide_column != \"\" else \"\"\n",
    "                protein = row[protein_column] if protein_column != \"\" else \"\"\n",
    "                 \n",
    "                if \"{}{}{}{}{}{}{}\".format(function_status, activity_score, pmid, evidence, finding, nucleotide, protein) != \"\":\n",
    "                    allele_function_dict[allele] = {\n",
    "                        \"function_status\": function_status,\n",
    "                        \"activity_score\": activity_score,\n",
    "                        \"pmid\": pmid,\n",
    "                        \"evidence\": evidence,\n",
    "                        \"finding\": finding,\n",
    "                        \"nucleotide\": nucleotide,\n",
    "                        \"protein\": protein\n",
    "                    }\n",
    "    return allele_function_dict\n",
    "\n",
    "for key,value in sorted(functionality_path_dict.items(), key=lambda x: x[0]):\n",
    "    func_dict = parse_functionality(value)\n",
    "    print(key)\n",
    "    print(func_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1abdc58",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa9bcd69",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38bf0cea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
