{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2ce42d76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'D:\\\\drug KG\\\\pgkb'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0347328b",
   "metadata": {},
   "source": [
    "## 药物与基因变异知识图谱搭建\n",
    "\n",
    "### 4种数据源的药物与基因变异关系数据\n",
    "\n",
    "- #### 临床数据 clinical annotation (保留置信级别 1A，1B)\n",
    "- #### FDA drug label\n",
    "- #### 科研数据 variant annotation (保留 p_value < 0.01)\n",
    "- #### 协会指南数据 Guideline annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3cd22e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e3d6fc80",
   "metadata": {},
   "source": [
    "## 读取临床数据 Read clinical annotation\n",
    "### Only remain evidence level in (1A, 1B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b143e5a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(265, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Level of Evidence</th>\n",
       "      <th>Level Override</th>\n",
       "      <th>Level Modifiers</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>PMID Count</th>\n",
       "      <th>Evidence Count</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>Phenotype(s)</th>\n",
       "      <th>Latest History Date (YYYY-MM-DD)</th>\n",
       "      <th>URL</th>\n",
       "      <th>Specialty Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>981755803</td>\n",
       "      <td>rs75527207</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>1A</td>\n",
       "      <td></td>\n",
       "      <td>Rare Variant; Tier 1 VIP</td>\n",
       "      <td>234.875</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>28</td>\n",
       "      <td>30</td>\n",
       "      <td>ivacaftor</td>\n",
       "      <td>Cystic Fibrosis</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>https://www.pharmgkb.org/clinicalAnnotation/98...</td>\n",
       "      <td>Pediatric</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1449191690</td>\n",
       "      <td>rs141033578</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>1A</td>\n",
       "      <td></td>\n",
       "      <td>Rare Variant; Tier 1 VIP</td>\n",
       "      <td>200</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>ivacaftor</td>\n",
       "      <td>Cystic Fibrosis</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>https://www.pharmgkb.org/clinicalAnnotation/14...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Clinical Annotation ID Variant/Haplotypes  Gene Level of Evidence  \\\n",
       "0               981755803         rs75527207  CFTR                1A   \n",
       "1              1449191690        rs141033578  CFTR                1A   \n",
       "\n",
       "  Level Override           Level Modifiers    Score Phenotype Category  \\\n",
       "0                 Rare Variant; Tier 1 VIP  234.875           Efficacy   \n",
       "1                 Rare Variant; Tier 1 VIP      200           Efficacy   \n",
       "\n",
       "   PMID Count  Evidence Count    Drug(s)     Phenotype(s)  \\\n",
       "0          28              30  ivacaftor  Cystic Fibrosis   \n",
       "1           1               3  ivacaftor  Cystic Fibrosis   \n",
       "\n",
       "  Latest History Date (YYYY-MM-DD)  \\\n",
       "0                       2021-03-24   \n",
       "1                       2021-03-24   \n",
       "\n",
       "                                                 URL Specialty Population  \n",
       "0  https://www.pharmgkb.org/clinicalAnnotation/98...            Pediatric  \n",
       "1  https://www.pharmgkb.org/clinicalAnnotation/14...                       "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df_clinical_annotation = pd.read_csv('clinical_annotation/clinical_annotations.tsv', sep='\\t').fillna(\"\")\n",
    "df_clinical_annotation = df_clinical_annotation[(df_clinical_annotation[\"Level of Evidence\"] == \"1A\") | \n",
    "                                              (df_clinical_annotation[\"Level of Evidence\"] == \"1B\")]\n",
    "df_clinical_annotation.index = range(len(df_clinical_annotation))\n",
    "print(df_clinical_annotation.shape)\n",
    "df_clinical_annotation.to_csv(\"clinical_annotation_filter.tsv\", sep=\"\\t\", index=False)\n",
    "df_clinical_annotation[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "a2e0bdd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Dosage',\n",
       " 'Efficacy',\n",
       " 'Efficacy;Toxicity',\n",
       " 'Metabolism/PK',\n",
       " 'Other',\n",
       " 'Toxicity'}"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df_clinical_annotation[\"Phenotype Category\"].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b3e0d94",
   "metadata": {},
   "source": [
    "## 读取FDA药物标签 Read drug labels\n",
    "### Only remain drug source = \"FDA\"\n",
    "### ~~and testing level in (\"Testing required\", \"Testing recommend\")~~"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b3d0f7fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(373, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PharmGKB ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Source</th>\n",
       "      <th>Biomarker Flag</th>\n",
       "      <th>Testing Level</th>\n",
       "      <th>Has Prescribing Info</th>\n",
       "      <th>Has Dosing Info</th>\n",
       "      <th>Has Alternate Drug</th>\n",
       "      <th>Cancer Genome</th>\n",
       "      <th>Prescribing</th>\n",
       "      <th>Chemicals</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Variants/Haplotypes</th>\n",
       "      <th>Latest History Date (YYYY-MM-DD)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA166184637</td>\n",
       "      <td>Annotation of FDA Label for oxymetazoline and ...</td>\n",
       "      <td>FDA</td>\n",
       "      <td>On</td>\n",
       "      <td>Actionable PGx</td>\n",
       "      <td>Prescribing Info</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Prescribing</td>\n",
       "      <td>oxymetazoline and tetracaine</td>\n",
       "      <td>BCHE; CYB5R3; G6PD</td>\n",
       "      <td></td>\n",
       "      <td>2019-10-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA166153492</td>\n",
       "      <td>Annotation of FDA Label for brivaracetam and C...</td>\n",
       "      <td>FDA</td>\n",
       "      <td>On</td>\n",
       "      <td>Actionable PGx</td>\n",
       "      <td>Prescribing Info</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Prescribing</td>\n",
       "      <td>brivaracetam</td>\n",
       "      <td>CYP2C19</td>\n",
       "      <td></td>\n",
       "      <td>2017-11-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PharmGKB ID                                               Name Source  \\\n",
       "0  PA166184637  Annotation of FDA Label for oxymetazoline and ...    FDA   \n",
       "1  PA166153492  Annotation of FDA Label for brivaracetam and C...    FDA   \n",
       "\n",
       "  Biomarker Flag   Testing Level Has Prescribing Info Has Dosing Info  \\\n",
       "0             On  Actionable PGx     Prescribing Info                   \n",
       "1             On  Actionable PGx     Prescribing Info                   \n",
       "\n",
       "  Has Alternate Drug Cancer Genome  Prescribing                     Chemicals  \\\n",
       "0                                   Prescribing  oxymetazoline and tetracaine   \n",
       "1                                   Prescribing                  brivaracetam   \n",
       "\n",
       "                Genes Variants/Haplotypes Latest History Date (YYYY-MM-DD)  \n",
       "0  BCHE; CYB5R3; G6PD                                           2019-10-24  \n",
       "1             CYP2C19                                           2017-11-09  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df_drug_label = pd.read_csv('drug_label/drugLabels.tsv', sep='\\t').fillna(\"\")\n",
    "df_drug_label = df_drug_label[df_drug_label[\"Source\"] == \"FDA\"]\n",
    "# df_drug_label = df_drug_label[(df_drug_label[\"Testing Level\"] == \"Testing required\") |\n",
    "#                               (df_drug_label[\"Testing Level\"] == \"Testing recommended\")]\n",
    "df_drug_label.index = range(len(df_drug_label))\n",
    "print(df_drug_label.shape)\n",
    "df_drug_label.to_csv(\"drug_label_filter.tsv\", sep=\"\\t\", index=False)\n",
    "df_drug_label[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "101bf1d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Actionable PGx',\n",
       " 'Informative PGx',\n",
       " 'Testing recommended',\n",
       " 'Testing required'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df_drug_label[\"Testing Level\"].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d5f0469",
   "metadata": {},
   "source": [
    "## 读取科研中的变异实验数据 Read variant annotation\n",
    "### Only remain variant annotation significance = yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5fee5013",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5072, 11)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "b'Skipping line 4308: expected 11 fields, saw 13\\n'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>Specialty Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1183684657</td>\n",
       "      <td>CYP2D6 ultrarapid metabolizer genotype</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>tramadol</td>\n",
       "      <td>18204346</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Median (+)R,R-tramadol area under the curve wa...</td>\n",
       "      <td>CYP2D6 ultra-metabolizer genotype is associate...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1448997750</td>\n",
       "      <td>CYP2B6*1, CYP2B6*18</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>efavirenz</td>\n",
       "      <td>16495778</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note that in the paper the allele was r...</td>\n",
       "      <td>CYP2B6 *1/*18 is associated with increased con...</td>\n",
       "      <td>*1/*18</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Variant Annotation ID                      Variant/Haplotypes    Gene  \\\n",
       "0             1183684657  CYP2D6 ultrarapid metabolizer genotype  CYP2D6   \n",
       "1             1448997750                     CYP2B6*1, CYP2B6*18  CYP2B6   \n",
       "\n",
       "     Drug(s)      PMID Phenotype Category Significance  \\\n",
       "0   tramadol  18204346      Metabolism/PK          yes   \n",
       "1  efavirenz  16495778      Metabolism/PK          yes   \n",
       "\n",
       "                                               Notes  \\\n",
       "0  Median (+)R,R-tramadol area under the curve wa...   \n",
       "1  Please note that in the paper the allele was r...   \n",
       "\n",
       "                                            Sentence Alleles  \\\n",
       "0  CYP2D6 ultra-metabolizer genotype is associate...           \n",
       "1  CYP2B6 *1/*18 is associated with increased con...  *1/*18   \n",
       "\n",
       "  Specialty Population  \n",
       "0                       \n",
       "1                       "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_annotation = pd.read_csv('variant_annotation/var_drug_ann.tsv', sep='\\t', error_bad_lines=False).fillna(\"\")\n",
    "df_variant_annotation = df_variant_annotation[df_variant_annotation[\"Significance\"] == \"yes\"]\n",
    "df_variant_annotation.index = range(len(df_variant_annotation))\n",
    "print(df_variant_annotation.shape)\n",
    "df_variant_annotation[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cdf164a",
   "metadata": {},
   "source": [
    "## Merge variant annotation with study params\n",
    "### 用variant annotation表和variant实验数据表做join，设置p_value阈值为0.01，过滤掉不符合条件的数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0c7f8a27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3344, 21)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>...</th>\n",
       "      <th>Study Type</th>\n",
       "      <th>Study Cases</th>\n",
       "      <th>Study Controls</th>\n",
       "      <th>Characteristics</th>\n",
       "      <th>Characteristics Type</th>\n",
       "      <th>Frequency In Cases</th>\n",
       "      <th>Frequency In Controls</th>\n",
       "      <th>P Value</th>\n",
       "      <th>Biogeographical Groups</th>\n",
       "      <th>p_value_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1183684657</td>\n",
       "      <td>CYP2D6 ultrarapid metabolizer genotype</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>tramadol</td>\n",
       "      <td>18204346</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Median (+)R,R-tramadol area under the curve wa...</td>\n",
       "      <td>CYP2D6 ultra-metabolizer genotype is associate...</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>cohort</td>\n",
       "      <td>14</td>\n",
       "      <td></td>\n",
       "      <td>11 UM; 3 PM; healthy male non-smoking volunteers</td>\n",
       "      <td>Study Cohort</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt; 0.001</td>\n",
       "      <td>European</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1448997750</td>\n",
       "      <td>CYP2B6*1, CYP2B6*18</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>efavirenz</td>\n",
       "      <td>16495778</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note that in the paper the allele was r...</td>\n",
       "      <td>CYP2B6 *1/*18 is associated with increased con...</td>\n",
       "      <td>*1/*18</td>\n",
       "      <td>...</td>\n",
       "      <td>cohort</td>\n",
       "      <td>51</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt; 0.0001</td>\n",
       "      <td>Multiple groups</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Variant Annotation ID                      Variant/Haplotypes    Gene  \\\n",
       "0             1183684657  CYP2D6 ultrarapid metabolizer genotype  CYP2D6   \n",
       "1             1448997750                     CYP2B6*1, CYP2B6*18  CYP2B6   \n",
       "\n",
       "     Drug(s)      PMID Phenotype Category Significance  \\\n",
       "0   tramadol  18204346      Metabolism/PK          yes   \n",
       "1  efavirenz  16495778      Metabolism/PK          yes   \n",
       "\n",
       "                                               Notes  \\\n",
       "0  Median (+)R,R-tramadol area under the curve wa...   \n",
       "1  Please note that in the paper the allele was r...   \n",
       "\n",
       "                                            Sentence Alleles  ... Study Type  \\\n",
       "0  CYP2D6 ultra-metabolizer genotype is associate...          ...     cohort   \n",
       "1  CYP2B6 *1/*18 is associated with increased con...  *1/*18  ...     cohort   \n",
       "\n",
       "  Study Cases Study Controls  \\\n",
       "0          14                  \n",
       "1          51                  \n",
       "\n",
       "                                    Characteristics Characteristics Type  \\\n",
       "0  11 UM; 3 PM; healthy male non-smoking volunteers         Study Cohort   \n",
       "1                                                                Unknown   \n",
       "\n",
       "  Frequency In Cases Frequency In Controls   P Value Biogeographical Groups  \\\n",
       "0                                            < 0.001               European   \n",
       "1                                           < 0.0001        Multiple groups   \n",
       "\n",
       "  p_value_effect  \n",
       "0           True  \n",
       "1           True  \n",
       "\n",
       "[2 rows x 21 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_param = pd.read_csv('variant_annotation/study_parameters.tsv', sep='\\t', error_bad_lines=False).fillna(\"\")\n",
    "df_variant_param = df_variant_param[[\"Variant Annotation ID\", \"Study Type\", \"Study Cases\", \"Study Controls\", \"Characteristics\",\n",
    "                                     \"Characteristics Type\", \"Frequency In Cases\", \"Frequency In Controls\", \"P Value\", \"Biogeographical Groups\"]]\n",
    "df_variant_merge = pd.merge(df_variant_annotation, df_variant_param, how=\"left\", left_on=\"Variant Annotation ID\", right_on=\"Variant Annotation ID\")\n",
    "\n",
    "# set 0.01 as p_value threshold\n",
    "p_value_list = df_variant_merge[\"P Value\"].values\n",
    "p_value_list = list(map(lambda x: \"1\" if \">\" in str(x) else str(x).replace(\"<\", \"\").replace(\"=\", \"\").strip(), p_value_list))\n",
    "p_value_bool_list = []\n",
    "for x in p_value_list:\n",
    "    try:\n",
    "        if float(x) <= 0.01:\n",
    "            p_value_bool_list.append(True)\n",
    "        else:\n",
    "            p_value_bool_list.append(False)\n",
    "    except:\n",
    "        p_value_bool_list.append(False)\n",
    "df_variant_merge = df_variant_merge.assign(p_value_effect=p_value_bool_list)\n",
    "df_variant_merge = df_variant_merge[df_variant_merge[\"p_value_effect\"] == True]\n",
    "\n",
    "print(df_variant_merge.shape)\n",
    "df_variant_merge.to_csv(\"variant_annotation_filter.tsv\", sep=\"\\t\", index=False)\n",
    "df_variant_merge[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "531e16d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'',\n",
       " 'Dosage',\n",
       " 'Dosage,\"Efficacy\"',\n",
       " 'Dosage,\"Efficacy\",\"Toxicity\"',\n",
       " 'Dosage,\"Efficacy\",\"Toxicity\",\"Metabolism/PK\"',\n",
       " 'Dosage,\"Metabolism/PK\"',\n",
       " 'Dosage,\"Other\"',\n",
       " 'Dosage,\"Toxicity\"',\n",
       " 'Efficacy',\n",
       " 'Efficacy,\"Metabolism/PK\"',\n",
       " 'Efficacy,\"Other\"',\n",
       " 'Efficacy,\"Toxicity\"',\n",
       " 'Efficacy,\"Toxicity\",\"Metabolism/PK\"',\n",
       " 'Metabolism/PK',\n",
       " 'Other',\n",
       " 'Other,\"Metabolism/PK\"',\n",
       " 'PD',\n",
       " 'PD,\"Metabolism/PK\"',\n",
       " 'Toxicity',\n",
       " 'Toxicity,\"Metabolism/PK\"'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df_variant_merge[\"Phenotype Category\"].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac09da74",
   "metadata": {},
   "source": [
    "## 读取突变rsID与DNA改变位置的映射"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cfd33174",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_variants = pd.read_csv('variants/variants.tsv', sep='\\t', error_bad_lines=False).fillna(\"\")\n",
    "df_variants = df_variants[[\"Variant ID\", \"Variant Name\", \"Gene Symbols\",\n",
    "                           \"Location\", \"Synonyms\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6a83e31f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6400, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>Variant Name</th>\n",
       "      <th>Gene Symbols</th>\n",
       "      <th>Location</th>\n",
       "      <th>Synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA166156302</td>\n",
       "      <td>rs1000002</td>\n",
       "      <td>ABCC5</td>\n",
       "      <td>NC_000003.12:183917980</td>\n",
       "      <td>rs17623022, NG_047115.1:g.105031=, NC_000003.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA166156746</td>\n",
       "      <td>rs1000113</td>\n",
       "      <td>IRGM</td>\n",
       "      <td>NC_000005.10:150860514</td>\n",
       "      <td>1000113, NC_000005.9:g.150240076=, NC_000005.9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PA166195421</td>\n",
       "      <td>rs10006452</td>\n",
       "      <td>UGT2B7</td>\n",
       "      <td>NC_000004.12:69112090</td>\n",
       "      <td>10006452, NC_000004.12:g.69112090T&gt;A, 58882597...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PA166177121</td>\n",
       "      <td>rs10007051</td>\n",
       "      <td></td>\n",
       "      <td>NC_000004.12:129244309</td>\n",
       "      <td>NC_000004.11:g.130165464=, NC_000004.12:g.1292...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PA166156636</td>\n",
       "      <td>rs10008257</td>\n",
       "      <td></td>\n",
       "      <td>NC_000004.12:94435177</td>\n",
       "      <td>10008257, NC_000004.12:g.94435177=, rs10008257...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Variant ID Variant Name Gene Symbols                Location  \\\n",
       "0  PA166156302    rs1000002        ABCC5  NC_000003.12:183917980   \n",
       "1  PA166156746    rs1000113         IRGM  NC_000005.10:150860514   \n",
       "2  PA166195421   rs10006452       UGT2B7   NC_000004.12:69112090   \n",
       "3  PA166177121   rs10007051               NC_000004.12:129244309   \n",
       "4  PA166156636   rs10008257                NC_000004.12:94435177   \n",
       "\n",
       "                                            Synonyms  \n",
       "0  rs17623022, NG_047115.1:g.105031=, NC_000003.1...  \n",
       "1  1000113, NC_000005.9:g.150240076=, NC_000005.9...  \n",
       "2  10006452, NC_000004.12:g.69112090T>A, 58882597...  \n",
       "3  NC_000004.11:g.130165464=, NC_000004.12:g.1292...  \n",
       "4  10008257, NC_000004.12:g.94435177=, rs10008257...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df_variants.shape)\n",
    "df_variants[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a062594c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5440, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>guideline_consultant</th>\n",
       "      <th>guideline_name</th>\n",
       "      <th>drug</th>\n",
       "      <th>gene</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for abacavir and ...</td>\n",
       "      <td>abacavir</td>\n",
       "      <td>HLA_B</td>\n",
       "      <td>HLA-B*57:01:01</td>\n",
       "      <td>Presence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for allopurinol a...</td>\n",
       "      <td>allopurinol</td>\n",
       "      <td>HLA_B</td>\n",
       "      <td>HLA-B*58:01</td>\n",
       "      <td>Presence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 663A&gt;G</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 669T&gt;C</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 747A&gt;G</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 786G&gt;A</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 807A&gt;C</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 807A&gt;G</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 827A&gt;G</td>\n",
       "      <td>Normal risk of aminoglycoside-induced hearing ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CPIC_Guideline</td>\n",
       "      <td>Annotation of CPIC Guideline for amikacin, gen...</td>\n",
       "      <td>amikacin</td>\n",
       "      <td>MT_RNR1</td>\n",
       "      <td>MT-RNR1 839A&gt;G</td>\n",
       "      <td>Uncertain risk of aminoglycoside-induced heari...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  guideline_consultant                                     guideline_name  \\\n",
       "0       CPIC_Guideline  Annotation of CPIC Guideline for abacavir and ...   \n",
       "1       CPIC_Guideline  Annotation of CPIC Guideline for allopurinol a...   \n",
       "2       CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "3       CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "4       CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "5       CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "6       CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "7       CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "8       CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "9       CPIC_Guideline  Annotation of CPIC Guideline for amikacin, gen...   \n",
       "\n",
       "          drug     gene       haplotype  \\\n",
       "0     abacavir    HLA_B  HLA-B*57:01:01   \n",
       "1  allopurinol    HLA_B     HLA-B*58:01   \n",
       "2     amikacin  MT_RNR1  MT-RNR1 663A>G   \n",
       "3     amikacin  MT_RNR1  MT-RNR1 669T>C   \n",
       "4     amikacin  MT_RNR1  MT-RNR1 747A>G   \n",
       "5     amikacin  MT_RNR1  MT-RNR1 786G>A   \n",
       "6     amikacin  MT_RNR1  MT-RNR1 807A>C   \n",
       "7     amikacin  MT_RNR1  MT-RNR1 807A>G   \n",
       "8     amikacin  MT_RNR1  MT-RNR1 827A>G   \n",
       "9     amikacin  MT_RNR1  MT-RNR1 839A>G   \n",
       "\n",
       "                                                term  \n",
       "0                                           Presence  \n",
       "1                                           Presence  \n",
       "2  Uncertain risk of aminoglycoside-induced heari...  \n",
       "3  Uncertain risk of aminoglycoside-induced heari...  \n",
       "4  Uncertain risk of aminoglycoside-induced heari...  \n",
       "5  Uncertain risk of aminoglycoside-induced heari...  \n",
       "6  Uncertain risk of aminoglycoside-induced heari...  \n",
       "7  Uncertain risk of aminoglycoside-induced heari...  \n",
       "8  Normal risk of aminoglycoside-induced hearing ...  \n",
       "9  Uncertain risk of aminoglycoside-induced heari...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folder = \"guideline_annotation\"\n",
    "guideline_list = []\n",
    "haplotype_list = []\n",
    "term_list = []\n",
    "drug_list = []\n",
    "gene_list = []\n",
    "guideline_name_list = []\n",
    "\n",
    "for path in os.listdir(folder):\n",
    "    if path.endswith(\".json\") and \"Annotation_of_\" in path:\n",
    "        with open(os.path.join(folder, path), \"r\", encoding=\"utf-8\") as f:\n",
    "            guideline_dict = json.load(f)\n",
    "\n",
    "        path = path.replace(\"Annotation_of_\", \"\").replace(\".json\", \"\")\n",
    "        try:\n",
    "            guideline, drug_gene = path.split(\"_for_\")\n",
    "        except:\n",
    "            continue\n",
    "        drugs, genes = drug_gene.split(\"_and_\")\n",
    "        drugs = list(filter(lambda x: x!= \"\", [x.strip() for x in drugs.split(\"_\")]))\n",
    "\n",
    "        tmp_haplotype_list = []\n",
    "        tmp_term_list = []\n",
    "        for guideline_gene in guideline_dict[\"guideline\"][\"guidelineGenes\"]:\n",
    "            for allele in guideline_gene[\"alleles\"]:\n",
    "                if \"haplotype\" in allele.keys():\n",
    "                    tmp_haplotype_list.append(allele[\"haplotype\"][\"symbol\"])\n",
    "                    tmp_term_list.append(allele[\"function\"][\"term\"])\n",
    "        \n",
    "        guideline_name = guideline_dict[\"guideline\"][\"name\"]\n",
    "        \n",
    "        guideline_name_list.extend([guideline_name] * len(tmp_haplotype_list) * len(drugs))\n",
    "        guideline_list.extend([guideline] * len(tmp_haplotype_list) * len(drugs))\n",
    "        gene_list.extend([genes] * len(tmp_haplotype_list) * len(drugs))\n",
    "        # data alignment\n",
    "        for drug in drugs:\n",
    "            drug_list.extend([drug] * len(tmp_haplotype_list))\n",
    "            haplotype_list.extend(tmp_haplotype_list)\n",
    "            term_list.extend(tmp_term_list)\n",
    "        \n",
    "\n",
    "df_guideline = pd.DataFrame({\n",
    "    \"guideline_consultant\": guideline_list,\n",
    "    \"guideline_name\": guideline_name_list,\n",
    "    \"drug\": drug_list,\n",
    "    \"gene\": gene_list,\n",
    "    \"haplotype\": haplotype_list,\n",
    "    \"term\": term_list\n",
    "})\n",
    "\n",
    "print(df_guideline.shape)\n",
    "df_guideline[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b92b12a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "66dfaa61",
   "metadata": {},
   "source": [
    "## Read relationships data\n",
    "\n",
    "- Evidence\n",
    "    - VIP, VariantAnnotation, ClinicalAnnotation, DosingGuideline, DrugLabel or Pathway. Comma separated list\n",
    "        because the evidence for a relationship could come from multiple sources in PharmGKB.\n",
    "- Association\n",
    "    - \"associated\" means an association between the entities is supported by the \"Evidence\" and \"PMIDs\" columns.\n",
    "    - \"not associated\" means that the entities were evaluated but not found have a statistically significant association based on the “PMIDs” column.\n",
    "    - \"ambiguous\" means that some of the items in the \"Evidence\" and/or \"PMIDs\"  columns support an association and others do not.\n",
    "    \n",
    "#### In the filtering, only read Association = associated rows, then remain (Variant <-> Disease), (Haplotype <-> Disease),  (Variant <-> Chemical), (Haplotype <-> Chemical) relationships\n",
    "\n",
    "#### relationship数据的置信度不是很高，故从clinical annotation，drug label annotation以及variant annatation数据中提取基因变异与药物的关系数据。仅仅提取relationship中 变异与疾病的关系(为了丰富图谱，没有实际运用场景)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "be04c6f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13980, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entity1_id</th>\n",
       "      <th>Entity1_name</th>\n",
       "      <th>Entity1_type</th>\n",
       "      <th>Entity2_id</th>\n",
       "      <th>Entity2_name</th>\n",
       "      <th>Entity2_type</th>\n",
       "      <th>Evidence</th>\n",
       "      <th>Association</th>\n",
       "      <th>PK</th>\n",
       "      <th>PD</th>\n",
       "      <th>PMIDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA443334</td>\n",
       "      <td>Pain Insensitivity, Congenital</td>\n",
       "      <td>Disease</td>\n",
       "      <td>PA166155817</td>\n",
       "      <td>rs12478318</td>\n",
       "      <td>Variant</td>\n",
       "      <td>VariantAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>21939494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PA445553</td>\n",
       "      <td>Rheumatic Heart Disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>PA165816542</td>\n",
       "      <td>CYP2C9*1</td>\n",
       "      <td>Haplotype</td>\n",
       "      <td>VariantAnnotation</td>\n",
       "      <td>associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>27335128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Entity1_id                    Entity1_name Entity1_type   Entity2_id  \\\n",
       "0   PA443334  Pain Insensitivity, Congenital      Disease  PA166155817   \n",
       "1   PA445553         Rheumatic Heart Disease      Disease  PA165816542   \n",
       "\n",
       "  Entity2_name Entity2_type           Evidence Association PK PD     PMIDs  \n",
       "0   rs12478318      Variant  VariantAnnotation  associated        21939494  \n",
       "1     CYP2C9*1    Haplotype  VariantAnnotation  associated        27335128  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_relation = pd.read_csv('relationship/relationships_associated.tsv', sep='\\t').fillna(\"\")\n",
    "print(df_relation.shape)\n",
    "df_relation[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b77ff407",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bf63bb67",
   "metadata": {},
   "source": [
    "## Note\n",
    "\n",
    "#### 从PGKB中提取药物代谢与哪些基因有关\n",
    "| 数据源 | 头节点 | 关系 | 尾节点 | 注解 |\n",
    "| --- | --- | --- | --- | --- |\n",
    "| pgkb clinical annotation | Gene | affact (clinical) | Chemical | 在临床上，哪些基因会影响药物的药理反应 |\n",
    "| pgkb variant annotation | Gene | affact (research) | Chemical | 在科研中，哪些基因会影响药物的药理反应 |\n",
    "\n",
    "#### 从PGKB中提取rsID / haplotype 与变异位点的映射关系\n",
    "| 数据源 | 头节点 | 关系 | 尾节点 | 注解 |\n",
    "| --- | --- | --- | --- | --- |\n",
    "| pgkb variants| rsID | position_where | mutation position | rsID对应的突变位置是在哪里 |\n",
    "| pgkb allele definition | haplotype | position_where | mutation position | haplotyps对应的突变位置是在哪里 |\n",
    "\n",
    "#### 从PGKB中提取变异与药物的关系\n",
    "| 数据源 | 头节点 | 关系 | 尾节点 | 注解 |\n",
    "| --- | --- | --- | --- | --- |\n",
    "| pgkb clinical annotatoin | Variant / Haplotypes | affact_Metabolism/PK (clinical) | Chemical | 在临床上，某一个突变影响到对药物的代谢 |\n",
    "| pgkb clinical annotatoin | Variant / Haplotypes | affact_Dosage (clinical) | Chemical | 在临床上，某一个突变影响到对药物的剂量 |\n",
    "| pgkb clinical annotatoin | Variant / Haplotypes | affact_Efficacy (clinical) | Chemical | 在临床上，某一个突变影响到对药物的有效性 |\n",
    "| pgkb clinical annotatoin | Variant / Haplotypes | affact_Toxicity (clinical) | Chemical | 在临床上，某一个突变影响到对药物的毒性 |\n",
    "| pgkb clinical annotatoin | Variant / Haplotypes | affact_Other (clinical) | Chemical | 在临床上，某一个突变影响到对药物的其他因素 |\n",
    "| pgkb drug label annotation | Variant / Haplotypes | test_required (drug label) | Chemical | 要使用某种药，需要对特定基因检测后才可使用 |\n",
    "| pgkb drug label annotation | Variant / Haplotypes | test_recommended (drug label) | Chemical | 要使用某种药，建议对特定基因检测后才可使用 |\n",
    "| pgkb variant annotation | Variant / Haplotypes | affact_Metabolism/PK (research) | Chemical | 在研究性的药理反应实验中，某一个突变影响到对药物的代谢 |\n",
    "| pgkb variant annotation | Variant / Haplotypes | affact_Dosage (research) | Chemical | 在研究性的药理反应实验中，某一个突变影响到对药物的剂量 |\n",
    "| pgkb variant annotation | Variant / Haplotypes | affact_Efficacy (research) | Chemical | 在研究性的药理反应实验中，某一个突变影响到对药物的有效性 |\n",
    "| pgkb variant annotatoin | Variant / Haplotypes | affact_Toxicity (research) | Chemical | 在研究性的药理反应实验中，某一个突变影响到对药物的毒性 |\n",
    "| pgkb variant annotatoin | Variant / Haplotypes | affact_PD (research) | Chemical | 在研究性的药理反应实验中，某一个突变影响到对药物的PD |\n",
    "| pgkb variant annotatoin | Variant / Haplotypes | affact_Other (research) | Chemical | 在研究性的药理反应实验中，某一个突变影响到对药物的其他因素 |\n",
    "\n",
    "#### 基因与变异的关系\n",
    "| 数据源 | 头节点 | 关系 | 尾节点 | 注解 |\n",
    "| --- | --- | --- | --- | --- |\n",
    "| pgkb variants | Variant | is_from | Gene | 在某个基因上的变异 |\n",
    "| pgkb allele definition | Haplotypes | is_from | Gene | 在某个基因上的突变型别 |\n",
    "\n",
    "#### 疾病与突变的关系 (为了丰富图谱，没有实际运用场景)\n",
    "| 数据源 | 头节点 | 关系 | 尾节点 | 注解 |\n",
    "| --- | --- | --- | --- | --- |\n",
    "| pgkb relationship | Variant / Haplotypes | influence | disease | 某种疾病由突变或突变性别影响 |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "930fec24",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36959448",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
