{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9949cd4b",
   "metadata": {},
   "source": [
    "## 生成科研数据variant annotation的关系数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9d980acb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'D:\\\\pgkb_graph'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "import re\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1f9e85ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5072, 11)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "b'Skipping line 4308: expected 11 fields, saw 13\\n'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>Specialty Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1183684657</td>\n",
       "      <td>CYP2D6 ultrarapid metabolizer genotype</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>tramadol</td>\n",
       "      <td>18204346</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Median (+)R,R-tramadol area under the curve wa...</td>\n",
       "      <td>CYP2D6 ultra-metabolizer genotype is associate...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1448997750</td>\n",
       "      <td>CYP2B6*1, CYP2B6*18</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>efavirenz</td>\n",
       "      <td>16495778</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note that in the paper the allele was r...</td>\n",
       "      <td>CYP2B6 *1/*18 is associated with increased con...</td>\n",
       "      <td>*1/*18</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Variant Annotation ID                      Variant/Haplotypes    Gene  \\\n",
       "0             1183684657  CYP2D6 ultrarapid metabolizer genotype  CYP2D6   \n",
       "1             1448997750                     CYP2B6*1, CYP2B6*18  CYP2B6   \n",
       "\n",
       "     Drug(s)      PMID Phenotype Category Significance  \\\n",
       "0   tramadol  18204346      Metabolism/PK          yes   \n",
       "1  efavirenz  16495778      Metabolism/PK          yes   \n",
       "\n",
       "                                               Notes  \\\n",
       "0  Median (+)R,R-tramadol area under the curve wa...   \n",
       "1  Please note that in the paper the allele was r...   \n",
       "\n",
       "                                            Sentence Alleles  \\\n",
       "0  CYP2D6 ultra-metabolizer genotype is associate...           \n",
       "1  CYP2B6 *1/*18 is associated with increased con...  *1/*18   \n",
       "\n",
       "  Specialty Population  \n",
       "0                       \n",
       "1                       "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_annotation = pd.read_csv('variant_annotation/var_drug_ann.tsv', sep='\\t', error_bad_lines=False).fillna(\"\")\n",
    "df_variant_annotation = df_variant_annotation[df_variant_annotation[\"Significance\"] == \"yes\"]\n",
    "df_variant_annotation.index = range(len(df_variant_annotation))\n",
    "print(df_variant_annotation.shape)\n",
    "df_variant_annotation[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f8a02a80",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3344, 21)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>...</th>\n",
       "      <th>Study Type</th>\n",
       "      <th>Study Cases</th>\n",
       "      <th>Study Controls</th>\n",
       "      <th>Characteristics</th>\n",
       "      <th>Characteristics Type</th>\n",
       "      <th>Frequency In Cases</th>\n",
       "      <th>Frequency In Controls</th>\n",
       "      <th>P Value</th>\n",
       "      <th>Biogeographical Groups</th>\n",
       "      <th>p_value_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1183684657</td>\n",
       "      <td>CYP2D6 ultrarapid metabolizer genotype</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>tramadol</td>\n",
       "      <td>18204346</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Median (+)R,R-tramadol area under the curve wa...</td>\n",
       "      <td>CYP2D6 ultra-metabolizer genotype is associate...</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>cohort</td>\n",
       "      <td>14</td>\n",
       "      <td></td>\n",
       "      <td>11 UM; 3 PM; healthy male non-smoking volunteers</td>\n",
       "      <td>Study Cohort</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt; 0.001</td>\n",
       "      <td>European</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1448997750</td>\n",
       "      <td>CYP2B6*1, CYP2B6*18</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>efavirenz</td>\n",
       "      <td>16495778</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note that in the paper the allele was r...</td>\n",
       "      <td>CYP2B6 *1/*18 is associated with increased con...</td>\n",
       "      <td>*1/*18</td>\n",
       "      <td>...</td>\n",
       "      <td>cohort</td>\n",
       "      <td>51</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt; 0.0001</td>\n",
       "      <td>Multiple groups</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1183684661</td>\n",
       "      <td>CYP2D6 normal metabolizer genotype</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>tramadol</td>\n",
       "      <td>18204346</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Median (+)R,R-tramadol area under the curve wa...</td>\n",
       "      <td>CYP2D6 normal metabolizer genotype is associat...</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>cohort</td>\n",
       "      <td>14</td>\n",
       "      <td></td>\n",
       "      <td>11 UM; 3 PM; healthy male non-smoking volunteers</td>\n",
       "      <td>Study Cohort</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&lt; 0.001</td>\n",
       "      <td>European</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1449191920</td>\n",
       "      <td>rs113993960</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ivacaftor / lumacaftor</td>\n",
       "      <td>28325531</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Study found a significant improvement in FEV i...</td>\n",
       "      <td>Genotype del/del is associated with response t...</td>\n",
       "      <td>del/del</td>\n",
       "      <td>...</td>\n",
       "      <td>cohort</td>\n",
       "      <td>53</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>= 0.009</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1183632366</td>\n",
       "      <td>CYP2D6*1, CYP2D6*4</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>codeine</td>\n",
       "      <td>19940985</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Lower plasma concentrations of the morphine me...</td>\n",
       "      <td>CYP2D6 *4/*4 is associated with decreased meta...</td>\n",
       "      <td>*4/*4</td>\n",
       "      <td>...</td>\n",
       "      <td>cohort</td>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>End-stage renal disease patients with hemodial...</td>\n",
       "      <td>Study Cohort</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>= 0.01</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1449716526</td>\n",
       "      <td>CYP3A4*1, CYP3A4*1G</td>\n",
       "      <td>CYP3A4</td>\n",
       "      <td>fentanyl</td>\n",
       "      <td>21223952</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>rs2242480 (*1G) was the only variant genotyped...</td>\n",
       "      <td>CYP3A4 *1G/*1G is associated with increased co...</td>\n",
       "      <td>*1G/*1G</td>\n",
       "      <td>...</td>\n",
       "      <td>cohort</td>\n",
       "      <td>176</td>\n",
       "      <td></td>\n",
       "      <td>All participants were given fentanyl for analg...</td>\n",
       "      <td>Study Cohort</td>\n",
       "      <td>0.227</td>\n",
       "      <td></td>\n",
       "      <td>&lt; 0.01</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1451161540</td>\n",
       "      <td>CYP2D6*1, CYP2D6*2, CYP2D6*2xN</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>methadone</td>\n",
       "      <td>21589866</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>CYP2D6 ultrarapid metabolizers had significant...</td>\n",
       "      <td>CYP2D6 *1/*2xN + *2/*2xN (assigned as ultrarap...</td>\n",
       "      <td>*1/*2xN + *2/*2xN</td>\n",
       "      <td>...</td>\n",
       "      <td>cohort</td>\n",
       "      <td>105</td>\n",
       "      <td></td>\n",
       "      <td>P-value given for (R,S)-methadone concentrations</td>\n",
       "      <td>Study Cohort</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>= 0.002</td>\n",
       "      <td>European</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1448994156</td>\n",
       "      <td>CYP2C9*1, CYP2C9*3</td>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>phenytoin</td>\n",
       "      <td>14659971</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>No *3/*3 individuals were observed.</td>\n",
       "      <td>CYP2C9 *1/*3 is associated with increased conc...</td>\n",
       "      <td>*1/*3</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>28</td>\n",
       "      <td></td>\n",
       "      <td>serum phenytoin concentration to drug dose</td>\n",
       "      <td>Drug</td>\n",
       "      <td>0.107</td>\n",
       "      <td></td>\n",
       "      <td>= 0.01</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1451163180</td>\n",
       "      <td>rs4149056</td>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>atorvastatin</td>\n",
       "      <td>32128760</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td></td>\n",
       "      <td>Allele C is associated with increased exposure...</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>GWAS</td>\n",
       "      <td>590</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>= 2.21e-6</td>\n",
       "      <td>European</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1451163194</td>\n",
       "      <td>rs887829</td>\n",
       "      <td>UGT1A1,\"UGT1A10\",\"UGT1A3\",\"UGT1A4\",\"UGT1A5\",\"U...</td>\n",
       "      <td>atorvastatin</td>\n",
       "      <td>32128760</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>This variant was associated with increased met...</td>\n",
       "      <td>Allele T is associated with increased metaboli...</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>GWAS</td>\n",
       "      <td>590</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>= 7.25e-16</td>\n",
       "      <td>European</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Variant Annotation ID                      Variant/Haplotypes  \\\n",
       "0              1183684657  CYP2D6 ultrarapid metabolizer genotype   \n",
       "1              1448997750                     CYP2B6*1, CYP2B6*18   \n",
       "2              1183684661      CYP2D6 normal metabolizer genotype   \n",
       "3              1449191920                             rs113993960   \n",
       "4              1183632366                      CYP2D6*1, CYP2D6*4   \n",
       "13             1449716526                     CYP3A4*1, CYP3A4*1G   \n",
       "20             1451161540          CYP2D6*1, CYP2D6*2, CYP2D6*2xN   \n",
       "21             1448994156                      CYP2C9*1, CYP2C9*3   \n",
       "31             1451163180                               rs4149056   \n",
       "32             1451163194                                rs887829   \n",
       "\n",
       "                                                 Gene                 Drug(s)  \\\n",
       "0                                              CYP2D6                tramadol   \n",
       "1                                              CYP2B6               efavirenz   \n",
       "2                                              CYP2D6                tramadol   \n",
       "3                                                CFTR  ivacaftor / lumacaftor   \n",
       "4                                              CYP2D6                 codeine   \n",
       "13                                             CYP3A4                fentanyl   \n",
       "20                                             CYP2D6               methadone   \n",
       "21                                             CYP2C9               phenytoin   \n",
       "31                                            SLCO1B1            atorvastatin   \n",
       "32  UGT1A1,\"UGT1A10\",\"UGT1A3\",\"UGT1A4\",\"UGT1A5\",\"U...            atorvastatin   \n",
       "\n",
       "        PMID Phenotype Category Significance  \\\n",
       "0   18204346      Metabolism/PK          yes   \n",
       "1   16495778      Metabolism/PK          yes   \n",
       "2   18204346      Metabolism/PK          yes   \n",
       "3   28325531           Efficacy          yes   \n",
       "4   19940985      Metabolism/PK          yes   \n",
       "13  21223952      Metabolism/PK          yes   \n",
       "20  21589866      Metabolism/PK          yes   \n",
       "21  14659971      Metabolism/PK          yes   \n",
       "31  32128760      Metabolism/PK          yes   \n",
       "32  32128760      Metabolism/PK          yes   \n",
       "\n",
       "                                                Notes  \\\n",
       "0   Median (+)R,R-tramadol area under the curve wa...   \n",
       "1   Please note that in the paper the allele was r...   \n",
       "2   Median (+)R,R-tramadol area under the curve wa...   \n",
       "3   Study found a significant improvement in FEV i...   \n",
       "4   Lower plasma concentrations of the morphine me...   \n",
       "13  rs2242480 (*1G) was the only variant genotyped...   \n",
       "20  CYP2D6 ultrarapid metabolizers had significant...   \n",
       "21                No *3/*3 individuals were observed.   \n",
       "31                                                      \n",
       "32  This variant was associated with increased met...   \n",
       "\n",
       "                                             Sentence            Alleles  ...  \\\n",
       "0   CYP2D6 ultra-metabolizer genotype is associate...                     ...   \n",
       "1   CYP2B6 *1/*18 is associated with increased con...             *1/*18  ...   \n",
       "2   CYP2D6 normal metabolizer genotype is associat...                     ...   \n",
       "3   Genotype del/del is associated with response t...            del/del  ...   \n",
       "4   CYP2D6 *4/*4 is associated with decreased meta...              *4/*4  ...   \n",
       "13  CYP3A4 *1G/*1G is associated with increased co...            *1G/*1G  ...   \n",
       "20  CYP2D6 *1/*2xN + *2/*2xN (assigned as ultrarap...  *1/*2xN + *2/*2xN  ...   \n",
       "21  CYP2C9 *1/*3 is associated with increased conc...              *1/*3  ...   \n",
       "31  Allele C is associated with increased exposure...                  C  ...   \n",
       "32  Allele T is associated with increased metaboli...                  T  ...   \n",
       "\n",
       "   Study Type Study Cases Study Controls  \\\n",
       "0      cohort          14                  \n",
       "1      cohort          51                  \n",
       "2      cohort          14                  \n",
       "3      cohort          53                  \n",
       "4      cohort          11                  \n",
       "13     cohort         176                  \n",
       "20     cohort         105                  \n",
       "21                     28                  \n",
       "31       GWAS         590                  \n",
       "32       GWAS         590                  \n",
       "\n",
       "                                      Characteristics Characteristics Type  \\\n",
       "0    11 UM; 3 PM; healthy male non-smoking volunteers         Study Cohort   \n",
       "1                                                                  Unknown   \n",
       "2    11 UM; 3 PM; healthy male non-smoking volunteers         Study Cohort   \n",
       "3                                                                  Unknown   \n",
       "4   End-stage renal disease patients with hemodial...         Study Cohort   \n",
       "13  All participants were given fentanyl for analg...         Study Cohort   \n",
       "20   P-value given for (R,S)-methadone concentrations         Study Cohort   \n",
       "21         serum phenytoin concentration to drug dose                 Drug   \n",
       "31                                                                 Unknown   \n",
       "32                                                                 Unknown   \n",
       "\n",
       "   Frequency In Cases Frequency In Controls     P Value  \\\n",
       "0                                               < 0.001   \n",
       "1                                              < 0.0001   \n",
       "2                                               < 0.001   \n",
       "3                                               = 0.009   \n",
       "4                                                = 0.01   \n",
       "13              0.227                            < 0.01   \n",
       "20                                              = 0.002   \n",
       "21              0.107                            = 0.01   \n",
       "31                                            = 2.21e-6   \n",
       "32                                           = 7.25e-16   \n",
       "\n",
       "   Biogeographical Groups p_value_effect  \n",
       "0                European           True  \n",
       "1         Multiple groups           True  \n",
       "2                European           True  \n",
       "3                 Unknown           True  \n",
       "4                 Unknown           True  \n",
       "13             East Asian           True  \n",
       "20               European           True  \n",
       "21             East Asian           True  \n",
       "31               European           True  \n",
       "32               European           True  \n",
       "\n",
       "[10 rows x 21 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_param = pd.read_csv('variant_annotation/study_parameters.tsv', sep='\\t', error_bad_lines=False).fillna(\"\")\n",
    "df_variant_param = df_variant_param[[\"Variant Annotation ID\", \"Study Type\", \"Study Cases\", \"Study Controls\", \"Characteristics\",\n",
    "                                     \"Characteristics Type\", \"Frequency In Cases\", \"Frequency In Controls\", \"P Value\", \"Biogeographical Groups\"]]\n",
    "df_variant_merge = pd.merge(df_variant_annotation, df_variant_param, how=\"left\", left_on=\"Variant Annotation ID\", right_on=\"Variant Annotation ID\")\n",
    "\n",
    "# set 0.01 as p_value threshold to filter\n",
    "p_value_list = df_variant_merge[\"P Value\"].values\n",
    "p_value_list = list(map(lambda x: \"1\" if \">\" in str(x) else str(x).replace(\"<\", \"\").replace(\"=\", \"\").strip(), p_value_list))\n",
    "p_value_bool_list = []\n",
    "for x in p_value_list:\n",
    "    try:\n",
    "        if float(x) <= 0.01:\n",
    "            p_value_bool_list.append(True)\n",
    "        else:\n",
    "            p_value_bool_list.append(False)\n",
    "    except:\n",
    "        p_value_bool_list.append(False)\n",
    "df_variant_merge = df_variant_merge.assign(p_value_effect=p_value_bool_list)\n",
    "df_variant_merge = df_variant_merge[df_variant_merge[\"p_value_effect\"] == True]\n",
    "\n",
    "print(df_variant_merge.shape)\n",
    "df_variant_merge[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "61f93d23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Variant Annotation ID', 'Variant/Haplotypes', 'Gene', 'Drug(s)',\n",
       "       'PMID', 'Phenotype Category', 'Significance', 'Notes', 'Sentence',\n",
       "       'Alleles', 'Specialty Population', 'Study Type', 'Study Cases',\n",
       "       'Study Controls', 'Characteristics', 'Characteristics Type',\n",
       "       'Frequency In Cases', 'Frequency In Controls', 'P Value',\n",
       "       'Biogeographical Groups', 'p_value_effect'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_merge.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e2723356",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list(filter(lambda x: \",\" in x, df_variant_merge[\"Variant/Haplotypes\"].values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "063c1640",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_drug_research_list = []\n",
    "diplotype_drug_research_list = []\n",
    "\n",
    "for index, row in df_variant_merge.iterrows():\n",
    "    variant = row[\"Variant/Haplotypes\"]\n",
    "    if \"genotype\" in variant:\n",
    "        continue\n",
    "    variant_list = filter(lambda x: x!= \"\", [x.strip() for x in variant.split(\",\")])\n",
    "        \n",
    "    drugs = row[\"Drug(s)\"]\n",
    "    drug_list = list(filter(lambda x: x!= \"\", [x.replace(\"\\\"\", \"\").strip() for x in re.split(\"/|,\", drugs)]))\n",
    "    p_value = \"P value {}\".format(row[\"P Value\"]).replace(\"=\", \"\").replace(\"<\", \"\")\n",
    "    phenotype = row[\"Phenotype Category\"]\n",
    "    phenotype_list = set([x.replace(\"\\\"\", \"\").strip() for x in phenotype.split(\",\")])\n",
    "    PMID = row[\"PMID\"]\n",
    "    note = row[\"Notes\"]\n",
    "    sentence = row[\"Sentence\"]\n",
    "    biogeo_group = row[\"Biogeographical Groups\"].replace(\"\\\"\", \"'\")            \n",
    "    \n",
    "    if row[\"Alleles\"] != \"\" and \"/\" in row[\"Alleles\"]:\n",
    "        diplotype_list = [x.strip() for x in row[\"Alleles\"].split(\"+\")]\n",
    "        for d in drug_list:\n",
    "            for dip in diplotype_list:\n",
    "                for p in phenotype_list:\n",
    "                    if \"/\" in dip:\n",
    "                        diplotype = \"{} {}\".format(row[\"Gene\"], dip)\n",
    "                        diplotype_drug_research_list.append((d, p, p_value, biogeo_group, PMID, note, sentence, diplotype))\n",
    "    else:\n",
    "        for d in drug_list:\n",
    "            for v in variant_list:\n",
    "                for p in phenotype_list:\n",
    "                    variant_drug_research_list.append((d, p, p_value, biogeo_group, PMID, note, sentence, v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "34251919",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3147\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('atorvastatin',\n",
       "  'Metabolism/PK',\n",
       "  'P value  2.21e-6',\n",
       "  'European',\n",
       "  32128760,\n",
       "  '',\n",
       "  'Allele C is associated with increased exposure to atorvastatin as compared to allele T.',\n",
       "  'rs4149056'),\n",
       " ('atorvastatin',\n",
       "  'Metabolism/PK',\n",
       "  'P value  7.25e-16',\n",
       "  'European',\n",
       "  32128760,\n",
       "  'This variant was associated with increased metabolic ratios of both 2-OH ATV/ATV (P = 7.25 × 10−16) and 2-OH ATV L/ATV L (P = 3.95 × 10−15) at genome-wide significance. This variant is associated with increased hydroxylation of ATV, and not lactonization.',\n",
       "  'Allele T is associated with increased metabolism of atorvastatin as compared to allele C.',\n",
       "  'rs887829'),\n",
       " ('atorvastatin',\n",
       "  'Metabolism/PK',\n",
       "  'P value  6.18e-7',\n",
       "  'European',\n",
       "  32128760,\n",
       "  'This variant was associated with increased metabolic ratios of 2-OH ATV/ATV (P = 6.18 × 10−7) (increased hydroxylation of ATV) at genome-wide significance.',\n",
       "  'Allele G is associated with increased metabolism of atorvastatin as compared to allele T.',\n",
       "  'rs45446698'),\n",
       " ('aripiprazole',\n",
       "  'Metabolism/PK',\n",
       "  'P value  0.000',\n",
       "  'East Asian',\n",
       "  33586456,\n",
       "  'Authors state that Cmax, t1/2, Tmax and area under the concentration-time curve (AUC0-∞) were all significant although table shows p-values greater than 0.05 for t1/2.',\n",
       "  'Genotype GG is associated with increased concentrations of aripiprazole in healthy individuals as compared to genotypes CC + CG.',\n",
       "  'rs1058164'),\n",
       " ('aripiprazole',\n",
       "  'Metabolism/PK',\n",
       "  'P value  0.000',\n",
       "  'East Asian',\n",
       "  33586456,\n",
       "  'Authors state that Cmax, t1/2, Tmax and area under the concentration-time curve (AUC0-∞) were all significant although table shows p-values greater than 0.05 for t1/2.',\n",
       "  'Genotype AA is associated with increased concentrations of aripiprazole in healthy individuals as compared to genotypes AC + CC.',\n",
       "  'rs28371699')]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(variant_drug_research_list))\n",
    "variant_drug_research_list[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9f0dcf73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1394\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('efavirenz',\n",
       "  'Metabolism/PK',\n",
       "  'P value  0.0001',\n",
       "  'Multiple groups',\n",
       "  16495778,\n",
       "  'Please note that in the paper the allele was referred to as CYP2B6*16. CYP2B6*16 and *18 alleles have been consolidated by PharmVar in Jan 2020, with *16 now listed as a suballele of *18 (CYP2B6*18.002). This annotation is updated to be on CYP2B6*18, instead of CYP2B6*16.',\n",
       "  'CYP2B6 *1/*18 is associated with increased concentrations of efavirenz in people with HIV Infections as compared to CYP2B6 *1/*1.',\n",
       "  'CYP2B6 *1/*18'),\n",
       " ('ivacaftor',\n",
       "  'Efficacy',\n",
       "  'P value  0.009',\n",
       "  'Unknown',\n",
       "  28325531,\n",
       "  \"Study found a significant improvement in FEV in all patients following 3 months of treatment. There was no difference in patients' BMI.\",\n",
       "  'Genotype del/del is associated with response to ivacaftor / lumacaftor in people with Cystic Fibrosis.',\n",
       "  'CFTR del/del'),\n",
       " ('lumacaftor',\n",
       "  'Efficacy',\n",
       "  'P value  0.009',\n",
       "  'Unknown',\n",
       "  28325531,\n",
       "  \"Study found a significant improvement in FEV in all patients following 3 months of treatment. There was no difference in patients' BMI.\",\n",
       "  'Genotype del/del is associated with response to ivacaftor / lumacaftor in people with Cystic Fibrosis.',\n",
       "  'CFTR del/del'),\n",
       " ('codeine',\n",
       "  'Metabolism/PK',\n",
       "  'P value  0.01',\n",
       "  'Unknown',\n",
       "  19940985,\n",
       "  'Lower plasma concentrations of the morphine metabolite M3G and undetectable levels of M6G were found in the two poor metabolizers (*4/*4) compared to extensive metabolizers (*1/*1 +*1/*4). PM were determined on the basis of their CYP2D6*3, *4, and *6 alleles.',\n",
       "  'CYP2D6 *4/*4 is associated with decreased metabolism of codeine in people with Kidney Failure, Chronic as compared to CYP2D6 *1/*1 + *1/*4.',\n",
       "  'CYP2D6 *4/*4'),\n",
       " ('fentanyl',\n",
       "  'Metabolism/PK',\n",
       "  'P value  0.01',\n",
       "  'East Asian',\n",
       "  21223952,\n",
       "  'rs2242480 (*1G) was the only variant genotyped in the study.',\n",
       "  'CYP3A4 *1G/*1G is associated with increased concentrations of fentanyl in people with Pain, Postoperative as compared to CYP3A4 *1/*1 + *1/*1G.',\n",
       "  'CYP3A4 *1G/*1G')]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(diplotype_drug_research_list))\n",
    "diplotype_drug_research_list[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "4bfff7e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(variant_drug_research_list, \n",
    "             columns=[\"drug\", \"phenotype_category\", \"p_value\", \"bio_geo_group\", \"PMID\", \"note\", \"sentence\", \"variant\"]\n",
    "            ).assign(data_source=[\"research\"] * len(variant_drug_research_list)\n",
    "                    ).to_csv(\"processed/research_drug_variant_annotation.csv\", index=False)\n",
    "\n",
    "pd.DataFrame(diplotype_drug_research_list, \n",
    "             columns=[\"drug\", \"phenotype_category\", \"p_value\", \"bio_geo_group\", \"PMID\", \"note\", \"sentence\", \"diplotype\"]\n",
    "            ).assign(data_source=[\"research\"] * len(diplotype_drug_research_list)\n",
    "                    ).to_csv(\"processed/research_drug_diplotype_annotation.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "955379e0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed8e51bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
