{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "278676e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'D:\\\\pgkb_graph'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from py2neo import Graph \n",
    "import pandas as pd\n",
    "import re\n",
    "import os\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0f93652d",
   "metadata": {},
   "outputs": [],
   "source": [
    "session = Graph(\"neo4j://172.16.229.46:7687\", auth=(\"neo4j\", \"123456\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "78472a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "haplotype_guideline_cypher = \"\"\"\n",
    "MATCH (hap:haplotype)-[r:variant_guideline]->(che:chemical) \n",
    "WITH hap, r, che\n",
    "MATCH (hap)-[]->(ge:gene)\n",
    "RETURN ge.gene_name as gene_name, hap.variant_name as haplotype_name, hap.is_reference as is_reference,\n",
    "hap.NC_change_code as NC_code, hap.NG_change_code as NG_code, hap.nucleotide_change_code as nucleotide_code, \n",
    "hap.protein_change_code as protein_code, hap.mapped_rsID as mapped_rsID, \n",
    "r.phenotype as phenotype, r.implication as implication, r.implication_chn as implication_chn, \n",
    "r.recommendation as recommendation, r.recommendation_chn as recommendation_chn,\n",
    "r.genotype as genotype, r.genotype_chn as genotype_chn,\n",
    "r.organization as organization, r.link as link, \n",
    "che.chemical_name as chemical_name, che.atc_code as atc_code, \n",
    "che.L1_info as L1_info, che.L1_info_chn as L1_info_chn,\n",
    "che.L2_info as L2_info, che.L2_info_chn as L2_info_chn,\n",
    "che.L3_info as L3_info, che.L3_info_chn as L3_info_chn,\n",
    "che.L4_info as L4_info, che.L4_info_chn as L4_info_chn\n",
    "\"\"\"\n",
    "\n",
    "result_list = session.run(haplotype_guideline_cypher).data()\n",
    "df_haplotype_guideline = pd.DataFrame(result_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3e5589ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>haplotype_name</th>\n",
       "      <th>is_reference</th>\n",
       "      <th>NC_code</th>\n",
       "      <th>NG_code</th>\n",
       "      <th>nucleotide_code</th>\n",
       "      <th>protein_code</th>\n",
       "      <th>mapped_rsID</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>implication</th>\n",
       "      <th>...</th>\n",
       "      <th>chemical_name</th>\n",
       "      <th>atc_code</th>\n",
       "      <th>L1_info</th>\n",
       "      <th>L1_info_chn</th>\n",
       "      <th>L2_info</th>\n",
       "      <th>L2_info_chn</th>\n",
       "      <th>L3_info</th>\n",
       "      <th>L3_info_chn</th>\n",
       "      <th>L4_info</th>\n",
       "      <th>L4_info_chn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RYR1</td>\n",
       "      <td>RYR1 c.103T&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>Position at NC_000019.10 (Homo sapiens chromos...</td>\n",
       "      <td>Position at NG_008866.1 (RYR1 RefSeqGene; forw...</td>\n",
       "      <td>NM_000540.2; ['c.103T&gt;C']</td>\n",
       "      <td>Effect on protein (NP_000531.2); ['C35R']</td>\n",
       "      <td>rs193922747</td>\n",
       "      <td>Malignant Hyperthermia susceptible</td>\n",
       "      <td>Individuals are at increased risk of developin...</td>\n",
       "      <td>...</td>\n",
       "      <td>desflurane</td>\n",
       "      <td>N01AB07</td>\n",
       "      <td>N: NERVOUS SYSTEM</td>\n",
       "      <td>N：神经系统</td>\n",
       "      <td>N01: ANESTHETICS</td>\n",
       "      <td>N01：麻醉品</td>\n",
       "      <td>N01A: ANESTHETICS, GENERAL</td>\n",
       "      <td>N01A：麻醉品，一般</td>\n",
       "      <td>N01AB: Halogenated hydrocarbons</td>\n",
       "      <td>N01AB：卤化碳氢化合物</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RYR1</td>\n",
       "      <td>RYR1 c.130C&gt;T</td>\n",
       "      <td>False</td>\n",
       "      <td>Position at NC_000019.10 (Homo sapiens chromos...</td>\n",
       "      <td>Position at NG_008866.1 (RYR1 RefSeqGene; forw...</td>\n",
       "      <td>NM_000540.2; ['c.130C&gt;T']</td>\n",
       "      <td>Effect on protein (NP_000531.2); ['R44C']</td>\n",
       "      <td>rs193922748</td>\n",
       "      <td>Malignant Hyperthermia susceptible</td>\n",
       "      <td>Individuals are at increased risk of developin...</td>\n",
       "      <td>...</td>\n",
       "      <td>desflurane</td>\n",
       "      <td>N01AB07</td>\n",
       "      <td>N: NERVOUS SYSTEM</td>\n",
       "      <td>N：神经系统</td>\n",
       "      <td>N01: ANESTHETICS</td>\n",
       "      <td>N01：麻醉品</td>\n",
       "      <td>N01A: ANESTHETICS, GENERAL</td>\n",
       "      <td>N01A：麻醉品，一般</td>\n",
       "      <td>N01AB: Halogenated hydrocarbons</td>\n",
       "      <td>N01AB：卤化碳氢化合物</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name haplotype_name is_reference  \\\n",
       "0      RYR1  RYR1 c.103T>C        False   \n",
       "1      RYR1  RYR1 c.130C>T        False   \n",
       "\n",
       "                                             NC_code  \\\n",
       "0  Position at NC_000019.10 (Homo sapiens chromos...   \n",
       "1  Position at NC_000019.10 (Homo sapiens chromos...   \n",
       "\n",
       "                                             NG_code  \\\n",
       "0  Position at NG_008866.1 (RYR1 RefSeqGene; forw...   \n",
       "1  Position at NG_008866.1 (RYR1 RefSeqGene; forw...   \n",
       "\n",
       "             nucleotide_code                               protein_code  \\\n",
       "0  NM_000540.2; ['c.103T>C']  Effect on protein (NP_000531.2); ['C35R']   \n",
       "1  NM_000540.2; ['c.130C>T']  Effect on protein (NP_000531.2); ['R44C']   \n",
       "\n",
       "   mapped_rsID                           phenotype  \\\n",
       "0  rs193922747  Malignant Hyperthermia susceptible   \n",
       "1  rs193922748  Malignant Hyperthermia susceptible   \n",
       "\n",
       "                                         implication  ... chemical_name  \\\n",
       "0  Individuals are at increased risk of developin...  ...    desflurane   \n",
       "1  Individuals are at increased risk of developin...  ...    desflurane   \n",
       "\n",
       "  atc_code            L1_info L1_info_chn           L2_info L2_info_chn  \\\n",
       "0  N01AB07  N: NERVOUS SYSTEM      N：神经系统  N01: ANESTHETICS     N01：麻醉品   \n",
       "1  N01AB07  N: NERVOUS SYSTEM      N：神经系统  N01: ANESTHETICS     N01：麻醉品   \n",
       "\n",
       "                      L3_info  L3_info_chn                          L4_info  \\\n",
       "0  N01A: ANESTHETICS, GENERAL  N01A：麻醉品，一般  N01AB: Halogenated hydrocarbons   \n",
       "1  N01A: ANESTHETICS, GENERAL  N01A：麻醉品，一般  N01AB: Halogenated hydrocarbons   \n",
       "\n",
       "     L4_info_chn  \n",
       "0  N01AB：卤化碳氢化合物  \n",
       "1  N01AB：卤化碳氢化合物  \n",
       "\n",
       "[2 rows x 27 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_haplotype_guideline[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b6451a66",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "diplotype_guideline_cypher = \"\"\"\n",
    "MATCH (dip:diplotype)-[r:variant_guideline]->(che:chemical) \n",
    "WITH dip, r, che\n",
    "MATCH (dip)-[:diplotype_consist_of]->(hap:haplotype)-[:mutation_at]->(ge:gene)\n",
    "RETURN distinct ge.gene_name as gene_name,\n",
    "dip.diplotype_name as diplotype_name, r.phenotype as phenotype, \n",
    "hap.variant_name as haplotype_name, hap.is_reference as is_reference, hap.NC_change_code as NC_code, hap.NG_change_code as NG_code, \n",
    "hap.nucleotide_change_code as nucleotide_code, hap.protein_change_code as protein_code, hap.mapped_rsID as mapped_rsID,\n",
    "r.implication as implication, r.implication_chn as implication_chn, \n",
    "r.recommendation as recommendation, r.recommendation_chn as recommendation_chn, \n",
    "r.genotype as genotype, r.genotype_chn as genotype_chn,\n",
    "r.organization as organization, r.link as link, \n",
    "che.chemical_name as chemical_name, che.atc_code as atc_code, \n",
    "che.L1_info as L1_info, che.L1_info_chn as L1_info_chn,\n",
    "che.L2_info as L2_info, che.L2_info_chn as L2_info_chn,\n",
    "che.L3_info as L3_info, che.L3_info_chn as L3_info_chn,\n",
    "che.L4_info as L4_info, che.L4_info_chn as L4_info_chn\n",
    "\"\"\"\n",
    "\n",
    "result_list = session.run(diplotype_guideline_cypher).data()\n",
    "df_diplotype_guideline = pd.DataFrame(result_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b1c2dfc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>diplotype_name</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>haplotype_name</th>\n",
       "      <th>is_reference</th>\n",
       "      <th>NC_code</th>\n",
       "      <th>NG_code</th>\n",
       "      <th>nucleotide_code</th>\n",
       "      <th>protein_code</th>\n",
       "      <th>mapped_rsID</th>\n",
       "      <th>...</th>\n",
       "      <th>chemical_name</th>\n",
       "      <th>atc_code</th>\n",
       "      <th>L1_info</th>\n",
       "      <th>L1_info_chn</th>\n",
       "      <th>L2_info</th>\n",
       "      <th>L2_info_chn</th>\n",
       "      <th>L3_info</th>\n",
       "      <th>L3_info_chn</th>\n",
       "      <th>L4_info</th>\n",
       "      <th>L4_info_chn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-B</td>\n",
       "      <td>HLA-B *57:01/*X</td>\n",
       "      <td>High risk of hypersensitivity (~6% of patients)</td>\n",
       "      <td>HLA-B*57:01</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>abacavir</td>\n",
       "      <td>J05AF06</td>\n",
       "      <td>J: ANTIINFECTIVES FOR SYSTEMIC USE</td>\n",
       "      <td>J：系统使用的抗助液</td>\n",
       "      <td>J05: ANTIVIRALS FOR SYSTEMIC USE</td>\n",
       "      <td>J05：用于系统性使用的抗病毒</td>\n",
       "      <td>J05A: DIRECT ACTING ANTIVIRALS</td>\n",
       "      <td>J05A：直接代理抗病毒</td>\n",
       "      <td>J05AF: Nucleoside and nucleotide reverse trans...</td>\n",
       "      <td>J05AF：核苷和核苷酸逆转录酶抑制剂</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HLA-B</td>\n",
       "      <td>HLA-B *57:01/*57:01</td>\n",
       "      <td>High risk of hypersensitivity (~6% of patients)</td>\n",
       "      <td>HLA-B*57:01</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>abacavir</td>\n",
       "      <td>J05AF06</td>\n",
       "      <td>J: ANTIINFECTIVES FOR SYSTEMIC USE</td>\n",
       "      <td>J：系统使用的抗助液</td>\n",
       "      <td>J05: ANTIVIRALS FOR SYSTEMIC USE</td>\n",
       "      <td>J05：用于系统性使用的抗病毒</td>\n",
       "      <td>J05A: DIRECT ACTING ANTIVIRALS</td>\n",
       "      <td>J05A：直接代理抗病毒</td>\n",
       "      <td>J05AF: Nucleoside and nucleotide reverse trans...</td>\n",
       "      <td>J05AF：核苷和核苷酸逆转录酶抑制剂</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name       diplotype_name  \\\n",
       "0     HLA-B      HLA-B *57:01/*X   \n",
       "1     HLA-B  HLA-B *57:01/*57:01   \n",
       "\n",
       "                                         phenotype haplotype_name  \\\n",
       "0  High risk of hypersensitivity (~6% of patients)    HLA-B*57:01   \n",
       "1  High risk of hypersensitivity (~6% of patients)    HLA-B*57:01   \n",
       "\n",
       "  is_reference NC_code NG_code nucleotide_code protein_code mapped_rsID  ...  \\\n",
       "0         None                                                           ...   \n",
       "1         None                                                           ...   \n",
       "\n",
       "  chemical_name atc_code                             L1_info L1_info_chn  \\\n",
       "0      abacavir  J05AF06  J: ANTIINFECTIVES FOR SYSTEMIC USE  J：系统使用的抗助液   \n",
       "1      abacavir  J05AF06  J: ANTIINFECTIVES FOR SYSTEMIC USE  J：系统使用的抗助液   \n",
       "\n",
       "                            L2_info      L2_info_chn  \\\n",
       "0  J05: ANTIVIRALS FOR SYSTEMIC USE  J05：用于系统性使用的抗病毒   \n",
       "1  J05: ANTIVIRALS FOR SYSTEMIC USE  J05：用于系统性使用的抗病毒   \n",
       "\n",
       "                          L3_info   L3_info_chn  \\\n",
       "0  J05A: DIRECT ACTING ANTIVIRALS  J05A：直接代理抗病毒   \n",
       "1  J05A: DIRECT ACTING ANTIVIRALS  J05A：直接代理抗病毒   \n",
       "\n",
       "                                             L4_info          L4_info_chn  \n",
       "0  J05AF: Nucleoside and nucleotide reverse trans...  J05AF：核苷和核苷酸逆转录酶抑制剂  \n",
       "1  J05AF: Nucleoside and nucleotide reverse trans...  J05AF：核苷和核苷酸逆转录酶抑制剂  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_diplotype_guideline[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "26a644cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "diplotype_metabolizer_cypher = \"\"\"\n",
    "MATCH (dip:diplotype)-[r:diplotype_metabolizer]->(che:chemical) \n",
    "WITH dip, r, che\n",
    "MATCH (dip)-[:diplotype_consist_of]->(hap:haplotype)-[:mutation_at]->(ge:gene)\n",
    "RETURN distinct ge.gene_name as gene_name,\n",
    "dip.diplotype_name as diplotype_name, r.phenotype as phenotype, r.phenotype_category as metabolizer,\n",
    "hap.variant_name as haplotype_name, hap.is_reference as is_reference, hap.NC_change_code as NC_code, \n",
    "hap.NG_change_code as NG_code, hap.nucleotide_change_code as nucleotide_code, \n",
    "hap.protein_change_code as protein_code, hap.mapped_rsID as mapped_rsID,\n",
    "r.implication as implication, r.implication_chn as implication_chn,\n",
    "r.recommendation as recommendation, r.recommendation_chn as recommendation_chn,\n",
    "r.genotype as genotype, \n",
    "r.organization as organization, r.link as link, \n",
    "che.chemical_name as chemical_name, che.atc_code as atc_code, \n",
    "che.L1_info as L1_info, che.L1_info_chn as L1_info_chn,\n",
    "che.L2_info as L2_info, che.L2_info_chn as L2_info_chn,\n",
    "che.L3_info as L3_info, che.L3_info_chn as L3_info_chn,\n",
    "che.L4_info as L4_info, che.L4_info_chn as L4_info_chn\n",
    "\"\"\"\n",
    "result_list = session.run(diplotype_metabolizer_cypher).data()\n",
    "df_diplotype_metabolizer = pd.DataFrame(result_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8ec17cfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>diplotype_name</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>metabolizer</th>\n",
       "      <th>haplotype_name</th>\n",
       "      <th>is_reference</th>\n",
       "      <th>NC_code</th>\n",
       "      <th>NG_code</th>\n",
       "      <th>nucleotide_code</th>\n",
       "      <th>protein_code</th>\n",
       "      <th>...</th>\n",
       "      <th>chemical_name</th>\n",
       "      <th>atc_code</th>\n",
       "      <th>L1_info</th>\n",
       "      <th>L1_info_chn</th>\n",
       "      <th>L2_info</th>\n",
       "      <th>L2_info_chn</th>\n",
       "      <th>L3_info</th>\n",
       "      <th>L3_info_chn</th>\n",
       "      <th>L4_info</th>\n",
       "      <th>L4_info_chn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TPMT</td>\n",
       "      <td>TPMT *1/*2</td>\n",
       "      <td>IM (one inactive allele: *2, *3, *4-*18)</td>\n",
       "      <td>Intermediate Metabolizer</td>\n",
       "      <td>TPMT*2</td>\n",
       "      <td>None</td>\n",
       "      <td>Position at NC_000006.12 (Homo sapiens chromos...</td>\n",
       "      <td>Position at NG_012137.2 (TPMT RefSeqGene); ['g...</td>\n",
       "      <td>Nucleotide change on cDNA from http://www.imh....</td>\n",
       "      <td>Effect on protein (NP_000358.1); ['A80P']</td>\n",
       "      <td>...</td>\n",
       "      <td>mercaptopurine</td>\n",
       "      <td>L01BB02</td>\n",
       "      <td>L: ANTINEOPLASTIC AND IMMUNOMODULATING AGENTS</td>\n",
       "      <td>L：抗肿瘤和免疫调节剂</td>\n",
       "      <td>L01: ANTINEOPLASTIC AGENTS</td>\n",
       "      <td>L01：抗肿瘤剂</td>\n",
       "      <td>L01B: ANTIMETABOLITES</td>\n",
       "      <td>L01B：抗体代谢物</td>\n",
       "      <td>L01BB: Purine analogues</td>\n",
       "      <td>L01BB：嘌呤类似物</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TPMT</td>\n",
       "      <td>TPMT *1/*2</td>\n",
       "      <td>IM (one inactive allele: *2, *3, *4-*18)</td>\n",
       "      <td>Intermediate Metabolizer</td>\n",
       "      <td>TPMT*1</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>mercaptopurine</td>\n",
       "      <td>L01BB02</td>\n",
       "      <td>L: ANTINEOPLASTIC AND IMMUNOMODULATING AGENTS</td>\n",
       "      <td>L：抗肿瘤和免疫调节剂</td>\n",
       "      <td>L01: ANTINEOPLASTIC AGENTS</td>\n",
       "      <td>L01：抗肿瘤剂</td>\n",
       "      <td>L01B: ANTIMETABOLITES</td>\n",
       "      <td>L01B：抗体代谢物</td>\n",
       "      <td>L01BB: Purine analogues</td>\n",
       "      <td>L01BB：嘌呤类似物</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name diplotype_name                                 phenotype  \\\n",
       "0      TPMT     TPMT *1/*2  IM (one inactive allele: *2, *3, *4-*18)   \n",
       "1      TPMT     TPMT *1/*2  IM (one inactive allele: *2, *3, *4-*18)   \n",
       "\n",
       "                metabolizer haplotype_name is_reference  \\\n",
       "0  Intermediate Metabolizer         TPMT*2         None   \n",
       "1  Intermediate Metabolizer         TPMT*1         None   \n",
       "\n",
       "                                             NC_code  \\\n",
       "0  Position at NC_000006.12 (Homo sapiens chromos...   \n",
       "1                                                      \n",
       "\n",
       "                                             NG_code  \\\n",
       "0  Position at NG_012137.2 (TPMT RefSeqGene); ['g...   \n",
       "1                                                      \n",
       "\n",
       "                                     nucleotide_code  \\\n",
       "0  Nucleotide change on cDNA from http://www.imh....   \n",
       "1                                                      \n",
       "\n",
       "                                protein_code  ...   chemical_name atc_code  \\\n",
       "0  Effect on protein (NP_000358.1); ['A80P']  ...  mercaptopurine  L01BB02   \n",
       "1                                             ...  mercaptopurine  L01BB02   \n",
       "\n",
       "                                         L1_info  L1_info_chn  \\\n",
       "0  L: ANTINEOPLASTIC AND IMMUNOMODULATING AGENTS  L：抗肿瘤和免疫调节剂   \n",
       "1  L: ANTINEOPLASTIC AND IMMUNOMODULATING AGENTS  L：抗肿瘤和免疫调节剂   \n",
       "\n",
       "                      L2_info L2_info_chn                L3_info L3_info_chn  \\\n",
       "0  L01: ANTINEOPLASTIC AGENTS    L01：抗肿瘤剂  L01B: ANTIMETABOLITES  L01B：抗体代谢物   \n",
       "1  L01: ANTINEOPLASTIC AGENTS    L01：抗肿瘤剂  L01B: ANTIMETABOLITES  L01B：抗体代谢物   \n",
       "\n",
       "                   L4_info  L4_info_chn  \n",
       "0  L01BB: Purine analogues  L01BB：嘌呤类似物  \n",
       "1  L01BB: Purine analogues  L01BB：嘌呤类似物  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_diplotype_metabolizer[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "51161f9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_position(nc_code):\n",
    "    find_list = re.findall(r\"chromosome [\\d]+\", nc_code)\n",
    "    if len(find_list) > 0:\n",
    "        chr_name = find_list[0].replace(\"chromosome \", \"chr\")\n",
    "        position_list = [\n",
    "            \"{}:{}\".format(\n",
    "                chr_name,\n",
    "                \"\".join([\"{}:{}\".format(x.strip().replace(\"g.\", \"\").replace(y, \"\"), y) \n",
    "                         for y in re.findall(r\"[\\D]+\", x.strip().replace(\"g.\", \"\"))]) if \"del\" not in x else x.strip().replace(\"g.\", \"\")\n",
    "                ) \n",
    "            for x in nc_code.split(\";\")[1].strip().replace(\"[\", \"\").replace(\"]\", \"\"). replace(\"'\", \"\").split(\",\")\n",
    "        ]\n",
    "        return \", \".join(position_list)\n",
    "        \n",
    "    else:\n",
    "        return \"\"\n",
    "    \n",
    "df_diplotype_metabolizer = df_diplotype_metabolizer.assign(\n",
    "    position=[get_position(x) for x in list(df_diplotype_metabolizer[\"NC_code\"].values)])\n",
    "\n",
    "df_diplotype_guideline = df_diplotype_guideline.assign(\n",
    "    position=[get_position(x) for x in list(df_diplotype_guideline[\"NC_code\"].values)])\n",
    "\n",
    "df_haplotype_guideline = df_haplotype_guideline.assign(\n",
    "    position=[get_position(x) for x in list(df_haplotype_guideline[\"NC_code\"].values)])\n",
    "\n",
    "df_diplotype_guideline.to_csv(\"query_file/guideline_diplotype.csv\", index=False)\n",
    "df_haplotype_guideline.to_csv(\"query_file/guideline_haplotype.csv\", index=False)\n",
    "df_diplotype_metabolizer.to_csv(\"query_file/guideline_metabolizer.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a45ae4ab",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "group_list = []\n",
    "for key, content in df_diplotype_metabolizer.groupby([\"gene_name\", \"metabolizer\", \"organization\", \"chemical_name\"]):\n",
    "    content_dict = {\n",
    "        \"gene_name\": key[0],\n",
    "        \"metabolizer\": key[1],\n",
    "        \"organization\": key[2],\n",
    "        \"chemical_name\": key[3],\n",
    "        \"diplotype\": list(set(content[\"diplotype_name\"].values)),\n",
    "        \"haplotype\": list(set(content[\"haplotype_name\"].values)),\n",
    "        \"nc_code\": list(set(content[\"NC_code\"].values)),\n",
    "        \"ng_code\": list(set(content[\"NG_code\"].values)),\n",
    "        \"nucleotide_code\": list(set(content[\"nucleotide_code\"].values)),\n",
    "        \"protein_code\": list(set(content[\"protein_code\"].values)),\n",
    "        \"recommendation\": list(set(content[\"recommendation\"].values))[0],\n",
    "        \"recommendation_chn\": list(set(content[\"recommendation_chn\"].values))[0],\n",
    "        \"implication\": list(set(content[\"implication\"].values))[0],\n",
    "        \"implication_chn\": list(set(content[\"implication_chn\"].values))[0],\n",
    "        \"phenotype\": list(set(content[\"phenotype\"].values))[0],\n",
    "        \"link\": list(set(content[\"link\"].values))[0],\n",
    "        \"atc_code\": list(set(content[\"atc_code\"].values))[0],\n",
    "        \"L1_info\": list(set(content[\"L1_info\"].values))[0],\n",
    "        \"L1_info_chn\": list(set(content[\"L1_info_chn\"].values))[0],\n",
    "        \"L2_info\": list(set(content[\"L2_info\"].values))[0],\n",
    "        \"L2_info_chn\": list(set(content[\"L2_info_chn\"].values))[0],\n",
    "        \"L3_info\": list(set(content[\"L3_info\"].values))[0],\n",
    "        \"L3_info_chn\": list(set(content[\"L3_info_chn\"].values))[0],\n",
    "        \"L4_info\": list(set(content[\"L4_info\"].values))[0],\n",
    "        \"L4_info_chn\": list(set(content[\"L4_info_chn\"].values))[0],\n",
    "        \"position\": list(set([x.strip() for x in \",\".join(list(set(content[\"position\"].values))).split(\",\")])),\n",
    "        \"rsID\": list(set([x.strip() for x in \",\".join(list(set(content[\"mapped_rsID\"].values))).split(\",\")]))\n",
    "    }\n",
    "    group_list.append(content_dict)\n",
    "    \n",
    "pd.DataFrame(group_list).to_csv(\"query_file/guideline_metabolizer_group.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c9de7868",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_list = []\n",
    "for key, content in df_diplotype_guideline.groupby([\"gene_name\", \"organization\", \"chemical_name\", \"diplotype_name\"]):\n",
    "    content_dict = {\n",
    "        \"gene_name\": key[0],\n",
    "        \"metabolizer\": \"\",\n",
    "        \"organization\": key[1],\n",
    "        \"chemical_name\": key[2],\n",
    "        \"diplotype\": key[3],\n",
    "        \"haplotype\": list(set(content[\"haplotype_name\"].values)),\n",
    "        \"nc_code\": list(set(content[\"NC_code\"].values)),\n",
    "        \"ng_code\": list(set(content[\"NG_code\"].values)),\n",
    "        \"nucleotide_code\": list(set(content[\"nucleotide_code\"].values)),\n",
    "        \"protein_code\": list(set(content[\"protein_code\"].values)),\n",
    "        \"recommendation\": list(set(content[\"recommendation\"].values))[0],\n",
    "        \"recommendation_chn\": list(set(content[\"recommendation_chn\"].values))[0],\n",
    "        \"implication\": list(set(content[\"implication\"].values))[0],\n",
    "        \"implication_chn\": list(set(content[\"implication_chn\"].values))[0],\n",
    "        \"genotype\": list(set(content[\"genotype\"].values))[0],\n",
    "        \"genotype_chn\": list(set(content[\"genotype_chn\"].values))[0],\n",
    "        \"phenotype\": list(set(content[\"phenotype\"].values))[0],\n",
    "        \"link\": list(set(content[\"link\"].values))[0],\n",
    "        \"atc_code\": list(set(content[\"atc_code\"].values))[0],\n",
    "        \"L1_info\": list(set(content[\"L1_info\"].values))[0],\n",
    "        \"L1_info_chn\": list(set(content[\"L1_info_chn\"].values))[0],\n",
    "        \"L2_info\": list(set(content[\"L2_info\"].values))[0],\n",
    "        \"L2_info_chn\": list(set(content[\"L2_info_chn\"].values))[0],\n",
    "        \"L3_info\": list(set(content[\"L3_info\"].values))[0],\n",
    "        \"L3_info_chn\": list(set(content[\"L3_info_chn\"].values))[0],\n",
    "        \"L4_info\": list(set(content[\"L4_info\"].values))[0],\n",
    "        \"L4_info_chn\": list(set(content[\"L4_info_chn\"].values))[0],\n",
    "        \"position\": list(set([x.strip() for x in \",\".join(list(set(content[\"position\"].values))).split(\",\")])),\n",
    "        \"rsID\": list(set([x.strip() for x in \",\".join(list(set(content[\"mapped_rsID\"].values))).split(\",\")]))\n",
    "    }\n",
    "    group_list.append(content_dict)\n",
    "    \n",
    "pd.DataFrame(group_list).to_csv(\"query_file/guideline_diplotype_group.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ac44ef9a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "group_list = []\n",
    "for key, content in df_haplotype_guideline.groupby([\"gene_name\", \"organization\", \"chemical_name\", \"haplotype_name\"]):\n",
    "    content_dict = {\n",
    "        \"gene_name\": key[0],\n",
    "        \"metabolizer\": \"\",\n",
    "        \"organization\": key[1],\n",
    "        \"chemical_name\": key[2],\n",
    "        \"diplotype\": \"\",\n",
    "        \"haplotype\": key[3],\n",
    "        \"nc_code\": list(set(content[\"NC_code\"].values)),\n",
    "        \"ng_code\": list(set(content[\"NG_code\"].values)),\n",
    "        \"nucleotide_code\": list(set(content[\"nucleotide_code\"].values)),\n",
    "        \"protein_code\": list(set(content[\"protein_code\"].values)),\n",
    "        \"recommendation\": list(set(content[\"recommendation\"].values))[0],\n",
    "        \"recommendation_chn\": list(set(content[\"recommendation_chn\"].values))[0],\n",
    "        \"implication\": list(set(content[\"implication\"].values))[0],\n",
    "        \"implication_chn\": list(set(content[\"implication_chn\"].values))[0],\n",
    "        \"genotype\": list(set(content[\"genotype\"].values))[0],\n",
    "        \"genotype_chn\": list(set(content[\"genotype_chn\"].values))[0],\n",
    "        \"phenotype\": list(set(content[\"phenotype\"].values))[0],\n",
    "        \"link\": list(set(content[\"link\"].values))[0],\n",
    "        \"atc_code\": list(set(content[\"atc_code\"].values))[0],\n",
    "        \"L1_info\": list(set(content[\"L1_info\"].values))[0],\n",
    "        \"L1_info_chn\": list(set(content[\"L1_info_chn\"].values))[0],\n",
    "        \"L2_info\": list(set(content[\"L2_info\"].values))[0],\n",
    "        \"L2_info_chn\": list(set(content[\"L2_info_chn\"].values))[0],\n",
    "        \"L3_info\": list(set(content[\"L3_info\"].values))[0],\n",
    "        \"L3_info_chn\": list(set(content[\"L3_info_chn\"].values))[0],\n",
    "        \"L4_info\": list(set(content[\"L4_info\"].values))[0],\n",
    "        \"L4_info_chn\": list(set(content[\"L4_info_chn\"].values))[0],\n",
    "        \"position\": list(set([x.strip() for x in \",\".join(list(set(content[\"position\"].values))).split(\",\")])),\n",
    "        \"rsID\": list(set([x.strip() for x in \",\".join(list(set(content[\"mapped_rsID\"].values))).split(\",\")]))\n",
    "    }\n",
    "    group_list.append(content_dict)\n",
    "    \n",
    "pd.DataFrame(group_list).to_csv(\"query_file/guideline_haplotype_group.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2cf675d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4b2473c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "cpic_drug_label_cypher = \"\"\"\n",
    "MATCH (ge:gene)-[r:cpic_guideline]->(che:chemical) \n",
    "RETURN ge.gene_name as gene_name, ge.OMIM as OMIM, ge.chromosome as chromosome, ge.refseq as refseq,\n",
    "ge.chromosomal_start_GRCh38 as chromosomal_start, ge.chromosomal_stop_GRCh38 as chromosomal_stop, ge.is_VIP as is_VIP,\n",
    "r.CPIC_level_status as CPIC_level_status, r.CPIC_level as CPIC_level, r.FDA_PGx_label as FDA_PGx_label,\n",
    "r.PGKB_evidence_level as PGKB_evidence_level, r.guideline_link as link,\n",
    "che.chemical_name as chemical_name, che.atc_code as atc_code, \n",
    "che.L1_info as L1_info, che.L1_info_chn as L1_info_chn,\n",
    "che.L2_info as L2_info, che.L2_info_chn as L2_info_chn,\n",
    "che.L3_info as L3_info, che.L3_info_chn as L3_info_chn,\n",
    "che.L4_info as L4_info, che.L4_info_chn as L4_info_chn,\n",
    "\"CPIC\" as organization\n",
    "\"\"\"\n",
    "\n",
    "result_list = session.run(cpic_drug_label_cypher).data()\n",
    "df_cpic_drug_label = pd.DataFrame(result_list).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "cd45b856",
   "metadata": {},
   "outputs": [],
   "source": [
    "pgkb_drug_label_cypher = \"\"\"\n",
    "MATCH (ge:gene)-[r:drug_label]->(che:chemical) \n",
    "RETURN ge.gene_name as gene_name, ge.OMIM as OMIM, ge.chromosome as chromosome, ge.refseq as refseq,\n",
    "ge.chromosomal_start_GRCh38 as chromosomal_start, ge.chromosomal_stop_GRCh38 as chromosomal_stop, ge.is_VIP as is_VIP,\n",
    "r.link as link, r.label_name as label_name, r.testing_level as testing_level,\n",
    "che.chemical_name as chemical_name, che.atc_code as atc_code, \n",
    "che.L1_info as L1_info, che.L1_info_chn as L1_info_chn,\n",
    "che.L2_info as L2_info, che.L2_info_chn as L2_info_chn,\n",
    "che.L3_info as L3_info, che.L3_info_chn as L3_info_chn,\n",
    "che.L4_info as L4_info, che.L4_info_chn as L4_info_chn,\n",
    "r.organization as organization\n",
    "\"\"\"\n",
    "\n",
    "result_list = session.run(pgkb_drug_label_cypher).data()\n",
    "df_pgkb_drug_label = pd.DataFrame(result_list).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3cd2ed6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pgkb_drug_label.to_csv(\"query_file/pgkb_drug_label.csv\", index=False)\n",
    "df_cpic_drug_label.to_csv(\"query_file/cpic_drug_label.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8d56e6c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_clinical_cypher = \"\"\"\n",
    "MATCH (var:variant)-[r:clinical_annotation]->(che:chemical)\n",
    "WITH var, r, che\n",
    "MATCH (var)-[:mutation_at]->(ge:gene)\n",
    "return ge.gene_name as gene_name, ge.chromosome as chromosome,\n",
    "var.variant_name as variant_name, var.is_reference as is_reference,\n",
    "var.NC_change_code as NC_code, var.NG_change_code as NG_code, \n",
    "var.nucleotide_change_code as nucleotide_code, \n",
    "var.protein_change_code as protein_code, var.mapped_rsID as mapped_rsID, \n",
    "r.phenotype_category as phenotype_category, r.evidence_level as evidence_level, \n",
    "r.level_modifier as level_modifier, r.phenotype as phenotype, \n",
    "r.phenotype_chn as phenotype_chn, r.score as score, r.link as link,\n",
    "che.chemical_name as chemical_name, che.atc_code as atc_code, \n",
    "che.L1_info as L1_info, che.L1_info_chn as L1_info_chn,\n",
    "che.L2_info as L2_info, che.L2_info_chn as L2_info_chn,\n",
    "che.L3_info as L3_info, che.L3_info_chn as L3_info_chn,\n",
    "che.L4_info as L4_info, che.L4_info_chn as L4_info_chn\n",
    "\"\"\"\n",
    "\n",
    "result_list = session.run(variant_clinical_cypher).data()\n",
    "df_variant_clinical = pd.DataFrame(result_list).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "474580b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "group_list = []\n",
    "for key, content in df_variant_clinical.groupby([\"variant_name\", \"phenotype_category\", \"chemical_name\"]):\n",
    "    content_dict = {\n",
    "        \"gene_name\": list(set(content[\"gene_name\"].values))[0],\n",
    "        \"chromosome\": list(set(content[\"chromosome\"].values))[0],\n",
    "        \"variant_name\": key[0],\n",
    "        \"phenotype_category\": key[1],\n",
    "        \"chemical_name\": key[2],\n",
    "        \"nc_code\": list(set(content[\"NC_code\"].values)),\n",
    "        \"ng_code\": list(set(content[\"NG_code\"].values)),\n",
    "        \"nucleotide_code\": list(set(content[\"nucleotide_code\"].values)),\n",
    "        \"protein_code\": list(set(content[\"protein_code\"].values)),\n",
    "        \"evidence_level\": list(set(content[\"evidence_level\"].values))[0],\n",
    "        \"level_modifier\": list(set(content[\"level_modifier\"].values))[0],\n",
    "        \"phenotype\": list(set(content[\"phenotype\"].values)),\n",
    "        \"phenotype_chn\": list(set(content[\"phenotype_chn\"].values)),\n",
    "        \"link\": list(set(content[\"link\"].values))[0],\n",
    "        \"atc_code\": list(set(content[\"atc_code\"].values))[0],\n",
    "        \"L1_info\": list(set(content[\"L1_info\"].values))[0],\n",
    "        \"L1_info_chn\": list(set(content[\"L1_info_chn\"].values))[0],\n",
    "        \"L2_info\": list(set(content[\"L2_info\"].values))[0],\n",
    "        \"L2_info_chn\": list(set(content[\"L2_info_chn\"].values))[0],\n",
    "        \"L3_info\": list(set(content[\"L3_info\"].values))[0],\n",
    "        \"L3_info_chn\": list(set(content[\"L3_info_chn\"].values))[0],\n",
    "        \"L4_info\": list(set(content[\"L4_info\"].values))[0],\n",
    "        \"L4_info_chn\": list(set(content[\"L4_info_chn\"].values))[0],\n",
    "        \"rsID\": list(set([x.strip() for x in \",\".join(list(set(content[\"mapped_rsID\"].values))).split(\",\")]))\n",
    "    }\n",
    "    group_list.append(content_dict)\n",
    "df_variant_clinical = pd.DataFrame(group_list)\n",
    "df_variant_clinical = df_variant_clinical.sort_values(by=[\"evidence_level\"])\n",
    "df_variant_clinical.index = range(len(df_variant_clinical))\n",
    "df_variant_clinical.to_csv(\"query_file/variant_clinical.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd74d47e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0242b5c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_str = \"\"\"\n",
    "match (hap:haplotype) \n",
    "return hap.variant_name as variant_name, hap.NC_change_code as NC_change_code, \n",
    "hap.mapped_rsID as mapped_rsID\n",
    "\"\"\"\n",
    "\n",
    "result_list = session.run(query_str).data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "15e80c0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_position(nc_code):\n",
    "    find_list = re.findall(r\"chromosome [\\dX]+\", nc_code)\n",
    "    if len(find_list) > 0:\n",
    "        chr_name = find_list[0].replace(\"chromosome \", \"chr\")\n",
    "        position_list = [\n",
    "            \"{}:{}\".format(\n",
    "                chr_name,\n",
    "                \"\".join([\"{}:{}\".format(x.strip().replace(\"g.\", \"\").replace(y, \"\"), y) \n",
    "                         for y in re.findall(r\"[\\D]+\", x.strip().replace(\"g.\", \"\"))]) if \"del\" not in x else x.strip().replace(\"g.\", \"\")\n",
    "                ) \n",
    "            for x in nc_code.split(\";\")[1].strip().replace(\"[\", \"\").replace(\"]\", \"\"). replace(\"'\", \"\").split(\",\")\n",
    "        ]\n",
    "        return position_list\n",
    "        \n",
    "    else:\n",
    "        return []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1d771236",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mapping = pd.DataFrame(result_list)\n",
    "df_mapping = df_mapping[(df_mapping[\"NC_change_code\"] != \"\") & (df_mapping[\"mapped_rsID\"] != \"\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "49fac926",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mapping = df_mapping.sort_values(by=[\"variant_name\"])\n",
    "position_list = [get_position(x) for x in df_mapping[\"NC_change_code\"].values]\n",
    "df_mapping = df_mapping.assign(position=position_list)\n",
    "rsID_list = [list(filter(lambda x: x!= \"\", [x.strip() for x in x.split(\",\")])) for x in df_mapping[\"mapped_rsID\"].values]\n",
    "df_mapping = df_mapping.assign(rsID=rsID_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "05e1a2bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mapping.to_csv(\"processed/position_map.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "59df41e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4aa29620",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsID_reverse_dict = defaultdict(list)\n",
    "rsID_dict = defaultdict(list)\n",
    "position_erverse_dict = defaultdict(list)\n",
    "position_dict = defaultdict(list)\n",
    "for index, row in df_mapping.iterrows():\n",
    "    for rs in row[\"rsID\"]:\n",
    "        rsID_reverse_dict[row[\"variant_name\"]].append(rs)\n",
    "        rsID_dict[rs].append(row[\"variant_name\"])\n",
    "    for pos in row[\"position\"]:\n",
    "        position_erverse_dict[row[\"variant_name\"]].append(pos)\n",
    "        position_dict[pos].append(row[\"variant_name\"])\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "276ddcb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rsID_to_haplotype(rsID_list):\n",
    "    result_list = []\n",
    "    rsID_list = list(filter(lambda x: x in rsID_dict.keys(), rsID_list))\n",
    "    rs_sub_dict = defaultdict(list)\n",
    "    \n",
    "    for rsID in rsID_list:\n",
    "        rs_sub_dict[rsID].extend(rsID_dict[rsID])\n",
    "    \n",
    "    rs_sub_dict = {x: list(set(y)) for x, y in rs_sub_dict.items()}\n",
    "    candidate_haplotype_list = []\n",
    "    sub_haplotype_dict = defaultdict(list)\n",
    "    \n",
    "    for x, y in rs_sub_dict.items():\n",
    "        candidate_haplotype_list.extend(y)\n",
    "        for i in y:\n",
    "            sub_haplotype_dict[i].append(x)\n",
    "            \n",
    "    candidate_haplotype_list = list(set(candidate_haplotype_list))\n",
    "    \n",
    "    candidate_dict = {x: rsID_reverse_dict[x] for x in candidate_haplotype_list}\n",
    "    sub_haplotype_dict = {x: list(set(y)) for x, y in sub_haplotype_dict.items()}\n",
    "    \n",
    "    for key in candidate_dict.keys():\n",
    "        if list(candidate_dict[key]) == list(sub_haplotype_dict[key]):\n",
    "            result_list.append(key)\n",
    "            \n",
    "    return result_list\n",
    "            \n",
    "    \n",
    "def position_to_haplotype(position_list):\n",
    "    result_list = []\n",
    "    position_list = list(filter(lambda x: x in position_dict.keys(), position_list))\n",
    "    pos_sub_list = defaultdict(list)\n",
    "    \n",
    "    for pos in position_list:\n",
    "        pos_sub_list[pos].extend(position_dict[pos])\n",
    "        \n",
    "    pos_sub_dict = {x: list(set(y)) for x, y in pos_sub_dict.items()}\n",
    "    candidate_haplotype_list = []\n",
    "    sub_haplotype_dict = defaultdict(list)\n",
    "    \n",
    "    for x, y in pos_sub_dict.items():\n",
    "        candidate_haplotype_list.extend(y)\n",
    "        for i in y:\n",
    "            sub_haplotype_dict[i].append(x)\n",
    "            \n",
    "    candidate_haplotype_list = list(set(candidate_haplotype_list))\n",
    "    \n",
    "    candidate_dict = {x: position_erverse_dict[x] for x in candidate_haplotype_list}\n",
    "    sub_haplotype_dict = {x: list(set(y)) for x, y in sub_haplotype_dict.items()}\n",
    "    \n",
    "    for key in candidate_dict.keys():\n",
    "        if list(candidate_dict[key]) == list(sub_haplotype_dict[key]):\n",
    "            result_list.append(key)\n",
    "            \n",
    "    return result_list\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cd34f20",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
