{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a80a998d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'D:\\\\drug KG\\\\pgkb'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "from collections import defaultdict\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e2be2758",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CACNA1S haplotype\\CACNA1S_haplotypes.xlsx\n",
      "CFTR haplotype\\CFTR_haplotypes.xlsx\n",
      "CYP2A6 haplotype\\CYP2A6_haplotypes.xlsx\n",
      "CYP2B6 haplotype\\CYP2B6_haplotypes.xlsx\n",
      "CYP2C19 haplotype\\CYP2C19_haplotypes.xlsx\n",
      "CYP2C9 haplotype\\CYP2C9_haplotypes.xlsx\n",
      "CYP2D6 haplotype\\CYP2D6_haplotypes.xlsx\n",
      "CYP3A4 haplotype\\CYP3A4_haplotypes.xlsx\n",
      "CYP3A5 haplotype\\CYP3A5_haplotypes.xlsx\n",
      "DPYD haplotype\\DPYD_haplotypes.xlsx\n",
      "G6PD haplotype\\G6PD_haplotypes.xlsx\n",
      "IFNL3 haplotype\\IFNL3_haplotypes.xlsx\n",
      "MT-RNR1 haplotype\\MT-RNR1_haplotypes.xlsx\n",
      "NAT2 haplotype\\NAT2_haplotypes.xlsx\n",
      "NUDT15 haplotype\\NUDT15_haplotypes.xlsx\n",
      "RYR1 haplotype\\RYR1_haplotypes.xlsx\n",
      "SLCO1B1 haplotype\\SLCO1B1_haplotypes.xlsx\n",
      "TPMT haplotype\\TPMT_haplotypes.xlsx\n",
      "UGT1A1 haplotype\\UGT1A1_haplotypes.xlsx\n"
     ]
    }
   ],
   "source": [
    "haplotype_folder = \"haplotype\"\n",
    "\n",
    "# haplotype <-> gene mentioned on clinical annnotation\n",
    "clinical_genes = [\n",
    "    'TPMT', 'NAT2', 'G6PD', 'CYP3A5', 'CYP2A6', 'CYP3A4', 'CYP2C19',\n",
    "    'UGT1A1', 'CYP2D6', 'NUDT15', 'CYP2B6', 'CYP2C9'\n",
    "] # and 'HLA-B', 'HLA-A'\n",
    "\n",
    "# haplotype <-> gene mentioned on guideline annotation\n",
    "guideline_genes = [\n",
    "    'CACNA1S', 'CFTR', 'CYP2C9', 'CYP2B6', 'CYP2D6', 'SLCO1B1',\n",
    "    'UGT1A1', 'DPYD', 'NUDT15', 'MT-RNR1', 'CYP3A5', 'TPMT',\n",
    "    'RYR1', 'CYP2C19', 'G6PD', 'IFNL3'\n",
    "] # and 'HLA-B', 'HLA-A'\n",
    "\n",
    "genes = set(clinical_genes + guideline_genes)\n",
    "\n",
    "# The position mapping for haplotype HLA-A and HLA-B is complex, skip these first.\n",
    "\n",
    "haplotype_path_dict = {g: \"\" for g in genes}\n",
    "for path in os.listdir(haplotype_folder):\n",
    "    for gene in genes:\n",
    "        if \"{}_haplotypes\".format(gene) in path:\n",
    "            haplotype_path_dict[gene] = os.path.join(haplotype_folder, path)\n",
    "for key,value in sorted(haplotype_path_dict.items(), key=lambda x: x[0]):\n",
    "    print(key, value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "448a2f8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "standard_haplotype_dict = {\n",
    "    \"CACNA1S\": \"Reference\",\n",
    "    \"CFTR\": \"standard\",\n",
    "    \"CYP2A6\": \"*1A\",\n",
    "    \"CYP2B6\": \"*1\",\n",
    "    \"CYP2C19\": \"*38\",\n",
    "    \"CYP2C9\": \"*1\",\n",
    "    \"CYP2D6\": \"*1\",\n",
    "    \"CYP3A4\": \"*1\",\n",
    "    \"CYP3A5\": \"*1\",\n",
    "    \"DPYD\": \"Reference\",\n",
    "    \"G6PD\": \"B (wildtype)\",\n",
    "    \"IFNL3\": \"\",\n",
    "    \"MT-RNR1\": \"Reference\",\n",
    "    \"NAT2\": \"*4\",\n",
    "    \"NUDT15\": \"*1\",\n",
    "    \"RYR1\": \"Reference\",\n",
    "    \"SLCO1B1\": \"*1A\",\n",
    "    \"TPMT\": \"*1\",\n",
    "    \"UGT1A1\": \"*1\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15591e71",
   "metadata": {},
   "source": [
    "### 翻转 dataframe，让第一列做column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "82340df9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def df_T_convert(df):\n",
    "    index = df.T.index\n",
    "    values = df.T.values\n",
    "    df_T = pd.DataFrame(values[1:, :], index=range(len(values) - 1), columns=values[0, :])\n",
    "    df_index = pd.DataFrame({index[0]: index[1:]})\n",
    "    df_T = pd.concat([df_index, df_T], axis=1)\n",
    "    return df_T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e8afacfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 处理TPMT gene\n",
    "df_tpmt = pd.read_excel(\n",
    "    haplotype_path_dict[\"TPMT\"],\n",
    "    engine=\"openpyxl\",\n",
    "    sheet_name=\"modified\"\n",
    ").fillna(\"\")\n",
    "df_tpmt_T = df_T_convert(df_tpmt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "293f990a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 处理 NAT2 gene\n",
    "df_nat2 = pd.read_excel(\n",
    "    haplotype_path_dict[\"NAT2\"],\n",
    "    engine=\"openpyxl\",\n",
    "    sheet_name=\"modified\"\n",
    ").fillna(\"\")\n",
    "df_nat2_T = df_T_convert(df_nat2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "603e1ca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 处理 G6PD gene\n",
    "df_G6PD = pd.read_excel(\n",
    "    haplotype_path_dict[\"G6PD\"],\n",
    "    engine=\"openpyxl\",\n",
    "    sheet_name=\"modified\"\n",
    ").fillna(\"\")\n",
    "df_G6PD_T = df_T_convert(df_G6PD)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cf91ded5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df_T_dict = {}\n",
    "for gene, path in haplotype_path_dict.items():\n",
    "    df_tmp = pd.read_excel(\n",
    "        path,\n",
    "        engine=\"openpyxl\",\n",
    "        sheet_name=\"modified\"\n",
    "    ).fillna(\"\")\n",
    "    df_tmp_T = df_T_convert(df_tmp)\n",
    "    gene_df_T_dict[gene] = df_tmp_T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bf78c180",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['IFNL3', 'TPMT', 'CACNA1S', 'CYP2B6', 'CYP2C9', 'NUDT15', 'MT-RNR1', 'CYP2C19', 'SLCO1B1', 'CYP3A4', 'CYP2D6', 'NAT2', 'RYR1', 'CFTR', 'DPYD', 'CYP2A6', 'UGT1A1', 'CYP3A5', 'G6PD'])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df_T_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fd78a004",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CACNA1S:\n",
      "Index(['Free Text', 'Protein Change NP_000060.2',\n",
      "       'Chromosome Position Change NC_000001.11',\n",
      "       'Gene position Change NG_009816.1 (minus)', 'rsID', 'Reference',\n",
      "       'c.520C>T', 'c.3257G>A'],\n",
      "      dtype='object')\n",
      "CFTR:\n",
      "Index(['Free Text', 'Protein Change NP_000483.3',\n",
      "       'Chromosome Position Change NC_000007.14',\n",
      "       'Gene Position Change NG_016465.4 (plus)', 'rsID', 'standard', 'E56K',\n",
      "       'P67L', 'R74W', 'D110E', 'D110H', 'R117C', 'R117H', 'G178R', 'E193K',\n",
      "       '711+3A->G', 'L206W', 'R347H', 'R352Q', 'A455E', 'S549N', 'S549R(A>C)',\n",
      "       'S549R(T>G)', 'G551S', 'G551D', 'D579G', 'E831X', '2789+5G->A', 'S945L',\n",
      "       'S977F', '3272-26A->G', 'F1052V', 'K1060T', 'A1067T', 'G1069R',\n",
      "       'R1070Q', 'R1070W', 'F1074L', 'D1152H', '3849+10kbC->T', 'G1244E',\n",
      "       'S1251N', 'S1255P', 'D1270N', 'G1349D'],\n",
      "      dtype='object')\n",
      "CYP2A6:\n",
      "Index(['Haplotype Set ID', 'Haplotype Set Name', 'rsID', '*1A', '*1B1', '*1B2',\n",
      "       '*1B3', '*1B4', '*1B5', '*1B6', '*1B7', '*1B8', '*1B9', '*1B10',\n",
      "       '*1B11', '*1B12', '*1B13', '*1B14', '*1B15', '*1B16', '*1B17', '*1C',\n",
      "       '*1D', '*1E', '*1F', '*1G', '*1H', '*1J', '*1K', '*1L', '*1X2A',\n",
      "       '*1X2B', '*2', '*3', '*4A', '*4B', '*4C', '*4D', '*4E', '*4F', '*4G',\n",
      "       '*4H', '*5', '*6', '*7', '*8', '*9', '*9A', '*9B', '*10', '*11', '*12',\n",
      "       '*12A', '*12B', '*12C', '*13', '*14', '*15', '*16', '*17', '*18',\n",
      "       '*18A', '*18B', '*18C', '*19', '*20', '*21', '*22', '*23', '*24A',\n",
      "       '*24B', '*25', '*26', '*27', '*28A', '*28B ', '*31A', '*31B', '*34',\n",
      "       '*35', '*35A', '*35B', '*36', '*37', '*38', '*39', '*40', '*41', '*42',\n",
      "       '*43', '*44', '*45'],\n",
      "      dtype='object')\n",
      "CYP2B6:\n",
      "Index(['Nucleotide change to gene', 'Effect on protein (NP_000758.1)',\n",
      "       'Position at NC_000019.10 (Homo sapiens chromosome 19, GRCh38.p2)',\n",
      "       'Position at NG_007929.1 (CYP2B6 RefSeqGene; forward relative to chromosome)',\n",
      "       'rsID', '*1', '*2', '*3', '*4', '*5', '*6', '*7', '*8', '*9', '*10',\n",
      "       '*11', '*12', '*13', '*14', '*15', '*17', '*18', '*19', '*20', '*21',\n",
      "       '*22', '*23', '*24', '*25', '*26', '*27', '*28', '*29', '*30', '*31',\n",
      "       '*32', '*33', '*34', '*35', '*36', '*37', '*38'],\n",
      "      dtype='object')\n",
      "CYP2C19:\n",
      "Index(['Nucleotide change to gene', 'Effect on protein (NP_000760.1)',\n",
      "       'Position at NC_000010.11 (Homo sapiens chromosome 10, GRCh38.p2)',\n",
      "       'Position at NG_008384.3 (CYP2C19 RefSeqGene; forward relative to chromosome)',\n",
      "       'rsID', '*38', '*1', '*2', '*3', '*4', '*5', '*6', '*7', '*8', '*9',\n",
      "       '*10', '*11', '*12', '*13', '*14', '*15', '*16', '*17', '*18', '*19',\n",
      "       '*22', '*23', '*24', '*25', '*26', '*28', '*29', '*30', '*31', '*32',\n",
      "       '*33', '*34', '*35', '*36', '*37', '*39'],\n",
      "      dtype='object')\n",
      "CYP2C9:\n",
      "Index(['Nucleotide change per gene', 'Effect on protein (NP_000762.2)',\n",
      "       'Position at NC_000010.11 (Homo sapiens chromosome 10, GRCh38.p2)',\n",
      "       'Position at NG_008385.2 (CYP2C9 RefSeqGene)', 'rsID', '*1', '*2', '*3',\n",
      "       '*4', '*5', '*6', '*7', '*8', '*9', '*10', '*11', '*12', '*13', '*14',\n",
      "       '*15', '*16', '*17', '*18', '*19', '*20', '*21', '*22', '*23', '*24',\n",
      "       '*25', '*26', '*27', '*28', '*29', '*30', '*31', '*32', '*33', '*34',\n",
      "       '*35', '*36', '*37', '*38', '*39', '*40', '*41', '*42', '*43', '*44',\n",
      "       '*45', '*46', '*47', '*48', '*49', '*50', '*51', '*52', '*53', '*54',\n",
      "       '*55', '*56', '*57', '*58', '*59', '*60', '*61', '*62', '*63', '*64',\n",
      "       '*65', '*66', '*67', '*68', '*69', '*70', '*71'],\n",
      "      dtype='object')\n",
      "CYP2D6:\n",
      "Index(['Nucleotide changes NG_008376.3 starting at ATG',\n",
      "       'Effect on protein (NP_000097.3)',\n",
      "       'Position at NC_000022.11 (Homo sapiens chromosome 22, GRCh38.p2)',\n",
      "       'Position at NG_008376.4 (CYP2D6 RefSeqGene; reverse relative to chromosome)',\n",
      "       'rsID', '*1', '*2', '*3', '*4', '*5',\n",
      "       ...\n",
      "       '*136', '*137', '*138', '*139', '*140', '*141', '*142', '*143', '*144',\n",
      "       '*145'],\n",
      "      dtype='object', length=142)\n",
      "CYP3A4:\n",
      "Index(['Haplotype Set ID', 'Haplotype Set Name', 'rsID', '*1', '*1A', '*1B',\n",
      "       '*1C', '*1D', '*1E', '*1F', '*1G', '*1H', '*1J', '*1K', '*1L', '*1M',\n",
      "       '*1N', '*1P', '*1Q', '*1R', '*1S', '*1T', '*2', '*3', '*4', '*5', '*6',\n",
      "       '*7', '*8', '*9', '*10', '*11', '*12', '*13', '*14', '*15', '*15A',\n",
      "       '*15B', '*16', '*16A', '*16B', '*17', '*18', '*18A', '*18B', '*19',\n",
      "       '*20', '*21', '*22', '*23', '*24'],\n",
      "      dtype='object')\n",
      "CYP3A5:\n",
      "Index(['Nucleotide change per gene', 'Effect on protein (NP_000768.1)',\n",
      "       'Position at NC_000007.14 (Homo sapiens chromosome 7, GRCh38.p2)',\n",
      "       'Position at NG_007938.2 (CYP3A5 RefSeqGene)', 'rsID', '*1', '*2', '*3',\n",
      "       '*4', '*5', '*6', '*7', '*8', '*9'],\n",
      "      dtype='object')\n",
      "DPYD:\n",
      "Index(['Free Text', 'Protein Change NP_000101.2',\n",
      "       'Chromosome Position Change NC_000001.11',\n",
      "       'Gene Position Change NG_008807.2 (minus)', 'rsID', 'Reference',\n",
      "       'c.46C>G', 'c.61C>T', 'c.62G>A', 'c.85T>C (*9A)',\n",
      "       ...\n",
      "       '', '', '', '', '', '', '', '', '', ''],\n",
      "      dtype='object', length=975)\n",
      "G6PD:\n",
      "Index(['Nucleotide NM_001042351.2', 'Effect on protein (c)',\n",
      "       'Position at NC_000023.11 (Homo sapiens chromosome X, GRCh38.p2)',\n",
      "       'Position at NG_009015.2 (G6PD RefSeqGene; reverse relative to chromosome)',\n",
      "       'rsID', 'B (wildtype)', 'No name', 'Sinnai', 'Lages', 'Gaohe',\n",
      "       ...\n",
      "       'Yunan', 'Kaiping, Anant, Dhon, Sapporo-like, Wosera', 'Neapolis',\n",
      "       'Laibin', 'Split', 'Fukaya', 'Campinas', 'Buenos Aires', 'Arakawa',\n",
      "       'Brighton'],\n",
      "      dtype='object', length=192)\n",
      "IFNL3:\n",
      "Index(['Haplotype Set ID', 'Haplotype Set Name', 'rsID', 'rs12979860C',\n",
      "       'rs12979860T'],\n",
      "      dtype='object')\n",
      "MT-RNR1:\n",
      "Index(['Free Text', 'Chromosome Position Change NC_012920.1', 'Allele Name',\n",
      "       'Reference', '663A>G', '669T>C', '747A>G', '786G>A', '807A>G', '807A>C',\n",
      "       '827A>G', '839A>G', '896A>G', '930A>G', '951G>A', '960C>del', '961T>G',\n",
      "       '961T>del', '961T>del+Cn', '988G>A', '1095T>C', '1189T>C', '1243T>C',\n",
      "       '1438A>G', '1494C>T', '1520T>C', '1537C>T', '1555A>G', '1556C>T'],\n",
      "      dtype='object')\n",
      "NAT2:\n",
      "Index(['Haplotype Set ID', 'Haplotype Set Name', 'rsID', '*4', '*5', '*5A',\n",
      "       '*5B', '*5C', '*5D', '*5E', '*5F', '*5G', '*5H', '*5I', '*5J', '*5K',\n",
      "       '*5L', '*5M', '*5N', '*5O', '*5P', '*5Q', '*5R', '*5S', '*5T', '*5U',\n",
      "       '*5V', '*6', '*6A', '*6B', '*6C', '*6D', '*6E', '*6F', '*6G', '*6H',\n",
      "       '*6I', '*6J', '*6K', '*6L', '*6M', '*6N', '*6O', '*6P', '*6Q', '*6R',\n",
      "       '*6S', '*6T', '*7', '*7A', '*7B', '*7C', '*7D', '*7E', '*7F', '*7G',\n",
      "       '*10', '*11', '*11A', '*11B', '*12', '*12A', '*12B', '*12C', '*12D',\n",
      "       '*12E', '*12F', '*12G', '*12H', '*12I', '*12J', '*12K', '*12L', '*12M',\n",
      "       '*13', '*13A', '*13B', '*13C', '*14', '*14A', '*14B', '*14C', '*14D',\n",
      "       '*14E', '*14F', '*14G', '*14H', '*14I', '*14J', '*17', '*18', '*19',\n",
      "       '*20', '*21', '*22', '*23', '*24', '*25'],\n",
      "      dtype='object')\n",
      "NUDT15:\n",
      "Index(['Nucleotide change on NM_018283.3 (ATG start)',\n",
      "       'Effect on protein (NP_060753.1)',\n",
      "       'Position at NC_000013.11 (Homo sapiens chromosome 6, GRCh38.p7)',\n",
      "       'Position at NG_047021.1 (NUDT15 RefSeqGene)', 'rsID', '*1', '*2', '*3',\n",
      "       '*4', '*5', '*6', '*7', '*8', '*9', '*10', '*11', '*12', '*13', '*14',\n",
      "       '*15', '*16', '*17', '*18', '*19', '*20'],\n",
      "      dtype='object')\n",
      "RYR1:\n",
      "Index(['Free Text', 'Protein Change NP_000531.2',\n",
      "       'Chromosome Position Change NC_000019.10',\n",
      "       'Gene Position Change NG_008866.1 (plus)', 'rsID', 'Reference',\n",
      "       'c.103T>C', 'c.130C>T', 'c.487C>T', 'c.488G>T', 'c.742G>C', 'c.742G>A',\n",
      "       'c.982C>T', 'c.1021G>A', 'c.1021G>C', 'c.1201C>T', 'c.1209C>G',\n",
      "       'c.1565A>C', 'c.1589G>A', 'c.1597C>T', 'c.1598G>A', 'c.1654C>T',\n",
      "       'c.1840C>T', 'c.1841G>T', 'c.6487C>T', 'c.6488G>A', 'c.6502G>A',\n",
      "       'c.6617C>G', 'c.6617C>T', 'c.7007G>A', 'c.7042_7044delGAG', 'c.7048G>A',\n",
      "       'c.7063C>T', 'c.7124G>C', 'c.7282G>A', 'c.7300G>A', 'c.7304G>A',\n",
      "       'c.7354C>T', 'c.7360C>T', 'c.7361G>A', 'c.7372C>T', 'c.7373G>A',\n",
      "       'c.7522C>T', 'c.7522C>G', 'c.7523G>A', 'c.9310G>A', 'c.11969G>T',\n",
      "       'c.14387A>G', 'c.14477C>T', 'c.14497C>T', 'c.14512C>G', 'c.14545G>A',\n",
      "       'c.14582G>A', 'c.14693T>C'],\n",
      "      dtype='object')\n",
      "SLCO1B1:\n",
      "Index(['Common Name', 'Effect on protein (NP_006437.3)',\n",
      "       'Position at NC_000012.12 (Homo sapiens chromosome 12, GRCh38.p2)',\n",
      "       'Position at NG_011745.1 (SLCO1B1 RefSeqGene)', 'rsID', '*1A', '*1B',\n",
      "       '*2', '*3', '*4', '*5', '*6', '*7', '*8', '*9', '*10', '*11', '*12',\n",
      "       '*13', '*14', '*15', '*16', '*17', '*18', '*19', '*20', '*21', '*22',\n",
      "       '*23', '*24', '*25', '*26', '*27', '*28', '*29', '*30', '*31', '*32',\n",
      "       '*33', '*34', '*35', '*36'],\n",
      "      dtype='object')\n",
      "TPMT:\n",
      "Index(['Nucleotide change on cDNA', 'Effect on protein (NP_000358.1)',\n",
      "       'Position at NC_000006.12 (Homo sapiens chromosome 6, GRCh38.p2)',\n",
      "       'Position at NG_012137.2 (TPMT RefSeqGene)', 'rsID', '*1', '*2', '*3A',\n",
      "       '*3B', '*3C', '*4', '*5', '*6', '*7', '*8', '*9', '*10', '*11', '*12',\n",
      "       '*13', '*14', '*15', '*16', '*17', '*18', '*19', '*20', '*21', '*22',\n",
      "       '*23', '*24', '*25', '*26', '*27', '*28', '*29', '*30', '*31', '*32',\n",
      "       '*33', '*34', '*35', '*36', '*37', '*38', '*39', '*40', '*41', '*42',\n",
      "       '*43', '*44'],\n",
      "      dtype='object')\n",
      "UGT1A1:\n",
      "Index(['Nucleotide changes', 'Effect on protein (NP_000454.1)',\n",
      "       'Position at NC_000002.12 (Homo sapiens chromosome 2, GRCh38.p2)',\n",
      "       'Position at NG_002601.2 (UGT1A1 RefSeqGene)', 'rsID', '*1', '*6',\n",
      "       '*27', '*28', '*36', '*37', '*80', '*80+*28', '*80+*37'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "for key, value in sorted(gene_df_T_dict.items(), key=lambda x: x[0]):\n",
    "    print(\"{}:\".format(key))\n",
    "    print(value.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "44a0adf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'nucleotide change': ['88C>T'], 'Effect on protein (NP_060753.1)': ['L30V'], 'Position at NC_000013.11 (Homo sapiens chromosome 6, GRCh38.p7)': ['g.48037834C>T'], 'Position at NG_047021.1 (NUDT15 RefSeqGene)': ['g.5268C>T']}\n",
      "{'rsID': ['rs1800462'], 'nucleotide change': ['238G>C'], 'Effect on protein (NP_000358.1)': ['A80P'], 'Position at NC_000006.12 (Homo sapiens chromosome 6, GRCh38.p2)': ['g.18143724C>G'], 'Position at NG_012137.2 (TPMT RefSeqGene)': ['g.16420G>C']}\n",
      "{'rsID': ['rs200591577'], 'nucleotide change': ['205C>G'], 'Effect on protein (NP_000358.1)': ['L69V'], 'Position at NC_000006.12 (Homo sapiens chromosome 6, GRCh38.p2)': ['g.18147851G>C'], 'Position at NG_012137.2 (TPMT RefSeqGene)': ['g.12293C>G']}\n",
      "{'rsID': ['rs33973337', 'rs33980385', 'rs33926104', 'rs34284776'], 'nucleotide change': ['76A>T', '83A>G', '85C>A', '86G>C'], 'Effect on protein (NP_000758.1)': ['T26S', 'D28G', 'p.R29S', 'p.R29P'], 'Position at NC_000019.10 (Homo sapiens chromosome 19, GRCh38.p2)': ['g.40991381A>T', 'g.40991388A>G', 'g.40991390C>A', 'g.40991391G>C'], 'Position at NG_007929.1 (CYP2B6 RefSeqGene; forward relative to chromosome)': ['g.5083A>T', 'g.5090A>G', 'g.5092C>A', 'g.5093G>C']}\n",
      "{}\n",
      "{'rsID': ['rs5030868'], 'nucleotide change': ['c.563C>T'], 'Effect on protein (c)': ['p.S188F'], 'Position at NC_000023.11 (Homo sapiens chromosome X, GRCh38.p2)': ['g.154534419G>A'], 'Position at NG_009015.2 (G6PD RefSeqGene; reverse relative to chromosome)': ['g.18154C>T']}\n"
     ]
    }
   ],
   "source": [
    "# input haplotype, return mapping dict\n",
    "def haplotype_mapping(gene, h_type):\n",
    "    if gene not in genes:\n",
    "        return {}\n",
    "    \n",
    "    # standard type\n",
    "    if standard_haplotype_dict[gene] == h_type:\n",
    "        return {}\n",
    "    \n",
    "    # check header\n",
    "    if h_type not in gene_df_T_dict[gene].columns:\n",
    "        return {}\n",
    "    \n",
    "    h_type_list = list(gene_df_T_dict[gene][h_type].values)\n",
    "    \n",
    "    mapping_dict = {}\n",
    "    # rsID mapping\n",
    "    rsID_mapping_list = []\n",
    "    rsID_list = list(gene_df_T_dict[gene][\"rsID\"].values)\n",
    "    filter_rs_ID_list = list(filter(lambda x: x[1] != \"\" and \"rs\" in x[0], zip(rsID_list, h_type_list)))\n",
    "    if len(filter_rs_ID_list) != 0:\n",
    "        rsID_mapping_list = [x[0].strip() for x in filter_rs_ID_list]\n",
    "        mapping_dict[\"rsID\"] = rsID_mapping_list\n",
    "    \n",
    "    # nuleotide change mapping\n",
    "    nucleotide_mapping_list = []\n",
    "    # NG position mapping\n",
    "    NG_position_mapping_list = []\n",
    "    # NC position mapping\n",
    "    NC_position_mapping_list = []\n",
    "    # protein mapping\n",
    "    protein_mapping_list = []\n",
    "    for column in gene_df_T_dict[gene].columns:\n",
    "        if \"nucleotide\" in column.lower():\n",
    "            nucleotide_list = gene_df_T_dict[gene][column]\n",
    "            filter_nucleotide_list = list(filter(lambda x: x[1] != \"\" and x[0] != \"\", zip(nucleotide_list, h_type_list)))\n",
    "            if len(filter_nucleotide_list) != 0:\n",
    "                nucleotide_mapping_list = [x[0].strip() for x in filter_nucleotide_list]\n",
    "                mapping_dict[\"nucleotide change\"] = nucleotide_mapping_list\n",
    "            \n",
    "        if \"NG\" in column and \"position\" in column.lower():\n",
    "            NG_position_list = gene_df_T_dict[gene][column]\n",
    "            filter_NG_position_list = list(filter(lambda x: x[1] != \"\" and x[0] != \"\", zip(NG_position_list, h_type_list)))\n",
    "            if len(filter_NG_position_list) != 0:\n",
    "                NG_position_mapping_list = [x[0].strip() for x in filter_NG_position_list]\n",
    "                mapping_dict[column] = NG_position_mapping_list\n",
    "                \n",
    "        if \"NC\" in column and \"position\" in column.lower():\n",
    "            NC_position_list = gene_df_T_dict[gene][column]\n",
    "            filter_NC_position_list = list(filter(lambda x: x[1] != \"\" and x[0] != \"\", zip(NC_position_list, h_type_list)))\n",
    "            if len(filter_NC_position_list) != 0:\n",
    "                NC_position_mapping_list = [x[0].strip() for x in filter_NC_position_list]\n",
    "                mapping_dict[column] = NC_position_mapping_list\n",
    "                \n",
    "        if \"protein\" in column.lower():\n",
    "            protein_list = gene_df_T_dict[gene][column]\n",
    "            filter_protein_list = list(filter(lambda x: x[1] != \"\" and x[0] != \"\", zip(protein_list, h_type_list)))\n",
    "            if len(filter_protein_list) != 0:\n",
    "                protein_mapping_list = [x[0].strip() for x in filter_protein_list]\n",
    "                mapping_dict[column] = protein_mapping_list \n",
    "                \n",
    "    return mapping_dict\n",
    "            \n",
    "    \n",
    "print(haplotype_mapping(\"NUDT15\", \"*16\"))\n",
    "print(haplotype_mapping(\"TPMT\", \"*2\"))\n",
    "print(haplotype_mapping(\"TPMT\", \"*21\"))\n",
    "print(haplotype_mapping(\"CYP2B6\", \"*17\"))\n",
    "print(haplotype_mapping(\"CYP3A4\", \"*1D\")) # 没有rsID\n",
    "print(haplotype_mapping(\"G6PD\", \"Mediterranean, Dallas, Panama‚ Sassari, Cagliari, Birmingham\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d0d6b3b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "da0c50b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"json/select_variants.json\", \"r\") as f:\n",
    "    select_dict = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5730420e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6f1ea10",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
